diff --git a/.github/workflows/ghci.yml b/.github/workflows/ghci.yml
index 6c1c182..d4bb180 100644
--- a/.github/workflows/ghci.yml
+++ b/.github/workflows/ghci.yml
@@ -9,7 +9,7 @@ jobs:
   build:
     runs-on: ${{ matrix.os }}
     name: build(${{ matrix.msystem || matrix.docker_image || matrix.os }},
-      ${{ matrix.cc }}${{ matrix.api && ', ' }}${{ matrix.api }}${{ matrix.jobname_suffix && ', ' }}${{ matrix.jobname_suffix }})
+      ${{ matrix.cc }}${{ matrix.jobname_suffix && ', ' }}${{ matrix.jobname_suffix }})
     strategy:
       fail-fast: false
       matrix:
@@ -47,7 +47,7 @@ jobs:
             # various regression test fail, assumed due to older deps
             skip_tests: yes
           # MacOS Intel
-          - os: macos-13
+          - os: macos-15-intel
             cc: clang
           # MacOS arm64
           - os: macos-14
@@ -62,14 +62,10 @@ jobs:
           - os: windows-latest
             msystem: MINGW32
             cc: gcc
-            api: desktop
             shell: 'msys2 {0}'
-          # Add a best-effort build for UWP apps for Microsoft Store
           - os: windows-latest
             msystem: UCRT64
             cc: gcc
-            api: app
-            extra_cflags: -DWINAPI_FAMILY=WINAPI_FAMILY_APP -specs=/tmp/windowsapp.specs
             shell: 'msys2 {0}'
 
     defaults:
@@ -102,10 +98,10 @@ jobs:
             output_stream.write(f"count={num_cpus}\n")
 
       - name: checkout code
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
 
       - name: download test samples
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
         with:
           repository: libass/libass-tests
           path: ${{ env.ART_SAMPLES }}
@@ -221,14 +217,6 @@ jobs:
           echo "SANFLAGS=$flags"
           echo "flags=${flags}" >> $GITHUB_OUTPUT
 
-      - name: Customize compiler
-        if: matrix.api == 'app' && matrix.cc == 'gcc'
-        run: >
-          gcc -dumpspecs
-          | sed 's/-lmsvcrt/-lucrtapp/g; s/-lkernel32/-lwindowsapp/g;
-          s/-ladvapi32//g; s/-lshell32//g; s/-luser32//g'
-          > /tmp/windowsapp.specs
-
       - name: configure
         run: |
           export CC="${{ matrix.cc }}\
diff --git a/.github/workflows/meson.yml b/.github/workflows/meson.yml
index df1da09..472f931 100644
--- a/.github/workflows/meson.yml
+++ b/.github/workflows/meson.yml
@@ -15,8 +15,9 @@ jobs:
       fail-fast: false
       matrix:
         config:
-          - name: Windows (MSVC)
+          - name: Windows (MSVC x64)
             os: windows-latest
+            arch: x64
             msvc: true
             shell: 'pwsh'
             args: '-Ddefault_library=shared'
@@ -30,6 +31,17 @@ jobs:
             shell: 'msys2 {0}'
             args: '-Ddefault_library=static -Db_sanitize=address,undefined'
 
+          - name: Windows (CLANGARM64)
+            os: windows-11-arm
+            msystem: CLANGARM64
+            args: '-Ddefault_library=shared'
+            shell: 'msys2 {0}'
+
+          - name: Windows (CLANGARM64)
+            os: windows-11-arm
+            msystem: CLANGARM64
+            shell: 'msys2 {0}'
+
           - name: Windows (UCRT64)
             os: windows-latest
             msystem: UCRT64
@@ -58,18 +70,17 @@ jobs:
         shell: ${{ matrix.config.shell || 'bash' }}
 
     env:
-      VS: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise"
       ART_SAMPLES: ${{ github.workspace }}/ext_art-samples
       SCCACHE_DIR: ${{ github.workspace }}/.sccache
       SCCACHE_MAXSIZE: 200M
 
     steps:
-      - uses: actions/checkout@v4
+      - uses: actions/checkout@v6
         with:
           fetch-depth: '0'
 
       - name: Download test samples
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
         with:
           repository: libass/libass-tests
           path: ${{ env.ART_SAMPLES }}
@@ -98,12 +109,21 @@ jobs:
 
       - name: Restore cache
         if: matrix.config.msvc
-        uses: actions/cache/restore@v4
+        uses: actions/cache/restore@v5
         with:
           path: ${{ env.SCCACHE_DIR }}
-          key: windows-msvc-${{ steps.get_time.outputs.timestamp }}
+          key: windows-msvc-${{ matrix.config.arch }}-${{ steps.get_time.outputs.timestamp }}
           restore-keys: windows-msvc-
 
+      - name: Find Visual Studio
+        if: matrix.config.msvc
+        run: |
+          $vswhere = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe"
+          $vsPath = & $vswhere -latest -products * -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -property installationPath
+          if (-not $vsPath) { throw "Could not find a Visual Studio installation" }
+          "VS=$vsPath"
+          "VS=$vsPath" >> $env:GITHUB_ENV
+
       - name: Install MSVC dependecies with choco
         if: matrix.config.msvc
         # Install meson through pip to add it to PATH easily
@@ -158,11 +178,11 @@ jobs:
         run: cat build/meson-logs/testlog.txt
 
       - name: Save Cache
-        uses: actions/cache/save@v4
+        uses: actions/cache/save@v5
         if: always() && matrix.config.msvc
         with:
           path: ${{ env.SCCACHE_DIR }}
-          key: windows-msvc-${{ steps.get_time.outputs.timestamp }}
+          key: windows-msvc-${{ matrix.config.arch }}-${{ steps.get_time.outputs.timestamp }}
 
   build-muon:
     runs-on: ubuntu-latest
@@ -188,7 +208,7 @@ jobs:
                   fontconfig-dev libpng-dev git
 
       - name: checkout code
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
 
       - name: configure
         run: |
@@ -250,7 +270,7 @@ jobs:
                   libpng-dev meson ninja-build
 
       - name: checkout code
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
 
       - name: setup distdir
         run: |
diff --git a/.github/workflows/qemu-arch.yml b/.github/workflows/qemu-arch.yml
index 5e73f9a..4bb45f1 100644
--- a/.github/workflows/qemu-arch.yml
+++ b/.github/workflows/qemu-arch.yml
@@ -25,31 +25,22 @@ jobs:
     #         (x32 is no longer supported in GHA kernels)
     #         (amd64 and arm64 already covered in regular CI runners)
     # Big:    s390x
-    # Little: riscv64, mips64el
+    # Little: riscv64
     strategy:
       fail-fast: false
       matrix:
-        arch: [s390x, mips64el]
-        suite: [bookworm]
+        arch: [i386, riscv64, s390x]
+        suite: [trixie]
         chroot-rev: [1]
         confopts: [""]
-        cflags: [""]
         include:
           # i386 needs some SSE usage enabled to get a bit-identical result (even without ASM).
           # This is due to different (higher) precision of 387 floating math (default for <=i686).
           # SSE2 was introduced in 2000 by Intel and is present in both 32bit and 64bit
           # CPUs from both AMD and Intel since 2003. Our ASM needs at least SSE2 anyway.
           - arch: i386
-            suite: bookworm
-            # 24.04’s kernel doesn't enable all 32bit compat options leading to segfaults
-            host_os: 'ubuntu-22.04'
             cflags: '-msse -msse2 -mfpmath=sse'
             sanity: sane?
-            chroot-rev: 1
-          - arch: riscv64
-            suite: sid
-            port: no
-            chroot-rev: 1
 
     steps:
       - name: Prepare System
@@ -131,7 +122,7 @@ jobs:
       # use it to store our (compressed) chroot dirs
       # (cache is branch scoped)
       - name: Retrieve Cached Chroots
-        uses: actions/cache@v4
+        uses: actions/cache@v5
         id: cache
         with:
           path: /var/chroot/imgs
@@ -223,12 +214,12 @@ jobs:
 
 
       - name: Checkout libass
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
         with:
           path: '${{ steps.config.outputs.chr_dir }}/home/artci/workarea/libass'
 
       - name: Checkout tests
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
         with:
           repository: ${{ github.event.inputs.test_repo || 'libass/libass-tests'}}
           ref: ${{ github.event.inputs.test_ref }}
diff --git a/.github/workflows/vm.yml b/.github/workflows/vm.yml
index 036d12d..ad551fe 100644
--- a/.github/workflows/vm.yml
+++ b/.github/workflows/vm.yml
@@ -38,7 +38,7 @@ jobs:
 
           - name: 'ART (freebsd64-qemu, FreeBSD-14, amd64)'
             os_image: freebsd
-            os_version: 14.2
+            os_version: 14.3
             # The default FreeBSD env places the libiconv header from ports in the default include path,
             # but doesn't put the corresponding library into the default library path.
             # Either inlcuding the lib path (presumably done by ports) or
@@ -51,40 +51,66 @@ jobs:
 
           - name: 'ART (openbsd64-qemu, OpenBSD-7, amd64)'
             os_image: openbsd
-            os_version: 7.6
+            os_version: 7.8
             CC: 'clang'
             confflags: ''
             ASAN_OPTIONS: ''
             ART_REG_SKIP: 'font_nonunicode'
             AUTOCONF_VERSION: '2.72'
-            AUTOMAKE_VERSION: '1.16'
-
-          # No illumos in cross-platform-actions/action yet, see
-          #   https://github.com/cross-platform-actions/action/issues/20
-          #- name: 'ART (illumos32-qemu, OpenIndiana, i686)'
-          #  vbox: openindiana/hipster
-          #  CC: gcc -m32 -msse -msse2 -mfpmath=sse
-          #  confflags: '--host=i686-pc-solaris2.11'
-          #  ASAN_OPTIONS: ''
-          #  ART_REG_SKIP: 'font_nonunicode'
+            AUTOMAKE_VERSION: '1.18'
+
+          - name: 'ART (haiku64-qemu, Haiku, amd64)'
+            os_image: haiku
+            os_version: r1beta5
+            # no libubsan available, thus just SIGILL
+            CC: 'gcc -fsanitize=undefined,float-cast-overflow -fno-sanitize-recover=all -fsanitize-undefined-trap-on-error'
+
+          - name: 'ART(omnios64-qemu, illumos/OmniOS, amd64)'
+            os_image: omnios
+            os_version: r151056
+            # errors by default, but illumos has spec-incompliant iconv (const vs restrict)
+            CC: 'gcc -Wno-error=incompatible-pointer-types'
+            confenv: 'PKG_CONFIG_PATH=/opt/ooce/lib/64/pkgconfig'
+            ART_REG_SKIP: 'font_nonunicode'
+
     env:
       ASAN_OPTIONS: ${{ matrix.ASAN_OPTIONS }}
       AUTOCONF_VERSION: ${{ matrix.AUTOCONF_VERSION }}
       AUTOMAKE_VERSION: ${{ matrix.AUTOMAKE_VERSION }}
 
     steps:
+      # Available core count changes on runner type and over time, but GHA
+      # doesn’t easily expose this, so we’ll have to query this ourselves. Based on:
+      # https://github.com/orgs/community/discussions/25057#discussioncomment-5957241
+      - name: determine logical cpu core count
+        id: cpus
+        shell: python
+        run: |
+          import os
+          import multiprocessing
+
+          try:
+            num_cpus = len(os.sched_getaffinity(0))
+          except AttributeError:
+            num_cpus = multiprocessing.cpu_count()
+          print(f"num_cpus={num_cpus}")
+
+          output_file = os.environ["GITHUB_OUTPUT"]
+          with open(output_file, "a", encoding="utf-8") as output_stream:
+            output_stream.write(f"count={num_cpus}\n")
+
       - name: checkout libass
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
 
       - name: download test samples
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
         with:
           repository: ${{ github.event.inputs.test_repo || 'libass/libass-tests'}}
           ref: ${{ github.event.inputs.test_ref }}
           path: './libass-tests'
 
       - name: Install Packages
-        uses: cross-platform-actions/action@v0.28.0
+        uses: cross-platform-actions/action@v1.0.0
         with:
           operating_system: ${{ matrix.os_image }}
           version: ${{ matrix.os_version }}
@@ -112,33 +138,40 @@ jobs:
               openbsd)
                 # while OpenBSD natively uses doas,
                 # cross-platform-action only preconfigured sudo.
-                # Using system-provided: compiler, fontconfig, freetype
+                # Using system-provided: compiler, fontconfig, freetype, pkgconf
                 sudo pkg_add \
-                  git nasm libtool pkgconf \
+                  git nasm libtool \
                   autoconf-"$AUTOCONF_VERSION"p0 \
-                  automake-"$AUTOMAKE_VERSION".5 \
+                  automake-"$AUTOMAKE_VERSION" \
                   fribidi harfbuzz libunibreak png
                 ;;
-              openindiana/*)
-                pfexec pkg install --no-backup-be \
-                    crypto/ca-certificates \
-                    developer/versioning/git \
-                    developer/gcc-10 \
-                    developer/assembler/nasm \
-                    developer/build/autoconf \
-                    developer/build/automake \
-                    developer/build/libtool \
-                    developer/build/pkg-config \
-                    system/library/freetype-2 \
-                    library/fribidi \
-                    library/c++/harfbuzz \
-                    system/library/fontconfig \
-                    image/library/libpng
+              haiku)
+                pkgman install -y \
+                  git nasm gcc make pkgconfig \
+                  libtool autoconf automake \
+                  freetype_devel fribidi_devel harfbuzz_devel \
+                  fontconfig_devel libunibreak_devel libpng16_devel
+                ;;
+              omnios)
+                # "ooce" extra packages are not in regular search paths.
+                # Activate via /opt/ooce/lib/64/pkgconfig etc.
+                # Harfbuzz and fribidi are bundled with pango (no standalone package).
+                sudo pkg install --deny-new-be --no-backup-be \
+                  pkg:/developer/gcc14 \
+                  pkg:/developer/nasm \
+                  pkg:/developer/build/make \
+                  pkg:/developer/pkg-config \
+                  pkg:/developer/build/autoconf \
+                  pkg:/developer/build/automake \
+                  pkg:/developer/build/libtool \
+                  pkg:/ooce/library/fontconfig \
+                  pkg:/ooce/library/pango \
+                  pkg:/ooce/library/libpng
                 ;;
             esac
 
       - name: Build libass
-        uses: cross-platform-actions/action@v0.28.0
+        uses: cross-platform-actions/action@v1.0.0
         with:
           operating_system: ${{ matrix.os_image }}
           version: ${{ matrix.os_version }}
@@ -147,11 +180,12 @@ jobs:
           sync_files: true
           run: |
             ./autogen.sh
-            ./configure CC="${{ matrix.CC }}" ${{ matrix.confflags }} --enable-fuzz --enable-compare
-            make -j 2
+            env ${{ matrix.confenv }} \
+              ./configure CC="${{ matrix.CC }}" ${{ matrix.confflags }} --enable-fuzz --enable-compare
+            make -j '${{ steps.cpus.outputs.count }}'
 
       - name: Run Tests
-        uses: cross-platform-actions/action@v0.28.0
+        uses: cross-platform-actions/action@v1.0.0
         with:
           operating_system: ${{ matrix.os_image }}
           version: ${{ matrix.os_version }}
@@ -165,7 +199,7 @@ jobs:
 
       - name: Shutdown VM
         if: always()
-        uses: cross-platform-actions/action@v0.28.0
+        uses: cross-platform-actions/action@v1.0.0
         with:
           operating_system: ${{ matrix.os_image }}
           version: ${{ matrix.os_version }}
diff --git a/Changelog b/Changelog
index 66e8083..79bec30 100644
--- a/Changelog
+++ b/Changelog
@@ -1,3 +1,15 @@
+libass (0.17.5)
+ * Fix limited OOB read and write in wrap_lines_measure (GHSA-pjjp-65r7-ppgm; CVE pending)
+ * Fix OOB bit clears for negative Matroska ReadOrder fields (GHSA-5gf7-wjfm-vmvm; CVE pending)
+ * Fix \fay with glyph clusters
+ * Fix small alpha changes not always splitting runs when combined with fade
+ * Fix compilation with MSVC-mode clang
+ * Fades are now applied to BorderStyle=4 boxes too
+ * Fonts using legacy arabic Windows charmaps are now supported
+ * ass_render_frame no longer returns fully transparent images
+ * Avoid MSVC’s subpar code generation for isnan to bring performance closer to other compilers
+ * Avoid SSE instructions if compiler baseline already includes AVX
+
 libass (0.17.4)
  * add new API to prune old events from memory
    * ass_prune_events for manual pruning
diff --git a/Makefile.am b/Makefile.am
index 208f860..2287369 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -15,6 +15,11 @@ AM_CFLAGS = -std=gnu99 -D_GNU_SOURCE -D_XPLATFORM_SOURCE
 
 EXTRA_DIST = libass.pc.in Changelog MAINTAINERS RELEASEVERSION ltnasm.sh
 
+# Source of pre-generated C files
+EXTRA_DIST += legacy_arabic_cmap/ArabicPUASimplified.txt \
+              legacy_arabic_cmap/ArabicPUATraditional.txt \
+              legacy_arabic_cmap/gen_arabic_charmap.py
+
 # Meson build setup
 EXTRA_DIST += gen_defs.py meson_options.txt meson.build \
               libass/meson.build libass/ass/meson.build \
diff --git a/README.md b/README.md
index f3715c3..889bb6e 100644
--- a/README.md
+++ b/README.md
@@ -4,7 +4,7 @@ libass is a portable subtitle renderer for the ASS/SSA (Advanced Substation Alph
 
 Get it
 ======
-See [GitHub releases](https://github.com/libass/libass/releases) for the latest release 0.17.4 (released 2025-06-07).
+See [GitHub releases](https://github.com/libass/libass/releases) for the latest release 0.17.5 (released 2026-06-24).
 See the [changelog](https://github.com/libass/libass/blob/master/Changelog) for a detailed list of changes.
 
 Source code is available from our [GitHub repository](https://github.com/libass/libass).
diff --git a/RELEASEVERSION b/RELEASEVERSION
index 44e33a4..8b5334d 100644
--- a/RELEASEVERSION
+++ b/RELEASEVERSION
@@ -1 +1 @@
-0.17.4
+0.17.5
diff --git a/checkasm/meson.build b/checkasm/meson.build
index e9a79d0..c8d8acf 100644
--- a/checkasm/meson.build
+++ b/checkasm/meson.build
@@ -27,6 +27,7 @@ libass_checkasm = executable(
     include_directories: incs,
     dependencies: deps,
     objects: libass.extract_all_objects(recursive: true),
+    link_with: libass_link_with,
     c_args: asm_args,
     build_by_default: false,
 )
diff --git a/compare/meson.build b/compare/meson.build
index b3edafc..4fbf90c 100644
--- a/compare/meson.build
+++ b/compare/meson.build
@@ -10,6 +10,7 @@ libass_compare = executable(
     include_directories: incs,
     dependencies: deps + png_deps,
     objects: libass.extract_all_objects(recursive: true),
+    link_with: libass_link_with,
 )
 
 art_samples = get_option('art-samples')
diff --git a/configure.ac b/configure.ac
index afd6350..27676b3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -325,6 +325,13 @@ AS_IF([test "x$enable_asm" != xno], [
                         ))
                     ]
                 )
+
+                # Convert SSE asm into (128-bit) AVX when compiler baseline is set to use AVX.
+                # Switching between AVX and SSE instructions can incur penalties.
+                AC_CHECK_DECL([__AVX__], [
+                    ASFLAGS="$ASFLAGS -DFORCE_VEX_ENCODING=1"
+                ])
+
                 AC_MSG_CHECKING([if $AS supports vpmovzxwd])
                 echo "vpmovzxwd ymm0, xmm0" > conftest.asm
                 AS_IF([$AS conftest.asm $ASFLAGS -o conftest.o >conftest.log 2>&1], [
diff --git a/fuzz/fuzz.c b/fuzz/fuzz.c
index 19b3432..db9e105 100644
--- a/fuzz/fuzz.c
+++ b/fuzz/fuzz.c
@@ -47,6 +47,9 @@
     #define LEN_IN_RANGE(s) true
 #endif
 
+#define STR_(x) #x
+#define STR(x) STR_(x)
+
 // MSAN: will trigger MSAN if any pixel in bitmap not written to (costly)
 #ifndef ASSFUZZ_HASH_WHOLEBITMAP
     #define ASSFUZZ_HASH_WHOLEBITMAP 0
@@ -356,6 +359,10 @@ int main(int argc, char *argv[])
     ssize_t len;
     unsigned char *buf;
 
+#ifdef ASSFUZZ_FONTCONFIG_SYSROOT
+    setenv("FONTCONFIG_SYSROOT", STR(ASSFUZZ_FONTCONFIG_SYSROOT), 1);
+#endif
+
     if (!init()) {
         printf("library init failed!\n");
         return 1;
@@ -390,6 +397,14 @@ int main(int argc, char *argv[])
     return 0;
 }
 #elif ASS_FUZZMODE == FUZZMODE_LIBFUZZER
+int LLVMFuzzerInitialize(int *argc, char ***argv)
+{
+#ifdef ASSFUZZ_FONTCONFIG_SYSROOT
+    setenv("FONTCONFIG_SYSROOT", STR(ASSFUZZ_FONTCONFIG_SYSROOT), 1);
+#endif
+    return 0;
+}
+
 int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
 {
     // OSS Fuzz docs recommend just returning 0 on too large input
@@ -402,9 +417,8 @@ int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
 
     ASS_Track *track = NULL;
 
-    // All return values but zero and -1 are reserved
     if (!init())
-        return 0;
+        exit(1);
 
     track = ass_read_memory(ass_library, (char *)data, size, NULL);
     if (track) {
diff --git a/fuzz/meson.build b/fuzz/meson.build
index 6d87e58..0cd01e8 100644
--- a/fuzz/meson.build
+++ b/fuzz/meson.build
@@ -3,6 +3,8 @@ fuzz_src = files(
     'writeout.c',
 )
 
+add_languages(get_option('fuzz-link-language'), native: false)
+
 libass_fuzz = executable(
     'fuzz',
     fuzz_src,
diff --git a/fuzz/writeout.c b/fuzz/writeout.c
index b6dd0ce..3981e59 100644
--- a/fuzz/writeout.c
+++ b/fuzz/writeout.c
@@ -138,8 +138,9 @@ static void write_styles(FILE *f, ASS_Track *track)
                    "%.3f,%.3f,%.3f,%.3f,%d,%.3f,%.3f,%d,%d,%d,%d,%d\n",
             s->Name, s->FontName, s->FontSize, s->PrimaryColour, s->SecondaryColour,
             s->OutlineColour, s->BackColour, s->Bold, s->Italic, s->Underline, s->StrikeOut,
-            s->ScaleX, s->ScaleY, s->Spacing, s->Angle, s->BorderStyle, s->Outline, s->Shadow,
-            ssa2ass_align(s->Alignment), s->MarginL, s->MarginR, s->MarginV, s->Encoding);
+            s->ScaleX * 100, s->ScaleY * 100, s->Spacing, s->Angle, s->BorderStyle,
+            s->Outline, s->Shadow, ssa2ass_align(s->Alignment),
+            s->MarginL, s->MarginR, s->MarginV, s->Encoding);
     }
 }
 
@@ -163,8 +164,11 @@ static void write_events(FILE *f, ASS_Track *track)
             continue;
         }
         fprintf(f, "Dialogue: %d,%s,%s,%s,%s,%03d,%03d,%03d,%s,%s\n",
-            e->Layer, start, end, (track->styles + e->Style)->Name, e->Name,
-            e->MarginL, e->MarginR, e->MarginV, e->Effect, e->Text);
+            e->Layer, start, end, (track->styles + e->Style)->Name,
+            e->Name ? e->Name : "",
+            e->MarginL, e->MarginR, e->MarginV,
+            e->Effect ? e->Effect : "",
+            e->Text);
     }
 }
 
diff --git a/legacy_arabic_cmap/ArabicPUASimplified.txt b/legacy_arabic_cmap/ArabicPUASimplified.txt
new file mode 100644
index 0000000..a74ef26
--- /dev/null
+++ b/legacy_arabic_cmap/ArabicPUASimplified.txt
@@ -0,0 +1,250 @@
+#
+#    Name:     Legacy Simplified Arabic encoding
+#
+#    Format: Three tab-separated columns
+#        Column #1 is the PUA code (in hex as 0xXXXX)
+#        Column #2 is the Unicode (in hex as 0xXXXX)
+#        Column #3 is the Unicode name (follows a comment sign, '#')
+#
+#    The entries are in PUA order
+#
+0xF100	0x063B	# ARABIC LETTER KEHEH WITH TWO DOTS ABOVE
+0xF100	0x063C	# ARABIC LETTER KEHEH WITH THREE DOTS BELOW
+0xF100	0x063D	# ARABIC LETTER FARSI YEH WITH INVERTED V
+0xF100	0x063E	# ARABIC LETTER FARSI YEH WITH TWO DOTS ABOVE
+0xF100	0x063F	# ARABIC LETTER FARSI YEH WITH THREE DOTS ABOVE
+0xF100	0x0653	# ARABIC MADDAH ABOVE
+0xF100	0x0654	# ARABIC HAMZA ABOVE
+0xF100	0x0655	# ARABIC HAMZA BELOW
+0xF100	0x0656	# ARABIC SUBSCRIPT ALEF
+0xF100	0x0657	# ARABIC INVERTED DAMMA
+0xF100	0x0658	# ARABIC MARK NOON GHUNNA
+0xF100	0x0659	# ARABIC ZWARAKAY
+0xF100	0x065A	# ARABIC VOWEL SIGN SMALL V ABOVE
+0xF100	0x065B	# ARABIC VOWEL SIGN INVERTED SMALL V ABOVE
+0xF100	0x065C	# ARABIC VOWEL SIGN DOT BELOW
+0xF100	0x065D	# ARABIC REVERSED DAMMA
+0xF100	0x065E	# ARABIC FATHA WITH TWO DOTS
+0xF10C	0x200C	# ZERO WIDTH NON-JOINER
+0xF10D	0x200D	# ZERO WIDTH JOINER
+0xF10E	0x200E	# LEFT-TO-RIGHT MARK
+0xF10F	0x200F	# RIGHT-TO-LEFT MARK
+0xF120	0x0020	# SPACE
+0xF121	0x0021	# EXCLAMATION MARK
+0xF122	0x0022	# QUOTATION MARK
+0xF123	0x00AB	# LEFT-POINTING DOUBLE ANGLE QUOTATION MARK
+0xF124	0x00BB	# RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK
+0xF125	0x0025	# PERCENT SIGN
+0xF126	0x00D7	# MULTIPLICATION SIGN
+0xF127	0x00F7	# DIVISION SIGN
+0xF128	0x0028	# LEFT PARENTHESIS
+0xF129	0x0029	# RIGHT PARENTHESIS
+0xF12A	0x002A	# ASTERISK
+0xF12B	0x002B	# PLUS SIGN
+0xF12C	0x060C	# ARABIC COMMA
+0xF12D	0x002D	# HYPHEN-MINUS
+0xF12E	0x002E	# FULL STOP
+0xF12F	0x002F	# SOLIDUS
+0xF130	0x0660	# ARABIC-INDIC DIGIT ZERO
+0xF131	0x0661	# ARABIC-INDIC DIGIT ONE
+0xF132	0x0662	# ARABIC-INDIC DIGIT TWO
+0xF133	0x0663	# ARABIC-INDIC DIGIT THREE
+0xF134	0x0664	# ARABIC-INDIC DIGIT FOUR
+0xF135	0x0665	# ARABIC-INDIC DIGIT FIVE
+0xF136	0x0666	# ARABIC-INDIC DIGIT SIX
+0xF137	0x0667	# ARABIC-INDIC DIGIT SEVEN
+0xF138	0x0668	# ARABIC-INDIC DIGIT EIGHT
+0xF139	0x0669	# ARABIC-INDIC DIGIT NINE
+0xF13A	0x003A	# COLON
+0xF13B	0x003B	# SEMICOLON
+0xF13B	0x061B	# ARABIC SEMICOLON
+0xF13C	0x2018	# LEFT SINGLE QUOTATION MARK
+0xF13D	0x003D	# EQUALS SIGN
+0xF13E	0x2019	# RIGHT SINGLE QUOTATION MARK
+0xF13F	0x003F	# QUESTION MARK
+0xF13F	0x061F	# ARABIC QUESTION MARK
+0xF141	0x0627	# ARABIC LETTER ALEF
+0xF141	0xFE8D	# ARABIC LETTER ALEF ISOLATED FORM
+0xF142	0xFE8E	# ARABIC LETTER ALEF FINAL FORM
+0xF143	0x0623	# ARABIC LETTER ALEF WITH HAMZA ABOVE
+0xF143	0xFE83	# ARABIC LETTER ALEF WITH HAMZA ABOVE ISOLATED FORM
+0xF144	0xFE84	# ARABIC LETTER ALEF WITH HAMZA ABOVE FINAL FORM
+0xF145	0x0622	# ARABIC LETTER ALEF WITH MADDA ABOVE
+0xF145	0xFE81	# ARABIC LETTER ALEF WITH MADDA ABOVE ISOLATED FORM
+0xF146	0xFE82	# ARABIC LETTER ALEF WITH MADDA ABOVE FINAL FORM
+0xF147	0x0625	# ARABIC LETTER ALEF WITH HAMZA BELOW
+0xF147	0xFE87	# ARABIC LETTER ALEF WITH HAMZA BELOW ISOLATED FORM
+0xF148	0xFE88	# ARABIC LETTER ALEF WITH HAMZA BELOW FINAL FORM
+0xF149	0xFE91	# ARABIC LETTER BEH INITIAL FORM
+0xF149	0xFE92	# ARABIC LETTER BEH MEDIAL FORM
+0xF14A	0x0628	# ARABIC LETTER BEH
+0xF14A	0xFE8F	# ARABIC LETTER BEH ISOLATED FORM
+0xF14A	0xFE90	# ARABIC LETTER BEH FINAL FORM
+0xF14B	0xFE97	# ARABIC LETTER TEH INITIAL FORM
+0xF14B	0xFE98	# ARABIC LETTER TEH MEDIAL FORM
+0xF14C	0x062A	# ARABIC LETTER TEH
+0xF14C	0xFE95	# ARABIC LETTER TEH ISOLATED FORM
+0xF14C	0xFE96	# ARABIC LETTER TEH FINAL FORM
+0xF14D	0xFE9B	# ARABIC LETTER THEH INITIAL FORM
+0xF14D	0xFE9C	# ARABIC LETTER THEH MEDIAL FORM
+0xF14E	0x062B	# ARABIC LETTER THEH
+0xF14E	0xFE99	# ARABIC LETTER THEH ISOLATED FORM
+0xF14E	0xFE9A	# ARABIC LETTER THEH FINAL FORM
+0xF14F	0xFE9F	# ARABIC LETTER JEEM INITIAL FORM
+0xF14F	0xFEA0	# ARABIC LETTER JEEM MEDIAL FORM
+0xF150	0xFE9E	# ARABIC LETTER JEEM FINAL FORM
+0xF151	0x062C	# ARABIC LETTER JEEM
+0xF151	0xFE9D	# ARABIC LETTER JEEM ISOLATED FORM
+0xF152	0xFEA3	# ARABIC LETTER HAH INITIAL FORM
+0xF152	0xFEA4	# ARABIC LETTER HAH MEDIAL FORM
+0xF153	0xFEA2	# ARABIC LETTER HAH FINAL FORM
+0xF154	0x062D	# ARABIC LETTER HAH
+0xF154	0xFEA1	# ARABIC LETTER HAH ISOLATED FORM
+0xF155	0xFEA7	# ARABIC LETTER KHAH INITIAL FORM
+0xF155	0xFEA8	# ARABIC LETTER KHAH MEDIAL FORM
+0xF156	0xFEA6	# ARABIC LETTER KHAH FINAL FORM
+0xF157	0x062E	# ARABIC LETTER KHAH
+0xF157	0xFEA5	# ARABIC LETTER KHAH ISOLATED FORM
+0xF158	0x062F	# ARABIC LETTER DAL
+0xF158	0xFEA9	# ARABIC LETTER DAL ISOLATED FORM
+0xF158	0xFEAA	# ARABIC LETTER DAL FINAL FORM
+0xF159	0x0630	# ARABIC LETTER THAL
+0xF159	0xFEAB	# ARABIC LETTER THAL ISOLATED FORM
+0xF159	0xFEAC	# ARABIC LETTER THAL FINAL FORM
+0xF15A	0x0631	# ARABIC LETTER REH
+0xF15A	0xFEAD	# ARABIC LETTER REH ISOLATED FORM
+0xF15A	0xFEAE	# ARABIC LETTER REH FINAL FORM
+0xF15B	0x005B	# LEFT SQUARE BRACKET
+0xF15C	0x005C	# REVERSE SOLIDUS
+0xF15D	0x005D	# RIGHT SQUARE BRACKET
+0xF15E	0x002C	# COMMA
+0xF15E	0x066B	# ARABIC DECIMAL SEPARATOR
+0xF15E	0x066C	# ARABIC THOUSANDS SEPARATOR
+0xF15F	0x0640	# ARABIC TATWEEL
+0xF160	0x0632	# ARABIC LETTER ZAIN
+0xF160	0xFEAF	# ARABIC LETTER ZAIN ISOLATED FORM
+0xF160	0xFEB0	# ARABIC LETTER ZAIN FINAL FORM
+0xF161	0xFEB3	# ARABIC LETTER SEEN INITIAL FORM
+0xF161	0xFEB4	# ARABIC LETTER SEEN MEDIAL FORM
+0xF162	0x0633	# ARABIC LETTER SEEN
+0xF162	0xFEB1	# ARABIC LETTER SEEN ISOLATED FORM
+0xF162	0xFEB2	# ARABIC LETTER SEEN FINAL FORM
+0xF163	0xFEB7	# ARABIC LETTER SHEEN INITIAL FORM
+0xF163	0xFEB8	# ARABIC LETTER SHEEN MEDIAL FORM
+0xF164	0x0634	# ARABIC LETTER SHEEN
+0xF164	0xFEB5	# ARABIC LETTER SHEEN ISOLATED FORM
+0xF164	0xFEB6	# ARABIC LETTER SHEEN FINAL FORM
+0xF165	0xFEBB	# ARABIC LETTER SAD INITIAL FORM
+0xF165	0xFEBC	# ARABIC LETTER SAD MEDIAL FORM
+0xF166	0x0635	# ARABIC LETTER SAD
+0xF166	0xFEB9	# ARABIC LETTER SAD ISOLATED FORM
+0xF166	0xFEBA	# ARABIC LETTER SAD FINAL FORM
+0xF167	0xFEBF	# ARABIC LETTER DAD INITIAL FORM
+0xF167	0xFEC0	# ARABIC LETTER DAD MEDIAL FORM
+0xF168	0x0636	# ARABIC LETTER DAD
+0xF168	0xFEBD	# ARABIC LETTER DAD ISOLATED FORM
+0xF168	0xFEBE	# ARABIC LETTER DAD FINAL FORM
+0xF169	0x0637	# ARABIC LETTER TAH
+0xF169	0xFEC1	# ARABIC LETTER TAH ISOLATED FORM
+0xF169	0xFEC2	# ARABIC LETTER TAH FINAL FORM
+0xF169	0xFEC3	# ARABIC LETTER TAH INITIAL FORM
+0xF169	0xFEC4	# ARABIC LETTER TAH MEDIAL FORM
+0xF16A	0x0638	# ARABIC LETTER ZAH
+0xF16A	0xFEC5	# ARABIC LETTER ZAH ISOLATED FORM
+0xF16A	0xFEC6	# ARABIC LETTER ZAH FINAL FORM
+0xF16A	0xFEC7	# ARABIC LETTER ZAH INITIAL FORM
+0xF16A	0xFEC8	# ARABIC LETTER ZAH MEDIAL FORM
+0xF16B	0xFECB	# ARABIC LETTER AIN INITIAL FORM
+0xF16C	0xFECC	# ARABIC LETTER AIN MEDIAL FORM
+0xF16D	0xFECA	# ARABIC LETTER AIN FINAL FORM
+0xF16E	0x0639	# ARABIC LETTER AIN
+0xF16E	0xFEC9	# ARABIC LETTER AIN ISOLATED FORM
+0xF16F	0xFECF	# ARABIC LETTER GHAIN INITIAL FORM
+0xF170	0xFED0	# ARABIC LETTER GHAIN MEDIAL FORM
+0xF171	0xFECE	# ARABIC LETTER GHAIN FINAL FORM
+0xF172	0x063A	# ARABIC LETTER GHAIN
+0xF172	0xFECD	# ARABIC LETTER GHAIN ISOLATED FORM
+0xF173	0xFED3	# ARABIC LETTER FEH INITIAL FORM
+0xF174	0xFED4	# ARABIC LETTER FEH MEDIAL FORM
+0xF175	0x0641	# ARABIC LETTER FEH
+0xF175	0xFED1	# ARABIC LETTER FEH ISOLATED FORM
+0xF175	0xFED2	# ARABIC LETTER FEH FINAL FORM
+0xF176	0xFED7	# ARABIC LETTER QAF INITIAL FORM
+0xF177	0xFED8	# ARABIC LETTER QAF MEDIAL FORM
+0xF178	0x0642	# ARABIC LETTER QAF
+0xF178	0xFED5	# ARABIC LETTER QAF ISOLATED FORM
+0xF178	0xFED6	# ARABIC LETTER QAF FINAL FORM
+0xF179	0xFEDB	# ARABIC LETTER KAF INITIAL FORM
+0xF179	0xFEDC	# ARABIC LETTER KAF MEDIAL FORM
+0xF17A	0x0643	# ARABIC LETTER KAF
+0xF17A	0xFED9	# ARABIC LETTER KAF ISOLATED FORM
+0xF17A	0xFEDA	# ARABIC LETTER KAF FINAL FORM
+0xF17B	0xFEDF	# ARABIC LETTER LAM INITIAL FORM
+0xF17B	0xFEE0	# ARABIC LETTER LAM MEDIAL FORM
+0xF17C	0x0644	# ARABIC LETTER LAM
+0xF17C	0xFEDD	# ARABIC LETTER LAM ISOLATED FORM
+0xF17C	0xFEDE	# ARABIC LETTER LAM FINAL FORM
+0xF17D	0xFEE3	# ARABIC LETTER MEEM INITIAL FORM
+0xF17D	0xFEE4	# ARABIC LETTER MEEM MEDIAL FORM
+0xF17E	0x0645	# ARABIC LETTER MEEM
+0xF17E	0xFEE1	# ARABIC LETTER MEEM ISOLATED FORM
+0xF17E	0xFEE2	# ARABIC LETTER MEEM FINAL FORM
+0xF17F	0xFEE7	# ARABIC LETTER NOON INITIAL FORM
+0xF17F	0xFEE8	# ARABIC LETTER NOON MEDIAL FORM
+0xF1A1	0xFEEB	# ARABIC LETTER HEH INITIAL FORM
+0xF1A2	0xFEEC	# ARABIC LETTER HEH MEDIAL FORM
+0xF1A3	0xFEEA	# ARABIC LETTER HEH FINAL FORM
+0xF1A4	0x0647	# ARABIC LETTER HEH
+0xF1A4	0xFEE9	# ARABIC LETTER HEH ISOLATED FORM
+0xF1A5	0x0648	# ARABIC LETTER WAW
+0xF1A5	0xFEED	# ARABIC LETTER WAW ISOLATED FORM
+0xF1A5	0xFEEE	# ARABIC LETTER WAW FINAL FORM
+0xF1A6	0xFEF3	# ARABIC LETTER YEH INITIAL FORM
+0xF1A6	0xFEF4	# ARABIC LETTER YEH MEDIAL FORM
+0xF1A7	0xFEF2	# ARABIC LETTER YEH FINAL FORM
+0xF1A8	0x064A	# ARABIC LETTER YEH
+0xF1A8	0xFEF1	# ARABIC LETTER YEH ISOLATED FORM
+0xF1A9	0x0629	# ARABIC LETTER TEH MARBUTA
+0xF1A9	0xFE93	# ARABIC LETTER TEH MARBUTA ISOLATED FORM
+0xF1AA	0xFE94	# ARABIC LETTER TEH MARBUTA FINAL FORM
+0xF1AB	0xFEF0	# ARABIC LETTER ALEF MAKSURA FINAL FORM
+0xF1AC	0x0649	# ARABIC LETTER ALEF MAKSURA
+0xF1AC	0xFEEF	# ARABIC LETTER ALEF MAKSURA ISOLATED FORM
+0xF1AD	0x0621	# ARABIC LETTER HAMZA
+0xF1AE	0xFE8B	# ARABIC LETTER YEH WITH HAMZA ABOVE INITIAL FORM
+0xF1AE	0xFE8C	# ARABIC LETTER YEH WITH HAMZA ABOVE MEDIAL FORM
+0xF1AF	0xFE8A	# ARABIC LETTER YEH WITH HAMZA ABOVE FINAL FORM
+0xF1B0	0x0030	# DIGIT ZERO
+0xF1B1	0x0031	# DIGIT ONE
+0xF1B2	0x0032	# DIGIT TWO
+0xF1B3	0x0033	# DIGIT THREE
+0xF1B4	0x0034	# DIGIT FOUR
+0xF1B5	0x0035	# DIGIT FIVE
+0xF1B6	0x0036	# DIGIT SIX
+0xF1B7	0x0037	# DIGIT SEVEN
+0xF1B8	0x0038	# DIGIT EIGHT
+0xF1B9	0x0039	# DIGIT NINE
+0xF1BA	0x0626	# ARABIC LETTER YEH WITH HAMZA ABOVE
+0xF1BA	0xFE89	# ARABIC LETTER YEH WITH HAMZA ABOVE ISOLATED FORM
+0xF1BB	0x0624	# ARABIC LETTER WAW WITH HAMZA ABOVE
+0xF1BB	0xFE85	# ARABIC LETTER WAW WITH HAMZA ABOVE ISOLATED FORM
+0xF1BB	0xFE86	# ARABIC LETTER WAW WITH HAMZA ABOVE FINAL FORM
+0xF1BC	0xFEFC	# ARABIC LIGATURE LAM WITH ALEF FINAL FORM
+0xF1BD	0xFEFB	# ARABIC LIGATURE LAM WITH ALEF ISOLATED FORM
+0xF1BE	0xFEF7	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA ABOVE ISOLATED FORM
+0xF1BF	0xFEF8	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA ABOVE FINAL FORM
+0xF1C0	0xFEF5	# ARABIC LIGATURE LAM WITH ALEF WITH MADDA ABOVE ISOLATED FORM
+0xF1C1	0xFEF6	# ARABIC LIGATURE LAM WITH ALEF WITH MADDA ABOVE FINAL FORM
+0xF1C2	0xFEF9	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA BELOW ISOLATED FORM
+0xF1C3	0xFEFA	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA BELOW FINAL FORM
+0xF1C4	0x064E	# ARABIC FATHA
+0xF1C5	0x064F	# ARABIC DAMMA
+0xF1C6	0x0652	# ARABIC SUKUN
+0xF1C7	0x064B	# ARABIC FATHATAN
+0xF1C8	0x064C	# ARABIC DAMMATAN
+0xF1C9	0x0651	# ARABIC SHADDA
+0xF1CA	0x0650	# ARABIC KASRA
+0xF1CB	0x064D	# ARABIC KASRATAN
+0xF1E1	0x0646	# ARABIC LETTER NOON
+0xF1E1	0xFEE5	# ARABIC LETTER NOON ISOLATED FORM
+0xF1E1	0xFEE6	# ARABIC LETTER NOON FINAL FORM
diff --git a/legacy_arabic_cmap/ArabicPUATraditional.txt b/legacy_arabic_cmap/ArabicPUATraditional.txt
new file mode 100644
index 0000000..5683b87
--- /dev/null
+++ b/legacy_arabic_cmap/ArabicPUATraditional.txt
@@ -0,0 +1,295 @@
+#
+#    Name:    Legacy Traditional Arabic encoding
+#
+#    Format: Three tab-separated columns
+#        Column #1 is the PUA code (in hex as 0xXXXX)
+#        Column #2 is the Unicode (in hex as 0xXXXX)
+#        Column #3 is the Unicode name (follows a comment sign, '#')
+#
+#    The entries are in PUA order
+#
+0xF200	0x063B	# ARABIC LETTER KEHEH WITH TWO DOTS ABOVE
+0xF200	0x063C	# ARABIC LETTER KEHEH WITH THREE DOTS BELOW
+0xF200	0x063D	# ARABIC LETTER FARSI YEH WITH INVERTED V
+0xF200	0x063E	# ARABIC LETTER FARSI YEH WITH TWO DOTS ABOVE
+0xF200	0x063F	# ARABIC LETTER FARSI YEH WITH THREE DOTS ABOVE
+0xF200	0x0653	# ARABIC MADDAH ABOVE
+0xF200	0x0654	# ARABIC HAMZA ABOVE
+0xF200	0x0655	# ARABIC HAMZA BELOW
+0xF200	0x0656	# ARABIC SUBSCRIPT ALEF
+0xF200	0x0657	# ARABIC INVERTED DAMMA
+0xF200	0x0658	# ARABIC MARK NOON GHUNNA
+0xF200	0x0659	# ARABIC ZWARAKAY
+0xF200	0x065A	# ARABIC VOWEL SIGN SMALL V ABOVE
+0xF200	0x065B	# ARABIC VOWEL SIGN INVERTED SMALL V ABOVE
+0xF200	0x065C	# ARABIC VOWEL SIGN DOT BELOW
+0xF200	0x065D	# ARABIC REVERSED DAMMA
+0xF200	0x065E	# ARABIC FATHA WITH TWO DOTS
+0xF202	0xFC08	# ARABIC LIGATURE BEH WITH MEEM ISOLATED FORM
+0xF203	0xFC0E	# ARABIC LIGATURE TEH WITH MEEM ISOLATED FORM
+0xF204	0xFC12	# ARABIC LIGATURE THEH WITH MEEM ISOLATED FORM
+0xF205	0xFC42	# ARABIC LIGATURE LAM WITH MEEM ISOLATED FORM
+0xF206	0xFC4E	# ARABIC LIGATURE NOON WITH MEEM ISOLATED FORM
+0xF20C	0x200C	# ZERO WIDTH NON-JOINER
+0xF20D	0x200D	# ZERO WIDTH JOINER
+0xF20E	0x200E	# LEFT-TO-RIGHT MARK
+0xF20F	0x200F	# RIGHT-TO-LEFT MARK
+0xF210	0xFD88	# ARABIC LIGATURE LAM WITH MEEM WITH HAH INITIAL FORM
+0xF212	0xFC3F	# ARABIC LIGATURE LAM WITH JEEM ISOLATED FORM
+0xF213	0xFC40	# ARABIC LIGATURE LAM WITH HAH ISOLATED FORM
+0xF214	0xFC41	# ARABIC LIGATURE LAM WITH KHAH ISOLATED FORM
+0xF215	0xFC6A	# ARABIC LIGATURE BEH WITH REH FINAL FORM
+0xF216	0xFC70	# ARABIC LIGATURE TEH WITH REH FINAL FORM
+0xF217	0xFC91	# ARABIC LIGATURE YEH WITH REH FINAL FORM
+0xF218	0xFCB0	# ARABIC LIGATURE SEEN WITH MEEM INITIAL FORM
+0xF219	0xFD30	# ARABIC LIGATURE SHEEN WITH MEEM INITIAL FORM
+0xF21A	0xFCCD	# ARABIC LIGATURE LAM WITH HEH INITIAL FORM
+0xF21C	0xFC44	# ARABIC LIGATURE LAM WITH YEH ISOLATED FORM
+0xF21D	0xFC0A	# ARABIC LIGATURE BEH WITH YEH ISOLATED FORM
+0xF21E	0xFC10	# ARABIC LIGATURE TEH WITH YEH ISOLATED FORM
+0xF21F	0xFC50	# ARABIC LIGATURE NOON WITH YEH ISOLATED FORM
+0xF220	0x0020	# SPACE
+0xF221	0x0021	# EXCLAMATION MARK
+0xF222	0x0022	# QUOTATION MARK
+0xF223	0x00AB	# LEFT-POINTING DOUBLE ANGLE QUOTATION MARK
+0xF224	0x00BB	# RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK
+0xF225	0x0025	# PERCENT SIGN
+0xF226	0x00D7	# MULTIPLICATION SIGN
+0xF227	0x00F7	# DIVISION SIGN
+0xF228	0x0028	# LEFT PARENTHESIS
+0xF229	0x0029	# RIGHT PARENTHESIS
+0xF22A	0x002A	# ASTERISK
+0xF22B	0x002B	# PLUS SIGN
+0xF22C	0x060C	# ARABIC COMMA
+0xF22D	0x002D	# HYPHEN-MINUS
+0xF22E	0x002E	# FULL STOP
+0xF22F	0x002F	# SOLIDUS
+0xF230	0x0660	# ARABIC-INDIC DIGIT ZERO
+0xF231	0x0661	# ARABIC-INDIC DIGIT ONE
+0xF232	0x0662	# ARABIC-INDIC DIGIT TWO
+0xF233	0x0663	# ARABIC-INDIC DIGIT THREE
+0xF234	0x0664	# ARABIC-INDIC DIGIT FOUR
+0xF235	0x0665	# ARABIC-INDIC DIGIT FIVE
+0xF236	0x0666	# ARABIC-INDIC DIGIT SIX
+0xF237	0x0667	# ARABIC-INDIC DIGIT SEVEN
+0xF238	0x0668	# ARABIC-INDIC DIGIT EIGHT
+0xF239	0x0669	# ARABIC-INDIC DIGIT NINE
+0xF23A	0x003A	# COLON
+0xF23B	0x003B	# SEMICOLON
+0xF23B	0x061B	# ARABIC SEMICOLON
+0xF23C	0x201C	# LEFT DOUBLE QUOTATION MARK
+0xF23D	0x003D	# EQUALS SIGN
+0xF23E	0x201D	# RIGHT DOUBLE QUOTATION MARK
+0xF23F	0x003F	# QUESTION MARK
+0xF23F	0x061F	# ARABIC QUESTION MARK
+0xF241	0x0627	# ARABIC LETTER ALEF
+0xF241	0xFE8D	# ARABIC LETTER ALEF ISOLATED FORM
+0xF242	0xFE8E	# ARABIC LETTER ALEF FINAL FORM
+0xF243	0x0623	# ARABIC LETTER ALEF WITH HAMZA ABOVE
+0xF243	0xFE83	# ARABIC LETTER ALEF WITH HAMZA ABOVE ISOLATED FORM
+0xF244	0xFE84	# ARABIC LETTER ALEF WITH HAMZA ABOVE FINAL FORM
+0xF245	0x0622	# ARABIC LETTER ALEF WITH MADDA ABOVE
+0xF245	0xFE81	# ARABIC LETTER ALEF WITH MADDA ABOVE ISOLATED FORM
+0xF246	0xFE82	# ARABIC LETTER ALEF WITH MADDA ABOVE FINAL FORM
+0xF247	0x0625	# ARABIC LETTER ALEF WITH HAMZA BELOW
+0xF247	0xFE87	# ARABIC LETTER ALEF WITH HAMZA BELOW ISOLATED FORM
+0xF248	0xFE88	# ARABIC LETTER ALEF WITH HAMZA BELOW FINAL FORM
+0xF249	0xFE91	# ARABIC LETTER BEH INITIAL FORM
+0xF24A	0xFE92	# ARABIC LETTER BEH MEDIAL FORM
+0xF24B	0xFE90	# ARABIC LETTER BEH FINAL FORM
+0xF24C	0x0628	# ARABIC LETTER BEH
+0xF24C	0xFE8F	# ARABIC LETTER BEH ISOLATED FORM
+0xF24D	0xFE97	# ARABIC LETTER TEH INITIAL FORM
+0xF24E	0xFE98	# ARABIC LETTER TEH MEDIAL FORM
+0xF24F	0xFE96	# ARABIC LETTER TEH FINAL FORM
+0xF250	0x062A	# ARABIC LETTER TEH
+0xF250	0xFE95	# ARABIC LETTER TEH ISOLATED FORM
+0xF251	0xFE9B	# ARABIC LETTER THEH INITIAL FORM
+0xF252	0xFE9C	# ARABIC LETTER THEH MEDIAL FORM
+0xF253	0xFE9A	# ARABIC LETTER THEH FINAL FORM
+0xF254	0x062B	# ARABIC LETTER THEH
+0xF254	0xFE99	# ARABIC LETTER THEH ISOLATED FORM
+0xF255	0xFE9F	# ARABIC LETTER JEEM INITIAL FORM
+0xF256	0xFEA0	# ARABIC LETTER JEEM MEDIAL FORM
+0xF257	0xFE9E	# ARABIC LETTER JEEM FINAL FORM
+0xF258	0x062C	# ARABIC LETTER JEEM
+0xF258	0xFE9D	# ARABIC LETTER JEEM ISOLATED FORM
+0xF259	0xFEA3	# ARABIC LETTER HAH INITIAL FORM
+0xF25A	0xFEA4	# ARABIC LETTER HAH MEDIAL FORM
+0xF25B	0x005B	# LEFT SQUARE BRACKET
+0xF25C	0xFEA2	# ARABIC LETTER HAH FINAL FORM
+0xF25D	0x005D	# RIGHT SQUARE BRACKET
+0xF25E	0x002C	# COMMA
+0xF25E	0x066B	# ARABIC DECIMAL SEPARATOR
+0xF25E	0x066C	# ARABIC THOUSANDS SEPARATOR
+0xF25F	0x0640	# ARABIC TATWEEL
+0xF260	0x062D	# ARABIC LETTER HAH
+0xF260	0xFEA1	# ARABIC LETTER HAH ISOLATED FORM
+0xF261	0xFEA7	# ARABIC LETTER KHAH INITIAL FORM
+0xF262	0xFEA8	# ARABIC LETTER KHAH MEDIAL FORM
+0xF263	0xFEA6	# ARABIC LETTER KHAH FINAL FORM
+0xF264	0x062E	# ARABIC LETTER KHAH
+0xF264	0xFEA5	# ARABIC LETTER KHAH ISOLATED FORM
+0xF265	0x062F	# ARABIC LETTER DAL
+0xF265	0xFEA9	# ARABIC LETTER DAL ISOLATED FORM
+0xF266	0xFEAA	# ARABIC LETTER DAL FINAL FORM
+0xF267	0x0630	# ARABIC LETTER THAL
+0xF267	0xFEAB	# ARABIC LETTER THAL ISOLATED FORM
+0xF268	0xFEAC	# ARABIC LETTER THAL FINAL FORM
+0xF269	0x0631	# ARABIC LETTER REH
+0xF269	0xFEAD	# ARABIC LETTER REH ISOLATED FORM
+0xF26A	0xFEAE	# ARABIC LETTER REH FINAL FORM
+0xF26B	0x0632	# ARABIC LETTER ZAIN
+0xF26B	0xFEAF	# ARABIC LETTER ZAIN ISOLATED FORM
+0xF26C	0xFEB0	# ARABIC LETTER ZAIN FINAL FORM
+0xF26D	0xFEB3	# ARABIC LETTER SEEN INITIAL FORM
+0xF26E	0xFEB4	# ARABIC LETTER SEEN MEDIAL FORM
+0xF26F	0xFEB2	# ARABIC LETTER SEEN FINAL FORM
+0xF270	0x0633	# ARABIC LETTER SEEN
+0xF270	0xFEB1	# ARABIC LETTER SEEN ISOLATED FORM
+0xF271	0xFEB7	# ARABIC LETTER SHEEN INITIAL FORM
+0xF272	0xFEB8	# ARABIC LETTER SHEEN MEDIAL FORM
+0xF273	0xFEB6	# ARABIC LETTER SHEEN FINAL FORM
+0xF274	0x0634	# ARABIC LETTER SHEEN
+0xF274	0xFEB5	# ARABIC LETTER SHEEN ISOLATED FORM
+0xF275	0xFEBB	# ARABIC LETTER SAD INITIAL FORM
+0xF276	0xFEBC	# ARABIC LETTER SAD MEDIAL FORM
+0xF277	0xFEBA	# ARABIC LETTER SAD FINAL FORM
+0xF278	0x0635	# ARABIC LETTER SAD
+0xF278	0xFEB9	# ARABIC LETTER SAD ISOLATED FORM
+0xF279	0xFEBF	# ARABIC LETTER DAD INITIAL FORM
+0xF27A	0xFEC0	# ARABIC LETTER DAD MEDIAL FORM
+0xF27B	0xFD3E	# ORNATE LEFT PARENTHESIS
+0xF27C	0xFEBE	# ARABIC LETTER DAD FINAL FORM
+0xF27D	0xFD3F	# ORNATE RIGHT PARENTHESIS
+0xF27E	0x0636	# ARABIC LETTER DAD
+0xF27E	0xFEBD	# ARABIC LETTER DAD ISOLATED FORM
+0xF27F	0xFEC3	# ARABIC LETTER TAH INITIAL FORM
+0xF280	0xFC9C	# ARABIC LIGATURE BEH WITH JEEM INITIAL FORM
+0xF281	0xFC9D	# ARABIC LIGATURE BEH WITH HAH INITIAL FORM
+0xF282	0xFC9E	# ARABIC LIGATURE BEH WITH KHAH INITIAL FORM
+0xF283	0xFCA1	# ARABIC LIGATURE TEH WITH JEEM INITIAL FORM
+0xF284	0xFCA2	# ARABIC LIGATURE TEH WITH HAH INITIAL FORM
+0xF285	0xFCA3	# ARABIC LIGATURE TEH WITH KHAH INITIAL FORM
+0xF286	0xFCC9	# ARABIC LIGATURE LAM WITH JEEM INITIAL FORM
+0xF287	0xFCCA	# ARABIC LIGATURE LAM WITH HAH INITIAL FORM
+0xF288	0xFCCB	# ARABIC LIGATURE LAM WITH KHAH INITIAL FORM
+0xF289	0xFCCE	# ARABIC LIGATURE MEEM WITH JEEM INITIAL FORM
+0xF28A	0xFCCF	# ARABIC LIGATURE MEEM WITH HAH INITIAL FORM
+0xF28B	0xFCD0	# ARABIC LIGATURE MEEM WITH KHAH INITIAL FORM
+0xF28D	0xFCD2	# ARABIC LIGATURE NOON WITH JEEM INITIAL FORM
+0xF28E	0xFCD3	# ARABIC LIGATURE NOON WITH HAH INITIAL FORM
+0xF28F	0xFCDA	# ARABIC LIGATURE YEH WITH JEEM INITIAL FORM
+0xF290	0xFCDB	# ARABIC LIGATURE YEH WITH HAH INITIAL FORM
+0xF291	0xFCDC	# ARABIC LIGATURE YEH WITH KHAH INITIAL FORM
+0xF292	0xFC6D	# ARABIC LIGATURE BEH WITH NOON FINAL FORM
+0xF293	0xFC73	# ARABIC LIGATURE TEH WITH NOON FINAL FORM
+0xF294	0xFC94	# ARABIC LIGATURE YEH WITH NOON FINAL FORM
+0xF295	0xFC43	# ARABIC LIGATURE LAM WITH ALEF MAKSURA ISOLATED FORM
+0xF296	0xFC9F	# ARABIC LIGATURE BEH WITH MEEM INITIAL FORM
+0xF297	0xFCA4	# ARABIC LIGATURE TEH WITH MEEM INITIAL FORM
+0xF298	0xFCD5	# ARABIC LIGATURE NOON WITH MEEM INITIAL FORM
+0xF299	0xFCDD	# ARABIC LIGATURE YEH WITH MEEM INITIAL FORM
+0xF29A	0xFCA8	# ARABIC LIGATURE JEEM WITH MEEM INITIAL FORM
+0xF29B	0xFCAA	# ARABIC LIGATURE HAH WITH MEEM INITIAL FORM
+0xF29C	0xFCAC	# ARABIC LIGATURE KHAH WITH MEEM INITIAL FORM
+0xF29D	0xFCCC	# ARABIC LIGATURE LAM WITH MEEM INITIAL FORM
+0xF29E	0xFCD1	# ARABIC LIGATURE MEEM WITH MEEM INITIAL FORM
+0xF29F	0xFC32	# ARABIC LIGATURE FEH WITH YEH ISOLATED FORM
+0xF2A1	0xFEC2	# ARABIC LETTER TAH FINAL FORM
+0xF2A2	0x0637	# ARABIC LETTER TAH
+0xF2A2	0xFEC1	# ARABIC LETTER TAH ISOLATED FORM
+0xF2A3	0x0638	# ARABIC LETTER ZAH
+0xF2A3	0xFEC7	# ARABIC LETTER ZAH INITIAL FORM
+0xF2A4	0xFEC8	# ARABIC LETTER ZAH MEDIAL FORM
+0xF2A5	0xFEC6	# ARABIC LETTER ZAH FINAL FORM
+0xF2A6	0xFEC5	# ARABIC LETTER ZAH ISOLATED FORM
+0xF2A7	0xFECB	# ARABIC LETTER AIN INITIAL FORM
+0xF2A8	0xFECC	# ARABIC LETTER AIN MEDIAL FORM
+0xF2A9	0xFECA	# ARABIC LETTER AIN FINAL FORM
+0xF2AA	0x0639	# ARABIC LETTER AIN
+0xF2AA	0xFEC9	# ARABIC LETTER AIN ISOLATED FORM
+0xF2AB	0xFECF	# ARABIC LETTER GHAIN INITIAL FORM
+0xF2AC	0xFED0	# ARABIC LETTER GHAIN MEDIAL FORM
+0xF2AD	0xFECE	# ARABIC LETTER GHAIN FINAL FORM
+0xF2AE	0x063A	# ARABIC LETTER GHAIN
+0xF2AE	0xFECD	# ARABIC LETTER GHAIN ISOLATED FORM
+0xF2AF	0xFED3	# ARABIC LETTER FEH INITIAL FORM
+0xF2B0	0xFED4	# ARABIC LETTER FEH MEDIAL FORM
+0xF2B1	0xFED2	# ARABIC LETTER FEH FINAL FORM
+0xF2B2	0x0641	# ARABIC LETTER FEH
+0xF2B2	0xFED1	# ARABIC LETTER FEH ISOLATED FORM
+0xF2B3	0xFED7	# ARABIC LETTER QAF INITIAL FORM
+0xF2B4	0xFED8	# ARABIC LETTER QAF MEDIAL FORM
+0xF2B5	0xFED6	# ARABIC LETTER QAF FINAL FORM
+0xF2B6	0x0642	# ARABIC LETTER QAF
+0xF2B6	0xFED5	# ARABIC LETTER QAF ISOLATED FORM
+0xF2B7	0xFEDB	# ARABIC LETTER KAF INITIAL FORM
+0xF2B8	0xFEDC	# ARABIC LETTER KAF MEDIAL FORM
+0xF2B9	0xFEDA	# ARABIC LETTER KAF FINAL FORM
+0xF2BA	0x0643	# ARABIC LETTER KAF
+0xF2BA	0xFED9	# ARABIC LETTER KAF ISOLATED FORM
+0xF2BB	0xFEDF	# ARABIC LETTER LAM INITIAL FORM
+0xF2BC	0xFEE0	# ARABIC LETTER LAM MEDIAL FORM
+0xF2BD	0xFEDE	# ARABIC LETTER LAM FINAL FORM
+0xF2BE	0x0644	# ARABIC LETTER LAM
+0xF2BE	0xFEDD	# ARABIC LETTER LAM ISOLATED FORM
+0xF2BF	0xFEE3	# ARABIC LETTER MEEM INITIAL FORM
+0xF2C0	0xFEE4	# ARABIC LETTER MEEM MEDIAL FORM
+0xF2C1	0xFEE2	# ARABIC LETTER MEEM FINAL FORM
+0xF2C2	0x0645	# ARABIC LETTER MEEM
+0xF2C2	0xFEE1	# ARABIC LETTER MEEM ISOLATED FORM
+0xF2C3	0xFEE7	# ARABIC LETTER NOON INITIAL FORM
+0xF2C4	0xFEE8	# ARABIC LETTER NOON MEDIAL FORM
+0xF2C5	0xFEE6	# ARABIC LETTER NOON FINAL FORM
+0xF2C6	0x0646	# ARABIC LETTER NOON
+0xF2C6	0xFEE5	# ARABIC LETTER NOON ISOLATED FORM
+0xF2C7	0xFEEB	# ARABIC LETTER HEH INITIAL FORM
+0xF2C8	0xFEEC	# ARABIC LETTER HEH MEDIAL FORM
+0xF2C9	0xFEEA	# ARABIC LETTER HEH FINAL FORM
+0xF2CA	0x0647	# ARABIC LETTER HEH
+0xF2CA	0xFEE9	# ARABIC LETTER HEH ISOLATED FORM
+0xF2CB	0x0648	# ARABIC LETTER WAW
+0xF2CB	0xFEED	# ARABIC LETTER WAW ISOLATED FORM
+0xF2CC	0xFEEE	# ARABIC LETTER WAW FINAL FORM
+0xF2CD	0xFEF3	# ARABIC LETTER YEH INITIAL FORM
+0xF2CE	0xFEF4	# ARABIC LETTER YEH MEDIAL FORM
+0xF2CF	0xFEF2	# ARABIC LETTER YEH FINAL FORM
+0xF2D0	0x064A	# ARABIC LETTER YEH
+0xF2D0	0xFEF1	# ARABIC LETTER YEH ISOLATED FORM
+0xF2D1	0x0629	# ARABIC LETTER TEH MARBUTA
+0xF2D1	0xFE93	# ARABIC LETTER TEH MARBUTA ISOLATED FORM
+0xF2D2	0xFE94	# ARABIC LETTER TEH MARBUTA FINAL FORM
+0xF2D3	0xFEF0	# ARABIC LETTER ALEF MAKSURA FINAL FORM
+0xF2D4	0x0649	# ARABIC LETTER ALEF MAKSURA
+0xF2D4	0xFEEF	# ARABIC LETTER ALEF MAKSURA ISOLATED FORM
+0xF2D5	0x0621	# ARABIC LETTER HAMZA
+0xF2D6	0xFE8B	# ARABIC LETTER YEH WITH HAMZA ABOVE INITIAL FORM
+0xF2D7	0xFE8C	# ARABIC LETTER YEH WITH HAMZA ABOVE MEDIAL FORM
+0xF2D8	0xFE8A	# ARABIC LETTER YEH WITH HAMZA ABOVE FINAL FORM
+0xF2D9	0x0626	# ARABIC LETTER YEH WITH HAMZA ABOVE
+0xF2D9	0xFE89	# ARABIC LETTER YEH WITH HAMZA ABOVE ISOLATED FORM
+0xF2DA	0x0624	# ARABIC LETTER WAW WITH HAMZA ABOVE
+0xF2DA	0xFE85	# ARABIC LETTER WAW WITH HAMZA ABOVE ISOLATED FORM
+0xF2DB	0xFE86	# ARABIC LETTER WAW WITH HAMZA ABOVE FINAL FORM
+0xF2DC	0xFEFB	# ARABIC LIGATURE LAM WITH ALEF ISOLATED FORM
+0xF2DD	0xFEFC	# ARABIC LIGATURE LAM WITH ALEF FINAL FORM
+0xF2DE	0xFEF7	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA ABOVE ISOLATED FORM
+0xF2DF	0xFEF8	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA ABOVE FINAL FORM
+0xF2E0	0xFEF5	# ARABIC LIGATURE LAM WITH ALEF WITH MADDA ABOVE ISOLATED FORM
+0xF2E1	0xFEF6	# ARABIC LIGATURE LAM WITH ALEF WITH MADDA ABOVE FINAL FORM
+0xF2E2	0xFEF9	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA BELOW ISOLATED FORM
+0xF2E3	0xFEFA	# ARABIC LIGATURE LAM WITH ALEF WITH HAMZA BELOW FINAL FORM
+0xF2E4	0x064E	# ARABIC FATHA
+0xF2E5	0x064F	# ARABIC DAMMA
+0xF2E6	0x0652	# ARABIC SUKUN
+0xF2E7	0x064B	# ARABIC FATHATAN
+0xF2E8	0x064C	# ARABIC DAMMATAN
+0xF2E9	0x0651	# ARABIC SHADDA
+0xF2EA	0x0650	# ARABIC KASRA
+0xF2EB	0x064D	# ARABIC KASRATAN
+0xF2EC	0xFC60	# ARABIC LIGATURE SHADDA WITH FATHA ISOLATED FORM
+0xF2ED	0xFC61	# ARABIC LIGATURE SHADDA WITH DAMMA ISOLATED FORM
+0xF2EF	0xFC5E	# ARABIC LIGATURE SHADDA WITH DAMMATAN ISOLATED FORM
+0xF2F0	0xFC62	# ARABIC LIGATURE SHADDA WITH KASRA ISOLATED FORM
+0xF2F1	0xFEC4	# ARABIC LETTER TAH MEDIAL FORM
diff --git a/legacy_arabic_cmap/gen_arabic_charmap.py b/legacy_arabic_cmap/gen_arabic_charmap.py
new file mode 100644
index 0000000..8091a33
--- /dev/null
+++ b/legacy_arabic_cmap/gen_arabic_charmap.py
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Self
+import textwrap
+
+
+class HashFunc(ABC):
+    @abstractmethod
+    def __call__(self, value: int) -> int: ...
+    @abstractmethod
+    def text(self) -> str: ...
+
+class Identity(HashFunc):
+    def __call__(self : Self, value : int) -> int:
+        return value
+    def text(self : Self) -> str:
+        return f"symbol"
+
+class Shift(HashFunc):
+    shift : int
+    def __init__(self : Self, shift : int):
+        self.shift = shift
+    def __call__(self : Self, value : int) -> int:
+        return value >> self.shift
+    def text(self : Self) -> str:
+        return f"symbol >> {self.shift}"
+
+class XorShift(HashFunc):
+    shift : int
+    def __init__(self : Self, shift : int):
+        self.shift = shift
+    def __call__(self : Self, value : int) -> int:
+        return value ^ (value >> self.shift)
+    def text(self : Self) -> str:
+        return f"symbol ^ symbol >> {self.shift}"
+
+def test_hash(table : list[list[int]], func : HashFunc, n : int) -> bool:
+    mask = (1 << n) - 1
+    check = [False] * (1 << n)
+    for code, _ in table:
+        index = func(code) & mask
+        if check[index]:
+            return False
+        check[index] = True
+    return True
+
+def hash_functions():
+    yield Identity()
+    for shift in range(1, 16):
+        yield Shift(shift)
+    for shift in range(1, 16):
+        yield XorShift(shift)
+
+def find_hash(table : list[list[int]]) -> tuple[HashFunc, int]:
+    m = len(table).bit_length()
+    for n in [m, m + 1]:
+        for func in hash_functions():
+            if test_hash(table, func, n):
+                return func, n
+    raise Exception("Cannot find suitable hash function!")
+
+
+LICENSE_HEADER = """\
+/*
+ * Copyright (C) 2026
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+// WARNING - THIS FILE IS AUTO-GENERATED. DO NOT EDIT IT MANUALLY.
+// Regenerate with: python3 gen_arabic_charmap.py
+"""
+
+def render_function(file_content: list[str], function_name: str, unicode_mapping_file: Path):
+    table = []
+    res_min = 1 << 32
+    with open(unicode_mapping_file, encoding="utf-8") as f:
+        lines = f.readlines()
+        for line in lines:
+            if line.startswith("#"):
+                continue
+            pua_code_str, unicode_str, _ = line.split("\t")
+            unicode = int(unicode_str, 0)
+            pua_code = int(pua_code_str, 0)
+            table.extend([[unicode, pua_code]])
+            res_min = min(res_min, pua_code)
+
+    func, n = find_hash(table)
+
+    n1 = n - 1
+    mask = (1 << n) - 1
+    checks = [1] * (1 << n1) + [0] * (1 << n1)
+    results = [0] * (1 << n)
+    for unicode, pua_code in table:
+        hash_val = func(unicode)
+        index = hash_val & mask
+        checks[index] = hash_val >> n1
+        results[index] = pua_code - res_min
+        if checks[index] > 255:
+            raise Exception("uint8_t isn't enough for check table!")
+        if results[index] > 255:
+            raise Exception("uint8_t isn't enough for result table!")
+
+    file_content.append(f"static const uint8_t {function_name}_check[{1 << n}] = {{")
+    for line in textwrap.wrap(", ".join(map(str, checks)), width = 100):
+        file_content.append(f"    {line}")
+    file_content.append("};")
+    file_content.append("")
+
+    file_content.append(f"static const uint8_t {function_name}_result[{1 << n}] = {{")
+    for line in textwrap.wrap(", ".join(map(str, results)), width = 100):
+        file_content.append(f"    {line}")
+    file_content.append("};")
+    file_content.append("")
+
+    expr = func.text()
+    file_content.append(f"uint32_t {function_name}(uint32_t symbol)")
+    file_content.append("{")
+    file_content.append(f"    unsigned hash = {expr}, index = hash & {mask}, test = hash >> {n1};")
+    file_content.append(f"    return test == {function_name}_check[index] ?")
+    file_content.append(f"      {res_min} + {function_name}_result[index] : symbol;")
+    file_content.append("}")
+    file_content.append("")
+
+
+def render_header(function_names: list[str]) -> list[str]:
+    file_content = []
+    guard = "LIBASS_ASS_ARABIC_CHARMAP_H"
+
+    file_content.append(LICENSE_HEADER)
+    file_content.append(f"#ifndef {guard}")
+    file_content.append(f"#define {guard}")
+    file_content.append("")
+    file_content.append("#include <stdint.h>")
+    file_content.append("")
+    for name in function_names:
+        file_content.append(f"uint32_t {name}(uint32_t symbol);")
+    file_content.append("")
+    file_content.append(f"#endif /* {guard} */")
+    file_content.append("")
+    return file_content
+
+
+def main():
+    SOURCES = [
+        {
+            "file": Path(__file__).parent.joinpath("ArabicPUASimplified.txt"),
+            "function": "ass_font_charmap_arabic_simplified",
+        },
+        {
+            "file": Path(__file__).parent.joinpath("ArabicPUATraditional.txt"),
+            "function": "ass_font_charmap_arabic_traditional",
+        },
+    ]
+
+    c_file_content = []
+    c_file_content.append(LICENSE_HEADER)
+    c_file_content.append("#include \"ass_arabic_charmap.h\"")
+    c_file_content.append("")
+
+    for source in SOURCES:
+        render_function(c_file_content, source["function"], source["file"])
+
+    c_file_path = Path(__file__).parent.parent.joinpath("libass", "ass_arabic_charmap.c")
+    c_file_path.write_text("\n".join(c_file_content), encoding="utf-8")
+
+    h_file_content = render_header([source["function"] for source in SOURCES])
+    h_file_path = Path(__file__).parent.parent.joinpath("libass", "ass_arabic_charmap.h")
+    h_file_path.write_text("\n".join(h_file_content), encoding="utf-8")
+
+
+if __name__ == "__main__":
+    main()
diff --git a/libass/Makefile_library.am b/libass/Makefile_library.am
index 738d935..ec7bff9 100644
--- a/libass/Makefile_library.am
+++ b/libass/Makefile_library.am
@@ -1,5 +1,5 @@
 LIBASS_LT_CURRENT = 13
-LIBASS_LT_REVISION = 1
+LIBASS_LT_REVISION = 2
 LIBASS_LT_AGE = 4
 
 .asm.lo:
@@ -24,6 +24,7 @@ libass_libass_internal_la_SOURCES = \
     libass/ass_rasterizer.h libass/ass_rasterizer.c \
     libass/ass_render.h libass/ass_render.c libass/ass_render_api.c \
     libass/ass_bitmap_engine.h libass/ass_bitmap_engine.c \
+    libass/ass_arabic_charmap.h libass/ass_arabic_charmap.c \
     libass/c/rasterizer_template.h libass/c/c_rasterizer.c \
     libass/c/c_blend_bitmaps.c \
     libass/c/c_be_blur.c \
diff --git a/libass/ass.c b/libass/ass.c
index 0b44649..b9ce811 100644
--- a/libass/ass.c
+++ b/libass/ass.c
@@ -200,6 +200,11 @@ static int test_and_set_read_order_bit(ASS_Track *track, int id)
 
 static inline void clear_read_order_bit(ASS_Track *track, int id)
 {
+    // we previously loaded a packet with an invalid ReadOrder field
+    // (or int overflow) and decided to just skip checking for duplicates
+    if (id < 0)
+        return;
+
     int index = id >> 5;
     if (index < track->parser_priv->read_order_elems) {
         uint32_t mask = ~(1u << (id & 0x1F));
@@ -1851,7 +1856,8 @@ void ass_lazy_track_init(ASS_Library *lib, ASS_Track *track)
             ass_msg(lib, MSGL_WARN,
                    "PlayResY undefined, setting to %d", track->PlayResY);
         } else if (track->PlayResY <= 0) {
-            track->PlayResY = FFMAX(1, track->PlayResX * 3LL / 4);
+            // overflow-safe `* 3/4`
+            track->PlayResY = FFMAX(1, (track->PlayResX - 1u) - (track->PlayResX - 1u) / 4u);
             ass_msg(lib, MSGL_WARN,
                    "PlayResY undefined, setting to %d", track->PlayResY);
         } else if (track->PlayResX <= 0 && track->PlayResY == 1024) {
@@ -1859,7 +1865,7 @@ void ass_lazy_track_init(ASS_Library *lib, ASS_Track *track)
             ass_msg(lib, MSGL_WARN,
                    "PlayResX undefined, setting to %d", track->PlayResX);
         } else if (track->PlayResX <= 0) {
-            track->PlayResX = FFMAX(1, track->PlayResY * 4LL / 3);
+            track->PlayResX = FFMIN(track->PlayResY + track->PlayResY / 3u, (unsigned) INT_MAX);
             ass_msg(lib, MSGL_WARN,
                    "PlayResX undefined, setting to %d", track->PlayResX);
         }
diff --git a/libass/ass.h b/libass/ass.h
index 3ff1598..851ca7c 100644
--- a/libass/ass.h
+++ b/libass/ass.h
@@ -24,7 +24,7 @@
 #include <stdarg.h>
 #include "ass_types.h"
 
-#define LIBASS_VERSION 0x01704000
+#define LIBASS_VERSION 0x01705000
 
 #ifdef __cplusplus
 extern "C" {
diff --git a/libass/ass_arabic_charmap.c b/libass/ass_arabic_charmap.c
new file mode 100644
index 0000000..d928dd7
--- /dev/null
+++ b/libass/ass_arabic_charmap.c
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2026
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+// WARNING - THIS FILE IS AUTO-GENERATED. DO NOT EDIT IT MANUALLY.
+// Regenerate with: python3 gen_arabic_charmap.py
+
+#include "ass_arabic_charmap.h"
+
+static const uint8_t ass_font_charmap_arabic_simplified_check[512] = {
+    1, 1, 1, 1, 1, 1, 1, 1, 48, 48, 48, 48, 1, 1, 1, 1, 1, 1, 1, 1, 48, 48, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 5, 5, 5, 5, 5, 5,
+    5, 5, 5, 5, 5, 5, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 5, 5, 5, 5, 5, 5, 5, 5, 0, 0, 5, 0, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 5, 5, 5,
+    5, 5, 5, 5, 5, 5, 5, 0, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 0, 0, 129, 0, 129, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 0, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129,
+    129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129, 129,
+    129
+};
+
+static const uint8_t ass_font_charmap_arabic_simplified_result[512] = {
+    0, 0, 0, 0, 0, 0, 0, 0, 15, 14, 12, 13, 0, 0, 0, 0, 0, 0, 0, 0, 60, 62, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 63, 0, 0, 61, 184, 185, 59, 58, 176, 177, 179, 178, 183, 182, 180, 181, 32, 33, 0, 34, 0, 0, 0,
+    37, 47, 46, 94, 45, 40, 41, 43, 42, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 92, 93, 0, 0,
+    91, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 39, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 38, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 36, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 35, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 44, 0, 0, 0, 0, 0, 63, 0, 0, 0, 0, 0, 59, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 106, 110, 0, 114, 89, 90, 98, 96, 105, 104, 100, 102, 0, 173, 67, 69,
+    65, 186, 187, 71, 88, 87, 81, 84, 74, 169, 78, 76, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    48, 49, 51, 50, 55, 54, 52, 53, 0, 0, 94, 0, 56, 57, 94, 0, 95, 117, 122, 120, 164, 225, 124, 126,
+    197, 196, 200, 203, 165, 172, 199, 168, 0, 0, 0, 0, 0, 0, 0, 0, 202, 201, 0, 198, 0, 0, 0, 0, 0, 0,
+    188, 0, 191, 194, 189, 195, 171, 168, 166, 167, 190, 193, 166, 192, 123, 126, 125, 126, 127, 225,
+    125, 225, 172, 165, 162, 165, 127, 164, 161, 163, 103, 105, 105, 105, 106, 106, 105, 106, 111, 113,
+    108, 114, 106, 110, 107, 109, 123, 124, 121, 124, 119, 122, 121, 122, 112, 117, 115, 117, 118, 120,
+    116, 120, 0, 69, 67, 70, 71, 187, 68, 187, 74, 66, 174, 65, 72, 186, 174, 175, 79, 80, 77, 81, 75,
+    78, 77, 78, 74, 73, 169, 73, 75, 76, 170, 76, 103, 104, 101, 104, 99, 102, 101, 102, 96, 98, 97, 98,
+    99, 100, 97, 100, 79, 84, 82, 83, 85, 86, 82, 87, 96, 90, 89, 90, 85, 88, 89, 88
+};
+
+uint32_t ass_font_charmap_arabic_simplified(uint32_t symbol)
+{
+    unsigned hash = symbol ^ symbol >> 1, index = hash & 511, test = hash >> 8;
+    return test == ass_font_charmap_arabic_simplified_check[index] ?
+      61696 + ass_font_charmap_arabic_simplified_result[index] : symbol;
+}
+
+static const uint8_t ass_font_charmap_arabic_traditional_check[1024] = {
+    1, 1, 1, 1, 1, 1, 1, 1, 24, 24, 24, 24, 1, 1, 1, 1, 1, 1, 24, 24, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2,
+    2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2, 1, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+    2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 64, 1, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 1, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 65, 0, 0, 65, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 65, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 65, 0, 65, 0, 0, 0, 0, 0, 65, 65, 0, 65, 0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 65, 65, 65, 65, 65, 0,
+    0, 65, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 65, 65, 65, 65, 0, 65, 65, 65, 0, 0, 65, 65, 0, 0, 65, 65, 65, 65, 65, 65, 0, 0, 0, 65, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 65, 65, 65, 65, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 65, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 65, 65, 65, 0, 0, 65, 0, 0, 0, 65, 0, 65, 0, 0, 65,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 65, 65, 0, 0, 0,
+    0, 0, 0, 65, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+static const uint8_t ass_font_charmap_arabic_traditional_result[1024] = {
+    0, 0, 0, 0, 0, 0, 0, 0, 15, 14, 12, 13, 0, 0, 0, 0, 0, 0, 60, 62, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 63, 0, 0, 61, 0, 0, 59, 58, 0, 0, 0, 0, 0, 0, 0, 0, 32, 33, 0, 34, 0, 0, 0, 37, 47, 46, 94, 45,
+    40, 41, 43, 42, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 93, 0, 0, 91, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 39, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 38, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 36, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 35, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 44, 0, 0, 0, 0, 0, 63, 0, 0, 0, 0, 0, 59, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 163, 170, 0, 174, 103, 105, 112, 107, 162, 126, 116, 120, 0, 213, 67, 69, 65, 217, 218, 71,
+    101, 100, 88, 96, 76, 209, 84, 80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 48, 49, 51, 50,
+    55, 54, 52, 53, 0, 0, 94, 0, 56, 57, 94, 0, 95, 178, 186, 182, 202, 198, 190, 194, 229, 228, 232,
+    235, 203, 212, 231, 208, 0, 0, 0, 0, 0, 0, 0, 0, 234, 233, 0, 230, 0, 0, 0, 0, 0, 0, 221, 0, 223,
+    226, 220, 227, 211, 208, 205, 207, 222, 225, 206, 224, 188, 194, 191, 193, 195, 197, 192, 198, 212,
+    204, 200, 203, 196, 202, 199, 201, 122, 162, 127, 161, 163, 165, 241, 166, 171, 173, 168, 174, 164,
+    170, 167, 169, 187, 189, 184, 190, 180, 186, 183, 185, 172, 178, 175, 177, 179, 181, 176, 182, 0,
+    69, 67, 70, 71, 219, 68, 218, 76, 66, 215, 65, 72, 217, 214, 216, 85, 87, 82, 88, 78, 84, 81, 83,
+    75, 73, 209, 74, 77, 79, 210, 80, 121, 124, 118, 126, 114, 120, 117, 119, 108, 112, 109, 111, 113,
+    115, 110, 116, 86, 96, 89, 92, 97, 99, 90, 100, 107, 106, 104, 105, 98, 101, 103, 102, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 3, 0, 0, 2, 0, 0, 29, 0, 0, 0, 0, 0, 0, 0, 0, 30, 0, 0, 4, 0, 0, 0, 0, 18, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 159, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 22, 0, 147, 0, 0, 0, 0, 0, 236, 237, 0, 240, 0, 0, 0, 0, 0, 0, 0, 146, 0, 0, 0, 21, 19, 20,
+    149, 5, 0, 0, 28, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 239, 0, 0, 0, 0, 0, 0, 31, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 138, 137, 157, 26, 0, 134, 136, 135, 0, 0, 145, 153, 0, 0, 144, 143, 139, 158,
+    142, 141, 0, 0, 0, 152, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 150, 130, 128, 129, 0, 0, 0,
+    0, 0, 23, 0, 0, 0, 0, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 24, 0, 0, 0, 0, 0, 0, 0, 0, 131, 133, 132, 0,
+    0, 151, 0, 0, 0, 156, 0, 154, 0, 0, 155, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 125, 123, 0, 0, 0, 0, 0, 0, 25, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0
+};
+
+uint32_t ass_font_charmap_arabic_traditional(uint32_t symbol)
+{
+    unsigned hash = symbol ^ symbol >> 1, index = hash & 1023, test = hash >> 9;
+    return test == ass_font_charmap_arabic_traditional_check[index] ?
+      61952 + ass_font_charmap_arabic_traditional_result[index] : symbol;
+}
diff --git a/libass/ass_arabic_charmap.h b/libass/ass_arabic_charmap.h
new file mode 100644
index 0000000..4666c50
--- /dev/null
+++ b/libass/ass_arabic_charmap.h
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2026
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+// WARNING - THIS FILE IS AUTO-GENERATED. DO NOT EDIT IT MANUALLY.
+// Regenerate with: python3 gen_arabic_charmap.py
+
+#ifndef LIBASS_ASS_ARABIC_CHARMAP_H
+#define LIBASS_ASS_ARABIC_CHARMAP_H
+
+#include <stdint.h>
+
+uint32_t ass_font_charmap_arabic_simplified(uint32_t symbol);
+uint32_t ass_font_charmap_arabic_traditional(uint32_t symbol);
+
+#endif /* LIBASS_ASS_ARABIC_CHARMAP_H */
diff --git a/libass/ass_font.c b/libass/ass_font.c
index 312fee2..aa12ca1 100644
--- a/libass/ass_font.c
+++ b/libass/ass_font.c
@@ -259,8 +259,20 @@ uint32_t ass_font_index_magic(FT_Face face, uint32_t symbol)
 
     if (face->charmap->platform_id == TT_PLATFORM_MICROSOFT) {
         switch (face->charmap->encoding) {
-        case FT_ENCODING_MS_SYMBOL:
+        case FT_ENCODING_MS_SYMBOL: {
+            TT_OS2 *os2 = FT_Get_Sfnt_Table(face, FT_SFNT_OS2);
+            if (os2) {
+                // See: https://learn.microsoft.com/en-us/typography/legacy/legacy_arabic_fonts
+                int ARABIC_CHARSET_SIMPLIFIED  = 178;
+                int ARABIC_CHARSET_TRADITIONAL = 179;
+                uint8_t charset = (os2->fsSelection >> 8) & 0xFF;
+                if (charset == ARABIC_CHARSET_SIMPLIFIED)
+                    return ass_font_charmap_arabic_simplified(symbol);
+                if (charset == ARABIC_CHARSET_TRADITIONAL)
+                    return ass_font_charmap_arabic_traditional(symbol);
+            }
             return 0xF000 | symbol;
+        }
         case FT_ENCODING_MS_SJIS:
         case FT_ENCODING_MS_GB2312:
         case FT_ENCODING_MS_BIG5:
diff --git a/libass/ass_font.h b/libass/ass_font.h
index 53ba8ea..2388249 100644
--- a/libass/ass_font.h
+++ b/libass/ass_font.h
@@ -30,6 +30,7 @@ typedef struct ass_font ASS_Font;
 #include "ass_fontselect.h"
 #include "ass_cache.h"
 #include "ass_outline.h"
+#include "ass_arabic_charmap.h"
 
 #define VERTICAL_LOWER_BOUND 0x02f1
 
diff --git a/libass/ass_fontconfig.c b/libass/ass_fontconfig.c
index 6624a80..0e6930b 100644
--- a/libass/ass_fontconfig.c
+++ b/libass/ass_fontconfig.c
@@ -100,7 +100,11 @@ static bool scan_fonts(FcConfig *config, ASS_FontProvider *provider)
     if (!pat)
         return false;
 
+#if FC_VERSION >= 21700
+    FcConfigSetDefaultSubstitute(config, pat);
+#else
     FcDefaultSubstitute(pat);
+#endif
     FcResult res;
     // trim=FcFalse returns all system fonts
     fonts = FcFontSort(config, pat, FcFalse, NULL, &res);
@@ -229,8 +233,12 @@ static void cache_fallbacks(ProviderPrivate *fc)
         return;
     FcPatternAddString(pat, FC_FAMILY, (FcChar8 *)"sans-serif");
     FcPatternAddBool(pat, FC_OUTLINE, FcTrue);
-    FcConfigSubstitute (fc->config, pat, FcMatchPattern);
-    FcDefaultSubstitute (pat);
+    FcConfigSubstitute(fc->config, pat, FcMatchPattern);
+#if FC_VERSION >= 21700
+    FcConfigSetDefaultSubstitute(fc->config, pat);
+#else
+    FcDefaultSubstitute(pat);
+#endif
 
     // FC_LANG is automatically set according to locale, but this results
     // in strange sorting sometimes, so remove the attribute completely.
diff --git a/libass/ass_outline.c b/libass/ass_outline.c
index cf9575d..188f07b 100644
--- a/libass/ass_outline.c
+++ b/libass/ass_outline.c
@@ -476,7 +476,7 @@ void ass_outline_update_min_transformed_x(const ASS_Outline *outline,
     for (size_t i = 0; i < outline->n_points; i++) {
         double z = m[2][0] * pt[i].x + m[2][1] * pt[i].y + m[2][2];
         double x = (m[0][0] * pt[i].x + m[0][1] * pt[i].y + m[0][2]) / FFMAX(z, 0.1);
-        if (isnan(x))
+        if (ass_isnan(x))
             continue;
         int32_t ix = ass_lrint(FFMINMAX(x, -OUTLINE_MAX, OUTLINE_MAX));
         *min_x = FFMIN(*min_x, ix);
diff --git a/libass/ass_parse.c b/libass/ass_parse.c
index b1f5cb9..28d49e7 100644
--- a/libass/ass_parse.c
+++ b/libass/ass_parse.c
@@ -120,7 +120,7 @@ void ass_update_font(RenderContext *state)
  */
 static inline int32_t dtoi32(double val)
 {
-    if (isnan(val) || val <= INT32_MIN || val >= INT32_MAX + 1LL)
+    if (ass_isnan(val) || val <= INT32_MIN || val >= INT32_MAX + 1LL)
         return INT32_MIN;
     return val;
 }
diff --git a/libass/ass_rasterizer.c b/libass/ass_rasterizer.c
index 0ac45d1..a0c958e 100644
--- a/libass/ass_rasterizer.c
+++ b/libass/ass_rasterizer.c
@@ -33,10 +33,10 @@
 
 static inline int ilog2(uint32_t n)
 {
-#ifdef __GNUC__
+#if defined(__GNUC__) || defined(__clang__)
     return __builtin_clz(n) ^ 31;
 #elif defined(_MSC_VER)
-    int res;
+    unsigned long res;
     _BitScanReverse(&res, n);
     return res;
 #else
diff --git a/libass/ass_render.c b/libass/ass_render.c
index 68b11fe..18f856a 100644
--- a/libass/ass_render.c
+++ b/libass/ass_render.c
@@ -1660,6 +1660,9 @@ static void trim_whitespace(RenderContext *state)
     GlyphInfo *cur;
     TextInfo *ti = &state->text_info;
 
+    if (!ti->length)
+      return;
+
     // Mark trailing spaces
     i = ti->length - 1;
     cur = ti->glyphs + i;
@@ -1681,17 +1684,21 @@ static void trim_whitespace(RenderContext *state)
         cur->starts_new_run = true;
 
     // Mark all extraneous whitespace inbetween
+    // XXX: should this really start at 0 again?
     for (i = 0; i < ti->length; ++i) {
         cur = ti->glyphs + i;
         if (cur->linebreak) {
             // Mark whitespace before
             j = i - 1;
             cur = ti->glyphs + j;
-            while (j && IS_WHITESPACE(cur)) {
+            // Use > instead of >= to avoid UB from moving the pointer outside valid range.
+            // White space at j == 0 was already trimmed in the "leading" loop before anyway.
+            while (j > 0 && IS_WHITESPACE(cur)) {
                 cur->skip = true;
                 cur->is_trimmed_whitespace = true;
                 cur = ti->glyphs + --j;
             }
+
             // A break itself can contain a whitespace, too
             cur = ti->glyphs + i;
             if (cur->symbol == ' ' || cur->symbol == '\n') {
@@ -1759,10 +1766,8 @@ wrap_lines_naive(RenderContext *state, double max_text_width, char *unibrks)
             last_breakable = i;
         }
 
-        if (break_at != -1) {
+        if (break_at != -1 && break_at + 1 < text_info->length) {
             // need to use one more line
-            // marking break_at+1 as start of a new line
-            int lead = break_at + 1;    // the first symbol of the new line
             if (text_info->n_lines >= text_info->max_lines) {
                 // Try to raise the maximum number of lines
                 bool success = false;
@@ -1777,12 +1782,13 @@ wrap_lines_naive(RenderContext *state, double max_text_width, char *unibrks)
                     text_info->n_lines--;
                 }
             }
-            if (lead < text_info->length) {
-                text_info->glyphs[lead].linebreak = break_type;
-                last_breakable = -1;
-                s1 = text_info->glyphs + lead;
-                text_info->n_lines++;
-            }
+
+            // marking break_at+1 as start of a new line
+            int lead = break_at + 1; // the first symbol of the new line
+            text_info->glyphs[lead].linebreak = break_type;
+            last_breakable = -1;
+            s1 = text_info->glyphs + lead;
+            text_info->n_lines++;
         }
     }
 }
@@ -1791,6 +1797,7 @@ wrap_lines_naive(RenderContext *state, double max_text_width, char *unibrks)
  * Rewind from a linestart position back to the first non-whitespace (0x20)
  * character. Trailing ASCII whitespace gets trimmed in rendering.
  * Assumes both arguments are part of the same array.
+ * start2 is never dereferenced.
  */
 static inline GlyphInfo *rewind_trailing_spaces(GlyphInfo *start1, GlyphInfo* start2)
 {
@@ -1823,6 +1830,8 @@ wrap_lines_rebalance(RenderContext *state, double max_text_width, char *unibrks)
             if ((i == text_info->length) || cur->linebreak) {
                 s1 = s2;
                 s2 = s3;
+                // WARNING: this may point one past the end and thus
+                // must ONLY be used for pointer comparison; never dereferenced!
                 s3 = cur;
                 if (s1 && (s2->linebreak == 1)) {       // have at least 2 lines, and linebreak is 'soft'
                     double l1, l2, l1_new, l2_new;
@@ -1875,40 +1884,6 @@ wrap_lines_rebalance(RenderContext *state, double max_text_width, char *unibrks)
 #undef DIFF
 }
 
-static void
-wrap_lines_measure(RenderContext *state, char *unibrks)
-{
-    TextInfo *text_info = &state->text_info;
-    int cur_line = 1;
-    int i = 0;
-
-    while (i < text_info->length && text_info->glyphs[i].skip)
-        ++i;
-    double pen_shift_x = d6_to_double(-text_info->glyphs[i].pos.x);
-    double pen_shift_y = 0.;
-
-    for (i = 0; i < text_info->length; ++i) {
-        GlyphInfo *cur = text_info->glyphs + i;
-        if (cur->linebreak) {
-            while (i < text_info->length && cur->skip && !FORCEBREAK(cur->symbol, i))
-                cur = text_info->glyphs + ++i;
-            double height =
-                text_info->lines[cur_line - 1].desc +
-                text_info->lines[cur_line].asc;
-            text_info->lines[cur_line - 1].len = i -
-                text_info->lines[cur_line - 1].offset;
-            text_info->lines[cur_line].offset = i;
-            cur_line++;
-            pen_shift_x = d6_to_double(-cur->pos.x);
-            pen_shift_y += height + state->renderer->settings.line_spacing;
-        }
-        cur->pos.x += double_to_d6(pen_shift_x);
-        cur->pos.y += double_to_d6(pen_shift_y);
-    }
-    text_info->lines[cur_line - 1].len =
-        text_info->length - text_info->lines[cur_line - 1].offset;
-}
-
 #undef ALLOWBREAK
 #undef FORCEBREAK
 
@@ -1955,7 +1930,6 @@ wrap_lines_smart(RenderContext *state, double max_text_width)
 
     trim_whitespace(state);
     measure_text(state);
-    wrap_lines_measure(state, unibrks);
 }
 
 /**
@@ -2145,12 +2119,8 @@ static bool parse_events(RenderContext *state, ASS_Event *event)
         // Fill glyph information
         info->symbol = code;
         info->font = state->font;
-        for (int i = 0; i < 4; i++) {
-            uint32_t clr = state->c[i];
-            info->a_pre_fade[i] = _a(clr);
-            ass_apply_fade(&clr, state->fade);
-            info->c[i] = clr;
-        }
+        for (int i = 0; i < 4; i++)
+            info->c[i] = state->c[i];
 
         info->effect_type = state->effect_type;
         info->effect_timing = state->effect_timing;
@@ -2316,9 +2286,11 @@ static void apply_baseline_shear(RenderContext *state)
             info->skip = true;
         if (info->skip)
             continue;
-        for (GlyphInfo *cur = info; cur; cur = cur->next)
-            cur->pos.y += shear;
-        shear += (info->fay / info->scale_x * info->scale_y) * info->cluster_advance.x;
+        double fay = info->fay / info->scale_x * info->scale_y;
+        for (GlyphInfo *cur = info; cur; cur = cur->next) {
+            cur->pos.y += shear + fay * cur->offset.x;
+            shear += fay * cur->advance.x;
+        }
     }
 }
 
@@ -2501,15 +2473,15 @@ static void render_and_combine_glyphs(RenderContext *state,
             if (flags & FILTER_NONZERO_SHADOW &&
                 (info->effect_type == EF_KARAOKE_KF ||
                  info->effect_type == EF_KARAOKE_KO ||
-                 (info->a_pre_fade[0]) != 0xFF ||
+                 _a(info->c[0]) != 0xFF ||
                  info->border_style == 3))
                 flags |= FILTER_FILL_IN_SHADOW;
             if (!(flags & FILTER_NONZERO_BORDER) &&
                 !(flags & FILTER_FILL_IN_SHADOW))
                 flags &= ~FILTER_NONZERO_SHADOW;
             if ((flags & FILTER_NONZERO_BORDER &&
-                 info->a_pre_fade[0] == 0 &&
-                 info->a_pre_fade[1] == 0 &&
+                 _a(info->c[0]) == 0 &&
+                 _a(info->c[1]) == 0 &&
                  info->fade == 0) ||
                 info->border_style == 3)
                 flags |= FILTER_FILL_IN_BORDER;
@@ -2526,6 +2498,9 @@ static void render_and_combine_glyphs(RenderContext *state,
                 current_info = &combined_info[nb_bitmaps];
 
                 memcpy(&current_info->c, &info->c, sizeof(info->c));
+                for (int i = 0; i < 4; i++)
+                    ass_apply_fade(&current_info->c[i], info->fade);
+
                 current_info->effect_type = info->effect_type;
                 current_info->effect_timing = info->effect_timing;
                 current_info->leftmost_x = OUTLINE_MAX;
@@ -2804,8 +2779,10 @@ static void add_background(RenderContext *state, EventImages *event_images)
     if (!nbuffer)
         return;
     memset(nbuffer, 0xFF, w * h);
+    uint32_t clr = state->c[3];
+    ass_apply_fade(&clr, state->fade);
     ASS_Image *img = my_draw_bitmap(nbuffer, w, h, w, left, top,
-                                    state->c[3], NULL);
+                                    clr, NULL);
     if (img) {
         img->next = event_images->imgs;
         event_images->imgs = img;
@@ -3362,6 +3339,25 @@ static int ass_detect_change(ASS_Renderer *priv)
     return diff;
 }
 
+/**
+ * \brief free a single image.
+ * \param img image as returned by ass_render_frame().
+ *        Only img will be freed, not img->next.
+ * \return img->next
+ * Should not be called outside of ass_frame_unref
+ * once the image list's refcounting is set up.
+ */
+static ASS_Image *ass_free_image(ASS_Image *img) {
+    ASS_Image *next = img->next;
+
+    ASS_ImagePriv *priv = (ASS_ImagePriv *) img;
+    ass_cache_dec_ref(priv->source);
+    ass_aligned_free(priv->buffer);
+    free(priv);
+
+    return next;
+}
+
 /**
  * \brief render a frame
  * \param priv library handle
@@ -3412,16 +3408,25 @@ ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
     if (cnt > 0)
         fix_collisions(priv, last, priv->eimg + cnt - last);
 
-    // concat lists
+    // concat lists, removing fully transparent bitmaps
     ASS_Image **tail = &priv->images_root;
     for (int i = 0; i < cnt; i++) {
         ASS_Image *cur = priv->eimg[i].imgs;
         while (cur) {
+            if (_a(cur->color) == 0xFF) {
+                cur = ass_free_image(cur);
+                continue;
+            }
+
             *tail = cur;
             tail = &cur->next;
             cur = cur->next;
         }
     }
+
+    // If the last image was skipped in the above loop, *tail may not be NULL and needs to be set to NULL.
+    *tail = NULL;
+
     ass_frame_ref(priv->images_root);
 
     if (detect_change)
@@ -3457,10 +3462,6 @@ void ass_frame_unref(ASS_Image *img)
     if (!img || --((ASS_ImagePriv *) img)->ref_count)
         return;
     do {
-        ASS_ImagePriv *priv = (ASS_ImagePriv *) img;
-        img = img->next;
-        ass_cache_dec_ref(priv->source);
-        ass_aligned_free(priv->buffer);
-        free(priv);
+        img = ass_free_image(img);
     } while (img);
 }
diff --git a/libass/ass_render.h b/libass/ass_render.h
index 7a157ab..e1eb7d2 100644
--- a/libass/ass_render.h
+++ b/libass/ass_render.h
@@ -98,7 +98,7 @@ typedef enum {
 // describes a combined bitmap
 typedef struct {
     FilterDesc filter;
-    uint32_t c[4];              // colors
+    uint32_t c[4];              // colors (with fade applied)
     Effect effect_type;
 
     // during render_and_combine_glyphs: distance in subpixels from the karaoke origin.
@@ -143,7 +143,6 @@ typedef struct glyph_info {
     char linebreak;             // the first (leading) glyph of some line ?
     bool starts_new_run;
     uint32_t c[4];              // colors
-    uint8_t a_pre_fade[4];      // alpha values before applying fades
     ASS_Vector advance;         // 26.6
     ASS_Vector cluster_advance;
     Effect effect_type;
@@ -183,7 +182,6 @@ typedef struct glyph_info {
 
 typedef struct {
     double asc, desc;
-    int offset, len;
 } LineInfo;
 
 typedef struct {
diff --git a/libass/ass_utils.h b/libass/ass_utils.h
index a5d4e5d..5b7cc7a 100644
--- a/libass/ass_utils.h
+++ b/libass/ass_utils.h
@@ -186,6 +186,12 @@ static inline long ass_lrint(double x)
 #endif
 }
 
+#if __has_builtin(__builtin_isnan)
+    #define ass_isnan __builtin_isnan
+#else
+    #define ass_isnan isnan
+#endif
+
 static inline int d6_to_int(int x)
 {
     return (x + 32) >> 6;
diff --git a/libass/meson.build b/libass/meson.build
index 5251c2f..9eda958 100644
--- a/libass/meson.build
+++ b/libass/meson.build
@@ -47,6 +47,7 @@ libass_src = files(
     'c/c_blur.c',
     'c/c_rasterizer.c',
     'ass.c',
+    'ass_arabic_charmap.c',
     'ass_bitmap.c',
     'ass_bitmap_engine.c',
     'ass_blur.c',
@@ -93,7 +94,7 @@ if enable_asm
         libass_src += asm_sources
     else
         asm_lib = static_library(
-            'libass_asm',
+            'ass_asm',
             config_h,
             sources: asm_sources,
             c_args: asm_args,
diff --git a/meson.build b/meson.build
index 4b1c24e..91a0435 100644
--- a/meson.build
+++ b/meson.build
@@ -310,6 +310,11 @@ if not asm_option.disabled()
                 )
             endif
 
+            # Convert SSE asm into (128-bit) AVX when compiler flags are set to use AVX instructions
+            if cc.get_define('__AVX__').strip() != ''
+                nasm_args += '-DFORCE_VEX_ENCODING=1'
+            endif
+
             add_project_arguments(nasm_args, language: 'nasm')
         endif
     elif generic_cpu_family == 'aarch64'
