diff --git a/Changelog b/Changelog
index 79bec30..2b77a40 100644
--- a/Changelog
+++ b/Changelog
@@ -1,6 +1,6 @@
 libass (0.17.5)
- * Fix limited OOB read and write in wrap_lines_measure (GHSA-pjjp-65r7-ppgm; CVE pending)
- * Fix OOB bit clears for negative Matroska ReadOrder fields (GHSA-5gf7-wjfm-vmvm; CVE pending)
+ * Fix limited OOB read and write in wrap_lines_measure (GHSA-pjjp-65r7-ppgm; CVE-2026-61627)
+ * Fix OOB bit clears for negative Matroska ReadOrder fields (GHSA-5gf7-wjfm-vmvm; CVE-2026-61626)
  * Fix \fay with glyph clusters
  * Fix small alpha changes not always splitting runs when combined with fade
  * Fix compilation with MSVC-mode clang
diff --git a/Makefile_util.am b/Makefile_util.am
index c2cd1c1..cf863fa 100644
--- a/Makefile_util.am
+++ b/Makefile_util.am
@@ -52,6 +52,7 @@ endif
 checkasm_checkasm_SOURCES = \
     checkasm/rasterizer.c \
     checkasm/blend_bitmaps.c \
+    checkasm/shift.c \
     checkasm/be_blur.c \
     checkasm/blur.c \
     checkasm/checkasm.h checkasm/checkasm.c \
diff --git a/checkasm/checkasm.c b/checkasm/checkasm.c
index 857938b..0ddd74d 100644
--- a/checkasm/checkasm.c
+++ b/checkasm/checkasm.c
@@ -59,6 +59,7 @@ static const struct {
 } tests[] = {
     { "rasterizer", checkasm_check_rasterizer },
     { "blend_bitmaps", checkasm_check_blend_bitmaps },
+    { "shift", checkasm_check_shift },
     { "be_blur", checkasm_check_be_blur },
     { "blur", checkasm_check_blur },
     { 0 }
diff --git a/checkasm/checkasm.h b/checkasm/checkasm.h
index 3cd7c24..9413019 100644
--- a/checkasm/checkasm.h
+++ b/checkasm/checkasm.h
@@ -61,6 +61,7 @@ int xor128_rand(void);
 
 void checkasm_check_rasterizer(unsigned cpu_flag);
 void checkasm_check_blend_bitmaps(unsigned cpu_flag);
+void checkasm_check_shift(unsigned cpu_flag);
 void checkasm_check_be_blur(unsigned cpu_flag);
 void checkasm_check_blur(unsigned cpu_flag);
 
diff --git a/checkasm/meson.build b/checkasm/meson.build
index c8d8acf..5a27ce2 100644
--- a/checkasm/meson.build
+++ b/checkasm/meson.build
@@ -2,6 +2,7 @@ checkasm_src = files(
     'checkasm.c',
     'rasterizer.c',
     'blend_bitmaps.c',
+    'shift.c',
     'be_blur.c',
     'blur.c',
 )
diff --git a/checkasm/shift.c b/checkasm/shift.c
new file mode 100644
index 0000000..9dca4ec
--- /dev/null
+++ b/checkasm/shift.c
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2026 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "ass_compat.h"
+
+#include "checkasm.h"
+
+#include <string.h>
+
+#define HEIGHT 8
+#define STRIDE 64
+#define MIN_WIDTH 2
+#define REP_COUNT 8
+
+static void check_shift(ShiftFunc func)
+{
+    ALIGN(uint8_t buf_ref[STRIDE * HEIGHT], 32);
+    ALIGN(uint8_t buf_new[STRIDE * HEIGHT], 32);
+    declare_func(void,
+                 uint8_t *buf, ptrdiff_t stride,
+                 size_t width, size_t height,
+                 int shift_x, int shift_y);
+
+    if (check_func(func, "shift")) {
+        for (int w = MIN_WIDTH; w <= STRIDE; w++) {
+            for(int rep = 0; rep < REP_COUNT; rep++) {
+                memset(buf_ref, 0, sizeof(buf_ref));
+                memset(buf_new, 0, sizeof(buf_new));
+                for (int y = 0; y < HEIGHT; y++) {
+                    for (int x = 0; x < w; x++)
+                        buf_ref[y * STRIDE + x] = buf_new[y * STRIDE + x] = rnd();
+                }
+
+                int shift_x = rnd() & 63;
+                int shift_y = rnd() & 63;
+
+                call_ref(buf_ref, STRIDE, w, HEIGHT, shift_x, shift_y);
+                call_new(buf_new, STRIDE, w, HEIGHT, shift_x, shift_y);
+
+                if (memcmp(buf_ref, buf_new, sizeof(buf_ref))) {
+                    fail();
+                    break;
+                }
+            }
+        }
+
+        bench_new(buf_new, STRIDE, STRIDE, HEIGHT, 42, 13);
+    }
+
+    report("shift");
+}
+
+void checkasm_check_shift(unsigned cpu_flag)
+{
+    BitmapEngine engine = ass_bitmap_engine_init(cpu_flag);
+    check_shift(engine.shift);
+}
diff --git a/configure.ac b/configure.ac
index 27676b3..50085c4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -16,7 +16,7 @@ AM_PROG_AS
 AC_CHECK_HEADERS_ONCE([iconv.h])
 
 # Checks for library functions.
-AC_CHECK_FUNCS([strdup strndup])
+AC_CHECK_FUNCS([strdup strndup strchrnul])
 
 # Query configuration parameters and set their description
 AC_ARG_ENABLE([test], AS_HELP_STRING([--enable-test],
diff --git a/libass/Makefile_library.am b/libass/Makefile_library.am
index ec7bff9..1830e08 100644
--- a/libass/Makefile_library.am
+++ b/libass/Makefile_library.am
@@ -11,7 +11,8 @@ libass_libass_internal_la_SOURCES = \
     libass/ass_string.h libass/ass_string.c \
     libass/ass_compat.h libass/ass_strtod.c \
     libass/ass_filesystem.h libass/ass_filesystem.c \
-    libass/ass_types.h libass/ass.h libass/ass_priv.h libass/ass.c \
+    libass/ass_types.h libass/ass_metrics.h \
+    libass/ass.h libass/ass_priv.h libass/ass.c \
     libass/ass_library.h libass/ass_library.c \
     libass/ass_cache_template.h libass/ass_cache.h libass/ass_cache.c \
     libass/ass_font.h libass/ass_font.c \
@@ -27,6 +28,7 @@ libass_libass_internal_la_SOURCES = \
     libass/ass_arabic_charmap.h libass/ass_arabic_charmap.c \
     libass/c/rasterizer_template.h libass/c/c_rasterizer.c \
     libass/c/c_blend_bitmaps.c \
+    libass/c/c_shift.c \
     libass/c/c_be_blur.c \
     libass/c/blur_template.h libass/c/c_blur.c \
     libass/wyhash.h
@@ -36,6 +38,7 @@ if X86
 libass_libass_internal_la_SOURCES += \
     libass/x86/rasterizer.asm \
     libass/x86/blend_bitmaps.asm \
+    libass/x86/shift.asm \
     libass/x86/be_blur.asm \
     libass/x86/blur.asm \
     libass/x86/cpuid.h libass/x86/cpuid.asm
@@ -44,6 +47,7 @@ if AARCH64
 libass_libass_internal_la_SOURCES += \
     libass/aarch64/rasterizer.S \
     libass/aarch64/blend_bitmaps.S \
+    libass/aarch64/shift.S \
     libass/aarch64/be_blur.S \
     libass/aarch64/blur.S \
     libass/aarch64/asm.S
@@ -72,7 +76,7 @@ libass_libass_la_LDFLAGS = -no-undefined -version-info $(LIBASS_LT_CURRENT):$(LI
 libass_libass_la_LDFLAGS += -export-symbols $(top_srcdir)/libass/libass.sym
 
 assheadersdir = $(includedir)/ass
-dist_assheaders_HEADERS = libass/ass_types.h libass/ass.h
+dist_assheaders_HEADERS = libass/ass_types.h libass/ass_metrics.h libass/ass.h
 
 EXTRA_DIST += \
     libass/x86/x86inc.asm libass/x86/utils.asm \
diff --git a/libass/aarch64/shift.S b/libass/aarch64/shift.S
new file mode 100644
index 0000000..90ee103
--- /dev/null
+++ b/libass/aarch64/shift.S
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2026 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "asm.S"
+
+/*
+ * shift_x
+ * Perform subpixel shift along X axis
+ */
+
+.macro shift_x dst1, dst2, src1, src2, mul, vtmp1, vtmp2
+    ext \vtmp1\().16b, \src1\().16b, \src2\().16b, 1
+    ushll \dst1\().8h, \src1\().8b, 6
+    usubl \vtmp2\().8h, \src1\().8b, \vtmp1\().8b
+    mls \dst1\().8h, \vtmp2\().8h, \mul\().h[0]
+    ushll2 \dst2\().8h, \src1\().16b, 6
+    usubl2 \vtmp2\().8h, \src1\().16b, \vtmp1\().16b
+    mls \dst2\().8h, \vtmp2\().8h, \mul\().h[0]
+.endm
+
+/*
+ * shift_y
+ * Perform subpixel shift along Y axis
+ */
+
+.macro shift_y src11, src12, src21, src22, mul, mode=
+.ifb \mode
+    sub \src21\().8h, \src11\().8h, \src21\().8h
+    sqdmulh \src21\().8h, \src21\().8h, \mul\().h[1]
+.else
+    sqdmulh \src21\().8h, \src11\().8h, \mul\().h[1]
+.endif
+    sub \src11\().8h, \src11\().8h, \src21\().8h
+    uqrshrn \src11\().8b, \src11\().8h, 6
+.ifb \mode
+    sub \src22\().8h, \src12\().8h, \src22\().8h
+    sqdmulh \src22\().8h, \src22\().8h, \mul\().h[1]
+.else
+    sqdmulh \src22\().8h, \src12\().8h, \mul\().h[1]
+.endif
+    sub \src12\().8h, \src12\().8h, \src22\().8h
+    uqrshrn2 \src11\().16b, \src12\().8h, 6
+.endm
+
+/*
+ * void ass_shift_c(uint8_t *buf, ptrdiff_t stride,
+ *                  size_t width, size_t height, int shift_x, int shift_y);
+ */
+
+function shift_neon, export=1
+    add w4, w4, w5, lsl 16 + 9
+    mov v6.s[0], w4
+    add x4, x0, x1
+    sub x1, x1, x2
+    and x1, x1, ~15
+    sub x3, x3, 1
+    movi v7.16b, 0
+
+0:
+    mov x6, x0
+    ld1 {v0.16b}, [x0], #16
+    ld1 {v1.16b}, [x4], #16
+    subs x5, x2, 16
+    b.ls 2f
+1:
+    mov v16.16b, v0.16b
+    ld1 {v0.16b}, [x0], #16
+    shift_x v2, v3, v16, v0, v6, v17, v18
+    mov v16.16b, v1.16b
+    ld1 {v1.16b}, [x4], #16
+    shift_x v4, v5, v16, v1, v6, v17, v18
+    shift_y v2, v3, v4, v5, v6
+    st1 {v2.16b}, [x6], #16
+    subs x5, x5, 16
+    b.hi 1b
+2:
+    shift_x v2, v3, v0, v7, v6, v17, v18
+    shift_x v4, v5, v1, v7, v6, v17, v18
+    shift_y v2, v3, v4, v5, v6
+    st1 {v2.16b}, [x6]
+    add x0, x0, x1
+    add x4, x4, x1
+    subs x3, x3, 1
+    b.hi 0b
+
+    mov x6, x0
+    ld1 {v0.16b}, [x0], #16
+    subs x5, x2, 16
+    b.ls 4f
+3:
+    mov v16.16b, v0.16b
+    ld1 {v0.16b}, [x0], #16
+    shift_x v2, v3, v16, v0, v6, v17, v18
+    shift_y v2, v3, v4, v5, v6, last
+    st1 {v2.16b}, [x6], #16
+    subs x5, x5, 16
+    b.hi 3b
+4:
+    shift_x v2, v3, v0, v7, v6, v17, v18
+    shift_y v2, v3, v4, v5, v6, last
+    st1 {v2.16b}, [x6]
+    ret
+endfunc
diff --git a/libass/ass.c b/libass/ass.c
index b9ce811..7f46769 100644
--- a/libass/ass.c
+++ b/libass/ass.c
@@ -56,6 +56,14 @@ static const char *const ssa_style_format =
 static const char *const ssa_event_format =
         "Marked, Start, End, Style, Name, "
         "MarginL, MarginR, MarginV, Effect, Text";
+static const char *const ass2_style_format =
+        "Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, "
+        "OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, "
+        "ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, "
+        "Alignment, MarginL, MarginR, MarginV, MarginB, Encoding, RelativeTo";
+static const char *const ass2_event_format =
+        "Layer, Start, End, Style, Name, "
+        "MarginL, MarginR, MarginV, MarginB, Effect, Text";
 
 #define ASS_STYLES_ALLOC 20
 
@@ -114,6 +122,11 @@ int ass_alloc_style(ASS_Track *track)
 
     sid = track->n_styles++;
     memset(track->styles + sid, 0, sizeof(ASS_Style));
+
+    // Explicitly flag newer values as unset
+    track->styles[sid].MarginB = INT_MIN;
+    track->styles[sid].RelativeTo = 2;
+
     return sid;
 }
 
@@ -244,6 +257,8 @@ static void set_default_style(ASS_Style *style)
     style->Shadow           = 3;
     style->Alignment        = 2;
     style->MarginL = style->MarginR = style->MarginV = 20;
+    style->MarginB = INT_MIN;
+    style->RelativeTo = 2;
 }
 
 static long long string2timecode(ASS_Library *library, char *p)
@@ -522,6 +537,7 @@ static int process_event_tail(ASS_Track *track, ASS_Event *event,
             INTVAL(MarginL)
             INTVAL(MarginR)
             INTVAL(MarginV)
+            INTVAL(MarginB)
             TIMEVAL(Start)
             TIMEVAL(Duration)
         PARSE_END
@@ -653,6 +669,8 @@ static int process_style(ASS_Track *track, char *str)
         // probably an ancient script version
         if (track->track_type == TRACK_TYPE_SSA)
             track->style_format = strdup(ssa_style_format);
+        else if (track->track_type == TRACK_TYPE_ASS2)
+            track->style_format = strdup(ass2_style_format);
         else
             track->style_format = strdup(ass_style_format);
         if (!track->style_format)
@@ -708,7 +726,8 @@ static int process_style(ASS_Track *track, char *str)
             FPVAL(Angle)
             INTVAL(BorderStyle)
             INTVAL(Alignment)
-                if (track->track_type == TRACK_TYPE_ASS)
+                if (track->track_type == TRACK_TYPE_ASS ||
+                    track->track_type == TRACK_TYPE_ASS2)
                     target->Alignment = numpad2align(target->Alignment);
                 // VSFilter compatibility
                 else if (target->Alignment == 8)
@@ -718,7 +737,9 @@ static int process_style(ASS_Track *track, char *str)
             INTVAL(MarginL)
             INTVAL(MarginR)
             INTVAL(MarginV)
+            INTVAL(MarginB)
             INTVAL(Encoding)
+            INTVAL(RelativeTo)
             FPVAL(ScaleX)
             FPVAL(ScaleY)
             FPVAL(Outline)
@@ -824,7 +845,7 @@ static void custom_format_line_compatibility(ASS_Track *const track,
 static int process_styles_line(ASS_Track *track, char *str)
 {
     int ret = 0;
-    if (!strncmp(str, "Format:", 7)) {
+    if (!strncmp(str, "Format:", 7) && track->track_type != TRACK_TYPE_ASS2) {
         char *p = str + 7;
         skip_spaces(&p);
         free(track->style_format);
@@ -860,6 +881,10 @@ static inline void parse_script_type(ASS_Track *track, const char *str)
         ver = TRACK_TYPE_ASS;
         --len; --p;
     }
+    if (*(p-1) == '+') {
+        ver = TRACK_TYPE_ASS2;
+        --len; --p;
+    }
 
     if (len >= 4 && !strncmp(p-4, "4.00", 4))
         track->track_type = ver;
@@ -928,6 +953,8 @@ static void event_format_fallback(ASS_Track *track)
     track->parser_priv->state = PST_EVENTS;
     if (track->track_type == TRACK_TYPE_SSA)
         track->event_format = strdup(ssa_event_format);
+    else if (track->track_type == TRACK_TYPE_ASS2)
+        track->event_format = strdup(ass2_event_format);
     else
         track->event_format = strdup(ass_event_format);
     ass_msg(track->library, MSGL_V,
@@ -992,7 +1019,7 @@ static bool detect_legacy_conv_subs(ASS_Track *track)
 
 static int process_events_line(ASS_Track *track, char *str)
 {
-    if (!strncmp(str, "Format:", 7)) {
+    if (!strncmp(str, "Format:", 7) && track->track_type != TRACK_TYPE_ASS2) {
         char *p = str + 7;
         skip_spaces(&p);
         free(track->event_format);
@@ -1183,6 +1210,9 @@ static int process_line(ASS_Track *track, char *str)
     } else if (!ass_strncasecmp(str, "[V4+ Styles]", 12)) {
         track->parser_priv->state = PST_STYLES;
         track->track_type = TRACK_TYPE_ASS;
+    } else if (!ass_strncasecmp(str, "[V4++ Styles]", 12)) {
+        track->parser_priv->state = PST_STYLES;
+        track->track_type = TRACK_TYPE_ASS2;
     } else if (!ass_strncasecmp(str, "[Events]", 8)) {
         track->parser_priv->state = PST_EVENTS;
     } else if (!ass_strncasecmp(str, "[Fonts]", 7)) {
diff --git a/libass/ass.h b/libass/ass.h
index 851ca7c..eb8d1e6 100644
--- a/libass/ass.h
+++ b/libass/ass.h
@@ -161,7 +161,7 @@ typedef enum {
      */
     ASS_OVERRIDE_BIT_FONT_SIZE_FIELDS = 1 << 2,
     /**
-     * On dialogue events override: FontName, treat_fontname_as_pattern
+     * On dialogue events override: FontName
      */
     ASS_OVERRIDE_BIT_FONT_NAME = 1 << 3,
     /**
@@ -613,6 +613,9 @@ void ass_set_cache_limits(ASS_Renderer *priv, int glyph_max,
  * \param now video timestamp in milliseconds
  * \param detect_change compare to the previous call and set to 1
  * if positions may have changed, or set to 2 if content may have changed.
+ *
+ * The returned ASS_Images live until the next call to ass_render_frame,
+ * ass_get_metrics, or ass_renderer_done.
  */
 ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
                             long long now, int *detect_change);
diff --git a/libass/ass_bitmap.c b/libass/ass_bitmap.c
index 41df5e1..4182b45 100644
--- a/libass/ass_bitmap.c
+++ b/libass/ass_bitmap.c
@@ -147,7 +147,8 @@ bool ass_copy_bitmap(const BitmapEngine *engine, Bitmap *dst, const Bitmap *src)
 }
 
 bool ass_outline_to_bitmap(RenderContext *state, Bitmap *bm,
-                           ASS_Outline *outline1, ASS_Outline *outline2)
+                           ASS_Outline *outline1, ASS_Outline *outline2,
+                           ASS_Rect *clip)
 {
     ASS_Renderer *render_priv = state->renderer;
     RasterizerData *rst = &state->rasterizer;
@@ -163,10 +164,23 @@ bool ass_outline_to_bitmap(RenderContext *state, Bitmap *bm,
         return false;
 
     // enlarge by 1/64th of pixel to bypass slow rasterizer path, add 1 pixel for shift_bitmap
-    int32_t x_min = (rst->bbox.x_min -   1) >> 6;
-    int32_t y_min = (rst->bbox.y_min -   1) >> 6;
-    int32_t x_max = (rst->bbox.x_max + 127) >> 6;
-    int32_t y_max = (rst->bbox.y_max + 127) >> 6;
+    int32_t x_min = (rst->bbox.x_min - 64) >> 6;
+    int32_t y_min = (rst->bbox.y_min - 64) >> 6;
+    int32_t x_max = (rst->bbox.x_max + 64) >> 6;
+    int32_t y_max = (rst->bbox.y_max + 64) >> 6;
+
+
+    if (clip) {
+        x_min = FFMAX(x_min, clip->x_min);
+        y_min = FFMAX(y_min, clip->y_min);
+        x_max = FFMIN(x_max, clip->x_max);
+        y_max = FFMIN(y_max, clip->y_max);
+
+        if (x_max <= x_min || y_max <= y_min) {
+            return false;
+        }
+    }
+
     int32_t w = x_max - x_min;
     int32_t h = y_max - y_min;
 
@@ -224,37 +238,20 @@ void ass_fix_outline(Bitmap *bm_g, Bitmap *bm_o)
 }
 
 /**
- * \brief Shift a bitmap by the fraction of a pixel in x and y direction
+ * \brief Shift a bitmap in x and y direction by amount
  * expressed in 26.6 fixed point
  */
-void ass_shift_bitmap(Bitmap *bm, int shift_x, int shift_y)
+void ass_shift_bitmap(const BitmapEngine *engine, Bitmap *bm,
+                      int32_t shift_x, int32_t shift_y)
 {
-    assert((shift_x & ~63) == 0 && (shift_y & ~63) == 0);
-
-    if (!bm->buffer)
+    // Works right even for negative offsets
+    // '>>' rounds toward negative infinity, '&' returns correct remainder
+    bm->left -= -shift_x >> 6;
+    bm->top  -= -shift_y >> 6;
+    int x = -shift_x & 63;
+    int y = -shift_y & 63;
+    if (!bm->buffer || !(x | y))
         return;
 
-    int32_t w = bm->w, h = bm->h;
-    ptrdiff_t s = bm->stride;
-    uint8_t *buf = bm->buffer;
-
-    // Shift in x direction
-    if (shift_x)
-        for (int32_t y = 0; y < h; y++) {
-            for (int32_t x = w - 1; x > 0; x--) {
-                uint8_t b = buf[x + y * s - 1] * shift_x >> 6;
-                buf[x + y * s - 1] -= b;
-                buf[x + y * s] += b;
-            }
-        }
-
-    // Shift in y direction
-    if (shift_y)
-        for (int32_t x = 0; x < w; x++) {
-            for (int32_t y = h - 1; y > 0; y--) {
-                uint8_t b = buf[x + y * s - s] * shift_y >> 6;
-                buf[x + y * s - s] -= b;
-                buf[x + y * s] += b;
-            }
-        }
+    engine->shift(bm->buffer, bm->stride, bm->w, bm->h, x, y);
 }
diff --git a/libass/ass_bitmap.h b/libass/ass_bitmap.h
index c337bf1..cf86543 100644
--- a/libass/ass_bitmap.h
+++ b/libass/ass_bitmap.h
@@ -42,13 +42,15 @@ void ass_free_bitmap(Bitmap *bm);
 struct render_context;
 
 bool ass_outline_to_bitmap(struct render_context *state, Bitmap *bm,
-                           ASS_Outline *outline1, ASS_Outline *outline2);
+                           ASS_Outline *outline1, ASS_Outline *outline2,
+                           ASS_Rect *clip);
 
 void ass_synth_blur(const BitmapEngine *engine, Bitmap *bm,
                     int be, double blur_r2x, double blur_r2y);
 
+uint32_t ass_blur_padding(double r2);
 bool ass_gaussian_blur(const BitmapEngine *engine, Bitmap *bm, double r2x, double r2y);
-void ass_shift_bitmap(Bitmap *bm, int shift_x, int shift_y);
+void ass_shift_bitmap(const BitmapEngine *engine, Bitmap *bm, int32_t shift_x, int32_t shift_y);
 void ass_fix_outline(Bitmap *bm_g, Bitmap *bm_o);
 
 #endif                          /* LIBASS_BITMAP_H */
diff --git a/libass/ass_bitmap_engine.c b/libass/ass_bitmap_engine.c
index 847adc4..e7cbdf3 100644
--- a/libass/ass_bitmap_engine.c
+++ b/libass/ass_bitmap_engine.c
@@ -47,6 +47,7 @@
     BitmapBlendFunc ass_add_bitmaps_  ## suffix; \
     BitmapBlendFunc ass_imul_bitmaps_ ## suffix; \
     BitmapMulFunc   ass_mul_bitmaps_  ## suffix; \
+    ShiftFunc       ass_shift_        ## suffix; \
     BeBlurFunc      ass_be_blur_      ## suffix;
 
 #define GENERIC_FUNCTION(name, suffix) \
@@ -56,6 +57,7 @@
     GENERIC_FUNCTION(add_bitmaps,  suffix) \
     GENERIC_FUNCTION(imul_bitmaps, suffix) \
     GENERIC_FUNCTION(mul_bitmaps,  suffix) \
+    GENERIC_FUNCTION(shift,        suffix) \
     GENERIC_FUNCTION(be_blur,      suffix)
 
 
@@ -180,6 +182,7 @@ BitmapEngine ass_bitmap_engine_init(unsigned mask)
         if (flags & ASS_CPU_FLAG_X86_SSSE3) {
             ALL_PROTOTYPES(16, ssse3)
             RASTERIZER_FUNCTION(fill_generic, ssse3)
+            GENERIC_FUNCTION(shift, ssse3)
             GENERIC_FUNCTION(be_blur, ssse3)
             BLUR_FUNCTION(shrink_horz, 16, ssse3)
             BLUR_FUNCTION(expand_horz, 16, ssse3)
diff --git a/libass/ass_bitmap_engine.h b/libass/ass_bitmap_engine.h
index 68d7aa8..a0958db 100644
--- a/libass/ass_bitmap_engine.h
+++ b/libass/ass_bitmap_engine.h
@@ -48,6 +48,8 @@ typedef void BitmapMulFunc(uint8_t *restrict dst, ptrdiff_t dst_stride,
                            const uint8_t *restrict src2, ptrdiff_t src2_stride,
                            size_t width, size_t height);
 
+typedef void ShiftFunc(uint8_t *restrict buf, ptrdiff_t stride,
+                       size_t width, size_t height, int shift_x, int shift_y);
 typedef void BeBlurFunc(uint8_t *restrict buf, ptrdiff_t stride,
                         size_t width, size_t height, uint16_t *restrict tmp);
 
@@ -76,6 +78,9 @@ typedef struct {
     BitmapBlendFunc *add_bitmaps, *imul_bitmaps;
     BitmapMulFunc *mul_bitmaps;
 
+    // shift function
+    ShiftFunc *shift;
+
     // be blur function
     BeBlurFunc *be_blur;
 
diff --git a/libass/ass_blur.c b/libass/ass_blur.c
index ec91a92..5adb16a 100644
--- a/libass/ass_blur.c
+++ b/libass/ass_blur.c
@@ -145,26 +145,54 @@ typedef struct {
     int16_t coeff[8];
 } BlurMethod;
 
-static void find_best_method(BlurMethod *blur, double r2)
+static void find_scale(BlurMethod *blur, double r2)
 {
-    double mu[8];
     if (r2 < 0.5) {
         blur->level = 0;
         blur->radius = 4;
-        mu[1] = 0.085 * r2 * r2 * r2;
-        mu[0] = 0.5 * r2 - 4 * mu[1];
-        mu[2] = mu[3] = 0;
     } else {
         double frac = frexp(sqrt(0.11569 * r2 + 0.20591047), &blur->level);
         double mul = pow(0.25, blur->level);
         blur->radius = 8 - (int) ((10.1525 + 0.8335 * mul) * (1 - frac));
         blur->radius = FFMAX(blur->radius, 4);
+    }
+}
+
+static void find_best_method(BlurMethod *blur, double r2)
+{
+    find_scale(blur, r2);
+
+    double mu[8];
+    if (r2 < 0.5) {
+        mu[1] = 0.085 * r2 * r2 * r2;
+        mu[0] = 0.5 * r2 - 4 * mu[1];
+        mu[2] = mu[3] = 0;
+    } else {
+        double mul = pow(0.25, blur->level);
         calc_coeff(mu, blur->radius, r2, mul);
     }
+
     for (int i = 0; i < blur->radius; i++)
         blur->coeff[i] = (int) (0x10000 * mu[i] + 0.5);
 }
 
+static int blur_offset(const BlurMethod *blur)
+{
+    return ((2 * blur->radius + 9) << blur->level) - 5;
+}
+
+static uint32_t blurred_dim(uint32_t d, const BlurMethod *blur)
+{
+    return ((d + blur_offset(blur)) & ~((1 << blur->level) - 1)) - 4;
+}
+
+uint32_t ass_blur_padding(double r2)
+{
+    BlurMethod blur;
+    find_scale(&blur, r2);
+    return (blur_offset(&blur) + 1) / 2;
+}
+
 /**
  * \brief Perform approximate gaussian blur
  * \param r2x in: desired standard deviation along X axis squared
@@ -179,10 +207,8 @@ bool ass_gaussian_blur(const BitmapEngine *engine, Bitmap *bm, double r2x, doubl
     else find_best_method(&blur_y, r2y);
 
     uint32_t w = bm->w, h = bm->h;
-    int offset_x = ((2 * blur_x.radius + 9) << blur_x.level) - 5;
-    int offset_y = ((2 * blur_y.radius + 9) << blur_y.level) - 5;
-    uint32_t end_w = ((w + offset_x) & ~((1 << blur_x.level) - 1)) - 4;
-    uint32_t end_h = ((h + offset_y) & ~((1 << blur_y.level) - 1)) - 4;
+    uint32_t end_w = blurred_dim(w, &blur_x);
+    uint32_t end_h = blurred_dim(h, &blur_y);
 
     const int stripe_width = 1 << (engine->align_order - 1);
     uint64_t size = (((uint64_t) end_w + stripe_width - 1) & ~(stripe_width - 1)) * end_h;
diff --git a/libass/ass_cache_template.h b/libass/ass_cache_template.h
index 84d6f4c..f9c928a 100644
--- a/libass/ass_cache_template.h
+++ b/libass/ass_cache_template.h
@@ -8,6 +8,8 @@
         ASS_StringView member;
 #define VECTOR(member) \
         ASS_Vector member;
+#define RECT(member) \
+        ASS_Rect member;
 #define END(typedefnamename) \
     } typedefnamename;
 
@@ -25,6 +27,10 @@
             ass_string_equal(a->member, b->member) &&
 #define VECTOR(member) \
             a->member.x == b->member.x && a->member.y == b->member.y &&
+#define RECT(member) \
+            a->member.x_min == b->member.x_min && a->member.y_min == b->member.y_min && \
+            a->member.x_max == b->member.x_max && a->member.y_max == b->member.y_max &&
+
 #define END(typedefname) \
             true; \
     }
@@ -40,6 +46,9 @@
 #define STRING(member) \
         hval = ass_hash_buf(p->member.str, p->member.len, hval);
 #define VECTOR(member) GENERIC(, member.x); GENERIC(, member.y);
+#define RECT(member) \
+        GENERIC(, member.x_min); GENERIC(, member.y_min); \
+        GENERIC(, member.x_max); GENERIC(, member.y_max);
 #define END(typedefname) \
         return hval; \
     }
@@ -66,6 +75,8 @@ START(bitmap, bitmap_hash_key)
     VECTOR(matrix_x)
     VECTOR(matrix_y)
     VECTOR(matrix_z)
+    // clip region
+    RECT(clip)
 END(BitmapHashKey)
 
 // font is refed when inserted and unrefed when dropped
@@ -135,4 +146,5 @@ END(BitmapRef)
 #undef GENERIC
 #undef STRING
 #undef VECTOR
+#undef RECT
 #undef END
diff --git a/libass/ass_compat.h b/libass/ass_compat.h
index cdb58ce..8a44ff2 100644
--- a/libass/ass_compat.h
+++ b/libass/ass_compat.h
@@ -30,6 +30,14 @@
 #define inline __inline
 #endif
 
+#ifndef __has_builtin
+#define __has_builtin(...) 0
+#endif
+
+#ifndef __has_attribute
+#define __has_attribute(...) 0
+#endif
+
 // Work around build failures on Windows with static FriBidi.
 // Possible because we only use FriBidi functions, not objects.
 #if defined(_WIN32) && !defined(FRIBIDI_LIB_STATIC)
diff --git a/libass/ass_metrics.h b/libass/ass_metrics.h
new file mode 100644
index 0000000..d2cf3b0
--- /dev/null
+++ b/libass/ass_metrics.h
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2024 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef LIBASS_METRICS_H
+#define LIBASS_METRICS_H
+
+#include <stdio.h>
+
+#include "ass_types.h"
+
+/*
+ * API to obtain text metrics and shape data for subtitles.
+ * This API is intended to ease the authoring of subtitles by providing
+ * a unified and platform-independent method to obtain metrics.
+ * However, there are no guarantees for the stability of the data returned by
+ * this API. With future changes to libass and its rendering, the returned
+ * metrics and shape data may change too.
+ *
+ * For example, subtitle authors could use this API to determine when and
+ * where libass automatically breaks a subtitle line into two, but there is
+ * no guarantee that newer or older versions of libass (as well as other
+ * renderers) will always break this line in the same way. Subtitle authors
+ * that want to ensure that a line is broken in a certain way should instead
+ * use manual line breaks and overrides (e.g. \q2).
+ *
+ * API Usage:
+ * After initializing and configuring an ASS_Renderer and ASS_Track in the
+ * same way as one would for normal subtitle rendering, users may call
+ * ass_get_metrics to obtain metrics for all events rendered at the given
+ * point in time.
+ * This returns a list of event metrics, each of which consisting of a list
+ * of runs, with metrics for those runs. A run contains lists of shapes
+ * which, when drawn together, make up the shape for the fill or border of
+ * the run.
+ * The individual shapes correspond to individual glyphs rendered in the run,
+ * but due to font shaping there is no guarantee for these glyphs to
+ * correspond one-to-one to characters in the run's text.
+ */
+
+
+/*
+ * An outline is represented by an array of points and an array of segments.
+ * Segments can be splines of order 1 (line), 2 (quadratic) or 3 (cubic).
+ * Each segment owns a number of points equal to its order in the point array
+ * and uses the first point owned by the next segment as its last point.
+ * The last segment in each contour uses the first point owned by the first
+ * segment in each contour as its last point.
+ * Correspondingly, the total number of points is equal to the sum of the
+ * spline orders of all segments.
+ */
+
+enum {
+    ASS_METRICS_OUTLINE_LINE_SEGMENT     = 1,  // line segment
+    ASS_METRICS_OUTLINE_QUADRATIC_SPLINE = 2,  // quadratic spline
+    ASS_METRICS_OUTLINE_CUBIC_SPLINE     = 3,  // cubic spline
+    ASS_METRICS_OUTLINE_COUNT_MASK       = 3,  // spline order mask
+    ASS_METRICS_OUTLINE_CONTOUR_END      = 4   // last segment in contour flag
+};
+
+typedef struct ass_metrics_outline {
+    size_t n_points;
+    size_t n_segments;
+    ASS_DVector *points;
+    char *segments;
+
+    struct ass_metrics_outline *next;   // Next outline, or NULL
+} ASS_Metrics_Outline;
+
+/* ASS_RunMetrics is a set of metrics and shape data corresponding to a single run. */
+typedef struct ass_run_metrics {
+    ASS_DVector pos;
+    ASS_DVector advance;
+    double asc;
+    double desc;
+    // TODO: Add more fields here?
+
+    ASS_Metrics_Outline *fill;
+    ASS_Metrics_Outline *border;
+
+    struct ass_run_metrics *next;    // Next run, or NULL
+} ASS_RunMetrics;
+
+/* ASS_Metrics is a set of metrics corresponding to a single event. */
+typedef struct ass_metrics {
+    ASS_Event *event;   // Becomes invalid once the containing ASS_Track is modified, pruned, or freed
+    ASS_RunMetrics *runs;
+    struct ass_metrics *next;   // Next set of metrics, or NULL
+} ASS_Metrics;
+
+/**
+ * \brief Get metrics for a frame, producing a list of ASS_Metrics.
+ * The metrics returned live until the next call to ass_get_metrics, ass_render_frame,
+ * or ass_renderer_done.
+ *
+ * \param priv renderer handle
+ * \param track subtitle track
+ * \param now video timestamp in milliseconds
+ */
+ASS_Metrics *ass_get_metrics(ASS_Renderer *priv, ASS_Track *track, long long now);
+
+#endif /* LIBASS_METRICS_H */
diff --git a/libass/ass_outline.c b/libass/ass_outline.c
index 188f07b..aa1fe82 100644
--- a/libass/ass_outline.c
+++ b/libass/ass_outline.c
@@ -77,6 +77,48 @@ void ass_outline_free(ASS_Outline *outline)
 }
 
 
+/*
+ * \brief Allocate an ASS_Metrics_Outline's memory and copy the given
+ * ASS_Outline's data into it.
+ */
+void ass_metric_outline_copy(ASS_Metrics_Outline *metrics_outline, ASS_Outline *outline)
+{
+    metrics_outline->n_points = outline->n_points;
+    metrics_outline->n_segments = outline->n_segments;
+    metrics_outline->points = malloc(sizeof(ASS_DVector) * outline->n_points);
+    metrics_outline->segments = malloc(outline->n_segments);
+    if (!metrics_outline->points || !metrics_outline->segments) {
+        free(metrics_outline->points);
+        free(metrics_outline->segments);
+        metrics_outline->n_points = 0;
+        metrics_outline->n_segments = 0;
+        return;
+    }
+
+    memcpy(metrics_outline->segments, outline->segments, outline->n_segments);
+
+    for (int i = 0; i < outline->n_points; i++) {
+        metrics_outline->points[i].x = d6_to_double(outline->points[i].x);
+        metrics_outline->points[i].y = d6_to_double(outline->points[i].y);
+    }
+}
+
+/*
+ * \brief Free the point and segments lists of an ASS_Metrics_Outline
+ */
+void ass_metric_outline_free(ASS_Metrics_Outline *metrics_outline)
+{
+    if (!metrics_outline)
+        return;
+
+    free(metrics_outline->points);
+    free(metrics_outline->segments);
+    metrics_outline->points = NULL;
+    metrics_outline->segments = NULL;
+    metrics_outline->n_points = 0;
+    metrics_outline->n_segments = 0;
+}
+
 static bool valid_point(const FT_Vector *pt)
 {
     return labs(pt->x) <= OUTLINE_MAX && labs(pt->y) <= OUTLINE_MAX;
diff --git a/libass/ass_outline.h b/libass/ass_outline.h
index 25848b5..0ec7549 100644
--- a/libass/ass_outline.h
+++ b/libass/ass_outline.h
@@ -24,6 +24,7 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#include "ass_metrics.h"
 #include "ass_utils.h"
 
 
@@ -31,10 +32,6 @@ typedef struct {
     int32_t x, y;
 } ASS_Vector;
 
-typedef struct {
-    double x, y;
-} ASS_DVector;
-
 typedef struct {
     int32_t x_min, y_min, x_max, y_max;
 } ASS_Rect;
@@ -91,6 +88,9 @@ void ass_outline_clear(ASS_Outline *outline);
 bool ass_outline_alloc(ASS_Outline *outline, size_t n_points, size_t n_segments);
 void ass_outline_free(ASS_Outline *outline);
 
+void ass_metric_outline_copy(ASS_Metrics_Outline *metrics_outline, ASS_Outline *outline);
+void ass_metric_outline_free(ASS_Metrics_Outline *metrics_outline);
+
 // expects preallocated outline and works inplace
 bool ass_outline_convert(ASS_Outline *outline, const FT_Outline *source);
 void ass_outline_add_rect(ASS_Outline *outline,
diff --git a/libass/ass_parse.c b/libass/ass_parse.c
index 28d49e7..99833c5 100644
--- a/libass/ass_parse.c
+++ b/libass/ass_parse.c
@@ -219,7 +219,7 @@ static bool parse_vector_clip(RenderContext *state,
 
     int scale = 1;
     if (nargs == 2)
-        scale = argtoi32(args[0]);
+        scale = FFMAX(argtoi32(args[0]), 1);
 
     struct arg text = args[nargs - 1];
     state->clip_drawing_text.str = text.start;
@@ -966,6 +966,8 @@ void ass_apply_transition_effects(RenderContext *state)
         delay = ((int) FFMAX(delay / scale_x, 1)) * scale_x;
         state->scroll_shift =
             (render_priv->time - event->Start) / delay;
+        state->scroll_fade_width =
+            cnt >= 3 ? ((int) (v[2] * scale_x)) / scale_x : 0;
         state->evt_type |= EVENT_HSCROLL;
         state->detect_collisions = 0;
         state->wrap_style = 2;
@@ -1005,6 +1007,8 @@ void ass_apply_transition_effects(RenderContext *state)
         }
         state->scroll_y0 = y0;
         state->scroll_y1 = y1;
+        state->scroll_fade_width =
+            cnt >= 4 ? ((int) (v[3] * scale_y)) / scale_y : 0;
         state->evt_type |= EVENT_VSCROLL;
         state->detect_collisions = 0;
     }
diff --git a/libass/ass_render.c b/libass/ass_render.c
index 18f856a..457e92c 100644
--- a/libass/ass_render.c
+++ b/libass/ass_render.c
@@ -29,17 +29,18 @@
 #endif
 
 #include "ass.h"
+#include "ass_metrics.h"
 #include "ass_outline.h"
 #include "ass_render.h"
 #include "ass_parse.h"
 #include "ass_priv.h"
 #include "ass_shaper.h"
+#include "ass_utils.h"
 
 #define MAX_GLYPHS_INITIAL 1024
 #define MAX_LINES_INITIAL 64
 #define MAX_BITMAPS_INITIAL 16
 #define MAX_SUB_BITMAPS_INITIAL 64
-#define SUBPIXEL_MASK 63
 #define STROKER_PRECISION 16     // stroker error in integer units, unrelated to final accuracy
 #define RASTERIZER_PRECISION 16  // rasterizer spline approximation error in 1/64 pixel units
 #define POSITION_PRECISION 8.0   // rough estimate of transform error in 1/64 pixel units
@@ -177,6 +178,7 @@ void ass_renderer_done(ASS_Renderer *render_priv)
     if (!render_priv)
         return;
 
+    ass_free_metrics(render_priv);
     ass_frame_unref(render_priv->images_root);
     ass_frame_unref(render_priv->prev_images_root);
 
@@ -235,6 +237,20 @@ static ASS_Image *my_draw_bitmap(unsigned char *bitmap, int bitmap_w,
     return &img->result;
 }
 
+static inline ASS_Image *ass_image_borrow_copy(const ASS_Image *img)
+{
+    ASS_Image *nimg = my_draw_bitmap(img->bitmap, img->w, img->h, img->stride,
+                                     img->dst_x, img->dst_y, img->color,
+                                     ((ASS_ImagePriv *) img)->source);
+    if (!nimg)
+        return NULL;
+    nimg->next = img->next;
+    nimg->type = img->type;
+    // Borrow the buffer from the original ASS_Image
+    ((ASS_ImagePriv *) nimg)->buffer = NULL;
+    return nimg;
+}
+
 /**
  * \brief Mapping between script and screen coordinates
  */
@@ -497,6 +513,7 @@ render_glyph(RenderContext *state, Bitmap *bm, int dst_x, int dst_y,
 
 static bool quantize_transform(double m[3][3], ASS_Vector *pos,
                                ASS_DVector *offset, bool first,
+                               ASS_Rect *clip,
                                BitmapHashKey *key)
 {
     // Full transform:
@@ -645,6 +662,33 @@ static bool quantize_transform(double m[3][3], ASS_Vector *pos,
     key->matrix_x.x = qm[0][0];  key->matrix_x.y = qm[0][1];
     key->matrix_y.x = qm[1][0];  key->matrix_y.y = qm[1][1];
     key->matrix_z.x = qm[2][0];  key->matrix_z.y = qm[2][1];
+
+    if (clip && !(
+        ((qr[0] + key->outline->cbox.x_min) < (int64_t)((uint64_t)clip->x_min << SUBPIXEL_ORDER)) ||
+        ((qr[1] + key->outline->cbox.y_min) < (int64_t)((uint64_t)clip->y_min << SUBPIXEL_ORDER)) ||
+        ((qr[0] + key->outline->cbox.x_max) > (int64_t)((uint64_t)clip->x_max << SUBPIXEL_ORDER)) ||
+        ((qr[1] + key->outline->cbox.y_max) > (int64_t)((uint64_t)clip->y_max << SUBPIXEL_ORDER))
+    ))
+        clip = NULL;
+
+#define CLIP_PRECISION 64
+#define CLIP_PRECISION_MASK (~(CLIP_PRECISION - 1))
+#define CLIP_MIN(x) ((x) & CLIP_PRECISION_MASK)
+#define CLIP_MAX(x) (((x) + CLIP_PRECISION) & CLIP_PRECISION_MASK)
+    if (clip) {
+        key->clip = (ASS_Rect){
+            CLIP_MIN(clip->x_min - pos->x),
+            CLIP_MIN(clip->y_min - pos->y),
+            CLIP_MAX(clip->x_max - pos->x),
+            CLIP_MAX(clip->y_max - pos->y),
+        };
+    } else {
+        key->clip = (ASS_Rect){
+            INT32_MIN, INT32_MIN,
+            INT32_MAX, INT32_MAX,
+        };
+    }
+
     return true;
 }
 
@@ -692,6 +736,135 @@ static inline size_t bitmap_size(const Bitmap *bm)
     return bm->stride * bm->h;
 }
 
+static void apply_hscroll_fade(RenderContext *state, ASS_Image *head)
+{
+    double fade_width = state->scroll_fade_width * state->screen_scale_x;
+    if (fade_width <= 0)
+        return;
+
+    ASS_Renderer *render_priv = state->renderer;
+
+    for (ASS_Image *cur = head; cur; cur = cur->next) {
+        int event_start = cur->dst_x - render_priv->settings.left_margin;
+        int event_end = event_start + cur->w;
+
+        double leading_fade_start = 0.5 * state->blur_scale_x - 0.5;
+        int leading_fade_start_int = floor(leading_fade_start) + 1;
+        double leading_fade_end = leading_fade_start + fade_width;
+        int leading_fade_end_int = ceil(leading_fade_end);
+
+        // int trailing_fade_end_int = render_priv->frame_content_width;
+        // double trailing_fade_end = trailing_fade_end_int + leading_fade_start;
+        double trailing_fade_end = render_priv->frame_content_width + leading_fade_start;
+        int trailing_fade_end_int = ceil(trailing_fade_end);
+        double trailing_fade_start = trailing_fade_end - fade_width;
+        int trailing_fade_start_int = floor(trailing_fade_start) + 1;
+
+        int full_a = _a(cur->color);
+        cur->color |= 0xFF;  // starts out fully transparent
+
+        for (int x = FFMAX(event_start, leading_fade_start_int); x < event_end && x <= trailing_fade_end_int; ) {
+            double exact_a = 0xFF - full_a;
+            if (x < leading_fade_end_int)
+                exact_a *= (x - leading_fade_start) / fade_width;
+            if (x >= trailing_fade_start_int)
+                exact_a *= FFMAX(trailing_fade_end - x, 0) / fade_width;
+
+            int a = 0xFF - lround(exact_a);
+            if (a != _a(cur->color)) {
+                if (x > event_start) {
+                    ASS_Image *nimg = ass_image_borrow_copy(cur);
+                    nimg->dst_x = x + render_priv->settings.left_margin;
+                    cur->w = nimg->dst_x - cur->dst_x;
+                    nimg->w -= cur->w;
+                    nimg->bitmap += cur->w;
+                    cur->next = nimg;
+                    cur = nimg;
+                }
+                cur->color = (cur->color & ~0xFF) | a;
+            }
+
+            if (x >= leading_fade_end_int && x < trailing_fade_start_int)
+                x = trailing_fade_start_int;
+            else
+                x++;
+        }
+    }
+}
+
+static void apply_vscroll_fade(RenderContext *state, ASS_Image *head)
+{
+    double fade_height = state->scroll_fade_width * state->screen_scale_y;
+    if (fade_height <= 0)
+        return;
+
+    ASS_Renderer *render_priv = state->renderer;
+
+    for (ASS_Image *cur = head; cur; cur = cur->next) {
+        int event_start = cur->dst_y - render_priv->settings.top_margin;
+        int event_end = event_start + cur->h;
+
+        ASS_Vector layout_res = ass_layout_res(render_priv);
+        double scale_y = ((double) layout_res.y) / render_priv->track->PlayResY;
+        double half_pixel = 0.5 * state->blur_scale_y - 0.5;
+
+        double leading_fade_start =
+            ((int32_t) (state->scroll_y0 * scale_y * 8) >> 3) * state->blur_scale_y + half_pixel;
+        int leading_fade_start_int = floor(leading_fade_start) + 1;
+        double leading_fade_end = leading_fade_start + fade_height;
+        int leading_fade_end_int = ceil(leading_fade_end);
+
+        double trailing_fade_end =
+            ((int32_t) (state->scroll_y1 * scale_y * 8) >> 3) * state->blur_scale_y + half_pixel;
+        int trailing_fade_end_int = ceil(trailing_fade_end);
+        double trailing_fade_start = trailing_fade_end - fade_height;
+        int trailing_fade_start_int = floor(trailing_fade_start) + 1;
+
+        int full_a = _a(cur->color);
+        cur->color |= 0xFF;  // starts out fully transparent
+
+        for (int y = FFMAX(event_start, leading_fade_start_int); y < event_end && y <= trailing_fade_end_int; ) {
+            double exact_a = 0xFF - full_a;
+            if (y < leading_fade_end_int)
+                exact_a *= (y - leading_fade_start) / fade_height;
+            if (y >= trailing_fade_start_int)
+                exact_a *= FFMAX(trailing_fade_end - y, 0) / fade_height;
+
+            int a = 0xFF - lround(exact_a);
+            if (a != _a(cur->color)) {
+                if (y > event_start) {
+                    ASS_Image *nimg = ass_image_borrow_copy(cur);
+                    nimg->dst_y = y + render_priv->settings.top_margin;
+                    cur->h = nimg->dst_y - cur->dst_y;
+                    nimg->h -= cur->h;
+                    nimg->bitmap += cur->h * cur->stride;
+                    cur->next = nimg;
+                    cur = nimg;
+                }
+                cur->color = (cur->color & ~0xFF) | a;
+            }
+
+            if (y >= leading_fade_end_int && y < trailing_fade_start_int)
+                y = trailing_fade_start_int;
+            else
+                y++;
+        }
+    }
+}
+
+/**
+ * Iterate through a list of bitmaps and apply the fade gradient
+ * from a Banner or Scroll effect, if applicable. Each image is
+ * replaced by multiple cropped copies with different alphas.
+ */
+static void apply_scroll_fade(RenderContext *state, ASS_Image *head)
+{
+    if (state->evt_type & EVENT_HSCROLL)
+        apply_hscroll_fade(state, head);
+    else if (state->evt_type & EVENT_VSCROLL)
+        apply_vscroll_fade(state, head);
+}
+
 /**
  * Iterate through a list of bitmaps and blend with clip vector, if
  * applicable. The blended bitmaps are added to a free list which is freed
@@ -718,11 +891,18 @@ static void blend_vector_clip(RenderContext *state, ASS_Image *head)
     m[0][2] = int_to_d6(render_priv->settings.left_margin);
     m[1][2] = int_to_d6(render_priv->settings.top_margin);
 
+    ASS_Rect clip = (ASS_Rect){
+        0,
+        0,
+        render_priv->settings.frame_width,
+        render_priv->settings.frame_height,
+    };
+
     ASS_Vector pos;
     BitmapHashKey key;
     key.outline = ass_cache_get(render_priv->cache.outline_cache, &ol_key, render_priv);
     if (!key.outline || !key.outline->valid ||
-            !quantize_transform(m, &pos, NULL, true, &key))
+            !quantize_transform(m, &pos, NULL, true, &clip, &key))
         return;
 
     Bitmap *clip_bm = ass_cache_get(render_priv->cache.bitmap_cache, &key, state);
@@ -879,6 +1059,7 @@ static ASS_Image *render_text(RenderContext *state)
 
     *tail = 0;
     blend_vector_clip(state, head);
+    apply_scroll_fade(state, head);
 
     return head;
 }
@@ -958,7 +1139,6 @@ static ASS_Style *handle_selective_style_overrides(RenderContext *state,
 
     if (requested & ASS_OVERRIDE_BIT_FONT_NAME) {
         new->FontName = user->FontName;
-        new->treat_fontname_as_pattern = user->treat_fontname_as_pattern;
     }
 
     if (requested & ASS_OVERRIDE_BIT_COLORS) {
@@ -993,9 +1173,13 @@ static ASS_Style *handle_selective_style_overrides(RenderContext *state,
     if (requested & ASS_OVERRIDE_BIT_MARGINS) {
         new->MarginL = user->MarginL;
         new->MarginR = user->MarginR;
-        new->MarginV = user->MarginV;
+        new->MarginT = user->MarginT;
+        new->MarginB = user->MarginB;
     }
 
+    // Never override RelativeTo; there are separate mechanisms for that
+    new->RelativeTo = rstyle->RelativeTo;
+
     if (!new->FontName)
         new->FontName = rstyle->FontName;
 
@@ -1089,7 +1273,6 @@ void ass_reset_render_context(RenderContext *state, ASS_Style *style)
 
     state->family.str = style->FontName;
     state->family.len = strlen(style->FontName);
-    state->treat_family_as_pattern = style->treat_fontname_as_pattern;
     state->bold = style->Bold;
     state->italic = style->Italic;
     ass_update_font(state);
@@ -1118,6 +1301,10 @@ init_render_context(RenderContext *state, ASS_Event *event)
 {
     ASS_Renderer *render_priv = state->renderer;
 
+    state->collect_metrics = false;
+    state->run_metrics = NULL;
+    state->current_run_metrics = NULL;
+
     state->event = event;
     state->parsed_tags = 0;
     state->evt_type = EVENT_NORMAL;
@@ -1372,6 +1559,35 @@ static void calc_transform_matrix(RenderContext *state,
     }
 }
 
+static void metrics_append_outline(RenderContext *state, ASS_Metrics_Outline **metrics_outline, BitmapHashKey *key, ASS_Vector *shift)
+{
+    if (!state->collect_metrics)
+        return;
+
+    while (*metrics_outline)
+        metrics_outline = &(**metrics_outline).next;
+
+    *metrics_outline = calloc(1, sizeof(ASS_Metrics_Outline));
+    if (!*metrics_outline)
+        return;
+
+    ASS_Metrics_Outline *result = *metrics_outline;
+
+    ASS_Outline outline[2];
+    ass_outline_apply_transform(outline, key);
+    ass_metric_outline_copy(result, &outline[0]);
+
+    ass_outline_free(&outline[0]);
+    ass_outline_free(&outline[1]);
+
+    for (int i = 0; i < result->n_points; i++) {
+        result->points[i].x += shift->x - state->current_run_metrics->pos.x;
+        result->points[i].y += shift->y - state->current_run_metrics->pos.y;
+    }
+}
+
+int ass_be_padding(int be);
+
 /**
  * \brief Get bitmaps for a glyph
  * \param info glyph info
@@ -1404,11 +1620,68 @@ get_bitmap_glyph(RenderContext *state, GlyphInfo *info,
     if (info->effect_type == EF_KARAOKE_KF)
         ass_outline_update_min_transformed_x(&info->outline->outline[0], m, leftmost_x);
 
+    ASS_Rect clip_rect = {
+        0,
+        0,
+        render_priv->settings.frame_width,
+        render_priv->settings.frame_height,
+    };
+
+#ifdef USE_RECT_CLIP
+    if (!state->clip_mode) {
+        clip_rect = (ASS_Rect){
+            state->clip_x0,
+            state->clip_y0,
+            state->clip_x1,
+            state->clip_y1,
+        };
+    }
+#endif
+
+    bool potentially_clipped = (
+        info->pos.x + info->outline->cbox.x_min < clip_rect.x_min << SUBPIXEL_ORDER ||
+        info->pos.y + info->outline->cbox.y_min < clip_rect.y_min << SUBPIXEL_ORDER ||
+        info->pos.x + info->outline->cbox.x_max > clip_rect.x_max << SUBPIXEL_ORDER ||
+        info->pos.y + info->outline->cbox.y_max > clip_rect.y_max << SUBPIXEL_ORDER
+    );
+
+    if (potentially_clipped) {
+        double blur_radius_scale = 2 / sqrt(log(256));
+        double blur_scale_x = state->blur_scale_x * blur_radius_scale;
+        double blur_scale_y = state->blur_scale_y * blur_radius_scale;
+        double blur_x = info->blur * blur_scale_x;
+        double blur_y = info->blur * blur_scale_y;
+
+        uint32_t xpad = ass_blur_padding(blur_x) + ass_be_padding(info->be);
+        uint32_t ypad = ass_blur_padding(blur_y) + ass_be_padding(info->be);
+
+        clip_rect.x_min -= xpad;
+        clip_rect.y_min -= ypad;
+        clip_rect.x_max += xpad;
+        clip_rect.y_max += ypad;
+
+        int32_t shadow_x = double_to_d6(info->shadow_x * state->border_scale_x);
+        int32_t shadow_y = double_to_d6(info->shadow_y * state->border_scale_y);
+
+        if (shadow_x > 0)
+            clip_rect.x_min -= (shadow_x + 63) >> 6;
+        else
+            clip_rect.x_max -= (shadow_x - 63) >> 6;
+        if (shadow_y > 0)
+            clip_rect.y_min -= (shadow_y + 63) >> 6;
+        else
+            clip_rect.y_max -= (shadow_y - 63) >> 6;
+    }
+
     BitmapHashKey key;
     key.outline = info->outline;
-    if (!quantize_transform(m, pos, offset, first, &key))
+    if (!quantize_transform(m, pos, offset, first, potentially_clipped ? &clip_rect : NULL, &key))
         return;
 
+    if (state->collect_metrics) {
+        metrics_append_outline(state, &state->current_run_metrics->fill, &key, pos);
+    }
+
     info->bm = ass_cache_get(render_priv->cache.bitmap_cache, &key, state);
     if (!info->bm || !info->bm->buffer)
         info->bm = NULL;
@@ -1530,9 +1803,13 @@ get_bitmap_glyph(RenderContext *state, GlyphInfo *info,
 
     key.outline = ass_cache_get(render_priv->cache.outline_cache, &ol_key, render_priv);
     if (!key.outline || !key.outline->valid ||
-            !quantize_transform(m, pos_o, offset, false, &key))
+            !quantize_transform(m, pos_o, offset, false, potentially_clipped ? &clip_rect : NULL, &key))
         return;
 
+    if (state->collect_metrics) {
+        metrics_append_outline(state, &state->current_run_metrics->border, &key, pos_o);
+    }
+
     info->bm_o = ass_cache_get(render_priv->cache.bitmap_cache, &key, state);
     if (!info->bm_o || !info->bm_o->buffer) {
         info->bm_o = NULL;
@@ -1546,16 +1823,11 @@ static inline size_t outline_size(const ASS_Outline* outline)
     return sizeof(ASS_Vector) * outline->n_points + outline->n_segments;
 }
 
-size_t ass_bitmap_construct(void *key, void *value, void *priv)
+void ass_outline_apply_transform(ASS_Outline *outline, BitmapHashKey *k)
 {
-    RenderContext *state = priv;
-    BitmapHashKey *k = key;
-    Bitmap *bm = value;
-
     double m[3][3];
     restore_transform(m, k);
 
-    ASS_Outline outline[2];
     if (k->matrix_z.x || k->matrix_z.y) {
         ass_outline_transform_3d(&outline[0], &k->outline->outline[0], m);
         ass_outline_transform_3d(&outline[1], &k->outline->outline[1], m);
@@ -1563,8 +1835,18 @@ size_t ass_bitmap_construct(void *key, void *value, void *priv)
         ass_outline_transform_2d(&outline[0], &k->outline->outline[0], m);
         ass_outline_transform_2d(&outline[1], &k->outline->outline[1], m);
     }
+}
+
+size_t ass_bitmap_construct(void *key, void *value, void *priv)
+{
+    RenderContext *state = priv;
+    BitmapHashKey *k = key;
+    Bitmap *bm = value;
+
+    ASS_Outline outline[2];
+    ass_outline_apply_transform(outline, k);
 
-    if (!ass_outline_to_bitmap(state, bm, &outline[0], &outline[1]))
+    if (!ass_outline_to_bitmap(state, bm, &outline[0], &outline[1], &k->clip))
         memset(bm, 0, sizeof(*bm));
     ass_outline_free(&outline[0]);
     ass_outline_free(&outline[1]);
@@ -2071,8 +2353,7 @@ static bool parse_events(RenderContext *state, ASS_Event *event)
                 q = p;
                 if (*p == '{')
                     q++;
-                while ((*q != '{') && (*q != 0))
-                    q++;
+                q = ass_strchrnul(q, '{');
                 drawing_text.str = p;
                 drawing_text.len = q - p;
                 code = 0xfffc; // object replacement character
@@ -2487,6 +2768,19 @@ static void render_and_combine_glyphs(RenderContext *state,
                 flags |= FILTER_FILL_IN_BORDER;
 
             if (new_run) {
+                if (state->collect_metrics) {
+                    ASS_RunMetrics *run_metrics = calloc(1, sizeof(ASS_RunMetrics));
+                    if (run_metrics) {
+                        if (!state->run_metrics)
+                            state->run_metrics = run_metrics;
+
+                        if (state->current_run_metrics)
+                            state->current_run_metrics->next = run_metrics;
+
+                        state->current_run_metrics = run_metrics;
+                    }
+                }
+
                 if (nb_bitmaps >= text_info->max_bitmaps) {
                     size_t new_size = 2 * text_info->max_bitmaps;
                     if (!ASS_REALLOC_ARRAY(text_info->combined_bitmaps, new_size))
@@ -2535,13 +2829,27 @@ static void render_and_combine_glyphs(RenderContext *state,
                 current_info->max_bitmap_count = MAX_SUB_BITMAPS_INITIAL;
 
                 nb_bitmaps++;
-                new_run = false;
             }
             assert(current_info);
 
             ASS_Vector pos, pos_o;
             info->pos.x = double_to_d6(device_x + d6_to_double(info->pos.x) * render_priv->par_scale_x);
             info->pos.y = double_to_d6(device_y) + info->pos.y;
+
+            if (state->collect_metrics) {
+                assert(state->current_run_metrics);
+                if (new_run) {
+                    state->current_run_metrics->pos.x = d6_to_double(info->pos.x);
+                    state->current_run_metrics->pos.y = d6_to_double(info->pos.y);
+                }
+                state->current_run_metrics->advance.x += d6_to_double(info->cluster_advance.x);
+                state->current_run_metrics->advance.y += d6_to_double(info->cluster_advance.y);
+                state->current_run_metrics->asc = FFMAX(state->current_run_metrics->asc, d6_to_double(info->asc));
+                state->current_run_metrics->desc = FFMAX(state->current_run_metrics->desc, d6_to_double(info->desc));
+            }
+
+            new_run = false;
+
             get_bitmap_glyph(state, info, &current_info->leftmost_x, &pos, &pos_o,
                              &offset, !current_info->bitmap_count, flags);
 
@@ -2740,12 +3048,7 @@ size_t ass_composite_construct(void *key, void *value, void *priv)
         } else {
             ass_copy_bitmap(&render_priv->engine, &v->bm_s, &v->bm);
         }
-
-        // Works right even for negative offsets
-        // '>>' rounds toward negative infinity, '&' returns correct remainder
-        v->bm_s.left += k->filter.shadow.x >> 6;
-        v->bm_s.top  += k->filter.shadow.y >> 6;
-        ass_shift_bitmap(&v->bm_s, k->filter.shadow.x & SUBPIXEL_MASK, k->filter.shadow.y & SUBPIXEL_MASK);
+        ass_shift_bitmap(&render_priv->engine, &v->bm_s, k->filter.shadow.x, k->filter.shadow.y);
     }
 
     if ((flags & FILTER_FILL_IN_SHADOW) && !(flags & FILTER_FILL_IN_BORDER))
@@ -2793,11 +3096,12 @@ static void add_background(RenderContext *state, EventImages *event_images)
  * \brief Main ass rendering function, glues everything together
  * \param event event to render
  * \param event_images struct containing resulting images, will also be initialized
+ * \param whether to collect metrics
  * Process event, appending resulting ASS_Image's to images_root.
  */
 static bool
 ass_render_event(RenderContext *state, ASS_Event *event,
-                 EventImages *event_images)
+                 EventImages *event_images, bool collect_metrics)
 {
     ASS_Renderer *render_priv = state->renderer;
     if (event->Style >= render_priv->track->n_styles) {
@@ -2811,6 +3115,7 @@ ass_render_event(RenderContext *state, ASS_Event *event,
 
     free_render_context(state);
     init_render_context(state, event);
+    state->collect_metrics = collect_metrics;
 
     if (!parse_events(state, event))
         return false;
@@ -2845,8 +3150,13 @@ ass_render_event(RenderContext *state, ASS_Event *event,
         (event->MarginL) ? event->MarginL : state->style->MarginL;
     int MarginR =
         (event->MarginR) ? event->MarginR : state->style->MarginR;
-    int MarginV =
-        (event->MarginV) ? event->MarginV : state->style->MarginV;
+    int MarginT =
+        (event->MarginT) ? event->MarginT : state->style->MarginT;
+    int MarginB =
+        (event->MarginB) ? event->MarginB : state->style->MarginB;
+
+    if (MarginB == INT_MIN)
+        MarginB = MarginT;
 
     // calculate max length of a line
     double max_text_width =
@@ -2918,7 +3228,7 @@ ass_render_event(RenderContext *state, ASS_Event *event,
         if (valign == VALIGN_TOP) {     // toptitle
             device_y =
                 y2scr_top(state,
-                          MarginV) + text_info->lines[0].asc;
+                          MarginT) + text_info->lines[0].asc;
         } else if (valign == VALIGN_CENTER) {   // midtitle
             double scr_y =
                 y2scr(state, render_priv->track->PlayResY / 2.0);
@@ -2932,7 +3242,7 @@ ass_render_event(RenderContext *state, ASS_Event *event,
                        "Invalid valign, assuming 0 (subtitle)");
             scr_bottom =
                 y2scr_sub(state,
-                          render_priv->track->PlayResY - MarginV);
+                          render_priv->track->PlayResY - MarginB);
             scr_top = y2scr_top(state, 0); //xxx not always 0?
             device_y = scr_bottom + (scr_top - scr_bottom) * line_pos / 100.0;
             device_y -= text_info->height;
@@ -3006,6 +3316,11 @@ ass_render_event(RenderContext *state, ASS_Event *event,
     event_images->event = event;
     event_images->imgs = render_text(state);
 
+    if (collect_metrics) {
+        event_images->metrics.event = event;
+        event_images->metrics.runs = state->run_metrics;
+    }
+
     if (state->border_style == 4)
         add_background(state, event_images);
 
@@ -3045,7 +3360,7 @@ static void setup_shaper(ASS_Shaper *shaper, ASS_Renderer *render_priv)
  */
 static bool
 ass_start_frame(ASS_Renderer *render_priv, ASS_Track *track,
-                long long now)
+                long long now, bool collect_metrics)
 {
     if (!render_priv->settings.frame_width
         && !render_priv->settings.frame_height)
@@ -3089,7 +3404,11 @@ ass_start_frame(ASS_Renderer *render_priv, ASS_Track *track,
     }
     render_priv->par_scale_x = par;
 
-    render_priv->prev_images_root = render_priv->images_root;
+    if (collect_metrics) {
+        ass_frame_unref(render_priv->images_root);
+    } else {
+        render_priv->prev_images_root = render_priv->images_root;
+    }
     render_priv->images_root = NULL;
 
     check_cache_limits(render_priv, &render_priv->cache);
@@ -3165,6 +3484,12 @@ shift_event(ASS_Renderer *render_priv, EventImages *ei, int shift)
         cur = cur->next;
     }
     ei->top += shift;
+
+    ASS_RunMetrics *run = ei->metrics.runs;
+    while (run) {
+        run->pos.y += shift;
+        run = run->next;
+    }
 }
 
 // dir: 1 - move down
@@ -3359,7 +3684,7 @@ static ASS_Image *ass_free_image(ASS_Image *img) {
 }
 
 /**
- * \brief render a frame
+ * \brief render a frame or collect metrics
  * \param priv library handle
  * \param track track
  * \param now current video timestamp (ms)
@@ -3367,14 +3692,16 @@ static ASS_Image *ass_free_image(ASS_Image *img) {
  *        0 if identical, 1 if different positions, 2 if different content.
  *        Can be NULL, in that case no detection is performed.
  */
-ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
-                            long long now, int *detect_change)
+int ass_render_frame_internal(ASS_Renderer *priv, ASS_Track *track,
+                              long long now, int *detect_change, bool collect_metrics)
 {
+    ass_free_metrics(priv);
+
     // init frame
-    if (!ass_start_frame(priv, track, now)) {
+    if (!ass_start_frame(priv, track, now, collect_metrics)) {
         if (detect_change)
             *detect_change = 2;
-        return NULL;
+        return -1;
     }
 
     // render events separately
@@ -3389,7 +3716,8 @@ ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
                     realloc(priv->eimg,
                             priv->eimg_size * sizeof(EventImages));
             }
-            if (ass_render_event(&priv->state, event, priv->eimg + cnt))
+            (priv->eimg + cnt)->metrics = (ASS_Metrics) { 0 };
+            if (ass_render_event(&priv->state, event, priv->eimg + cnt, collect_metrics))
                 cnt++;
         }
     }
@@ -3428,18 +3756,107 @@ ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
     *tail = NULL;
 
     ass_frame_ref(priv->images_root);
+    if (collect_metrics) {
+        ass_frame_unref(priv->images_root);
+        priv->images_root = NULL;
+    }
 
     if (detect_change)
         *detect_change = ass_detect_change(priv);
 
-    // free the previous image list
-    ass_frame_unref(priv->prev_images_root);
-    priv->prev_images_root = NULL;
+    if (!collect_metrics) {
+        // free the previous image list
+        ass_frame_unref(priv->prev_images_root);
+        priv->prev_images_root = NULL;
+    }
 
     if (track->parser_priv->prune_delay >= 0)
         ass_prune_events(track, now - track->parser_priv->prune_delay);
 
-    return priv->images_root;
+    return cnt;
+}
+
+/**
+ * \brief render a frame
+ * \param priv library handle
+ * \param track track
+ * \param now current video timestamp (ms)
+ * \param detect_change a value describing how the new images differ from the previous ones will be written here:
+ *        0 if identical, 1 if different positions, 2 if different content.
+ *        Can be NULL, in that case no detection is performed.
+ */
+ASS_Image *ass_render_frame(ASS_Renderer *priv, ASS_Track *track,
+                            long long now, int *detect_change)
+{
+    int cnt = ass_render_frame_internal(priv, track, now, detect_change, false);
+    return cnt >= 0 ? priv->images_root : NULL;
+}
+
+/**
+ * \brief Get metrics for a frame, producing a list of ASS_Metrics.
+ * \param priv renderer handle
+ * \param track subtitle track
+ * \param now video timestamp in milliseconds
+ */
+ASS_Metrics *ass_get_metrics(ASS_Renderer *priv, ASS_Track *track, long long now)
+{
+    int cnt = ass_render_frame_internal(priv, track, now, NULL, true);
+    if (cnt <= 0)
+        return NULL;
+
+    ASS_Metrics *root = &priv->eimg[0].metrics;
+    ASS_Metrics *cur = root;
+    for (int i = 1; i < cnt; i++) {
+        cur->next = &priv->eimg[i].metrics;
+        cur = cur->next;
+    }
+    cur->next = NULL;
+
+    return root;
+}
+
+
+/**
+ * \brief Free an ASS_Metrics_Outline's memory as well as all the
+ * memory of all ASS_Metrics_Outlines following it.
+ */
+void ass_free_metrics_outlines(ASS_Metrics_Outline *outline)
+{
+    while (outline) {
+        ASS_Metrics_Outline *next = outline->next;
+        ass_metric_outline_free(outline);
+        free(outline);
+        outline = next;
+    }
+}
+
+/**
+ * \brief Free all memory that was allocated for metrics.
+ */
+void ass_free_metrics(ASS_Renderer *priv)
+{
+    if (!priv->eimg_size || !priv->eimg)
+        return;
+
+    // The ASS_Metrics structs themselves are part of EventImages and
+    // are hence allocated and freed by ass_render_frame_internal.
+    // All contained pointers are owned by the ASS_Metrics struct
+    // and need to be freed here.
+    for (ASS_Metrics *metrics = &priv->eimg[0].metrics; metrics;) {
+        ASS_Metrics *next_metrics = metrics->next;
+        ASS_RunMetrics *run = metrics->runs;
+        while (run) {
+            ass_free_metrics_outlines(run->fill);
+            ass_free_metrics_outlines(run->border);
+            ASS_RunMetrics *next = run->next;
+            free(run);
+            run = next;
+        }
+        metrics->event = NULL;
+        metrics->runs = NULL;
+        metrics->next = NULL;
+        metrics = next_metrics;
+    }
 }
 
 /**
diff --git a/libass/ass_render.h b/libass/ass_render.h
index e1eb7d2..dae9494 100644
--- a/libass/ass_render.h
+++ b/libass/ass_render.h
@@ -30,6 +30,7 @@
 #include <hb.h>
 
 #include "ass.h"
+#include "ass_metrics.h"
 #include "ass_font.h"
 #include "ass_bitmap.h"
 #include "ass_cache.h"
@@ -86,6 +87,7 @@ typedef struct {
     int detect_collisions;
     int shift_direction;
     ASS_Event *event;
+    ASS_Metrics metrics;
 } EventImages;
 
 typedef enum {
@@ -215,6 +217,10 @@ struct render_context {
     ASS_Event *event;
     ASS_Style *style;
 
+    bool collect_metrics;
+    ASS_RunMetrics *run_metrics;
+    ASS_RunMetrics *current_run_metrics;
+
     ASS_Font *font;
     double font_size;
     int parsed_tags;
@@ -270,12 +276,12 @@ struct render_context {
     } scroll_direction;         // for EVENT_HSCROLL, EVENT_VSCROLL
     double scroll_shift;
     int scroll_y0, scroll_y1;
+    double scroll_fade_width;
 
     // face properties
     ASS_StringView family;
     unsigned bold;
     unsigned italic;
-    int treat_family_as_pattern;
     int wrap_style;
     int font_encoding;
 
@@ -353,6 +359,9 @@ typedef struct {
 } Rect;
 
 void ass_reset_render_context(RenderContext *state, ASS_Style *style);
+int ass_render_frame_internal(ASS_Renderer *priv, ASS_Track *track, long long now, int *detect_change, bool collect_metrics);
+void ass_outline_apply_transform(ASS_Outline *outline, BitmapHashKey *k);
+void ass_free_metrics(ASS_Renderer *priv);
 void ass_frame_ref(ASS_Image *img);
 void ass_frame_unref(ASS_Image *img);
 ASS_Vector ass_layout_res(ASS_Renderer *render_priv);
diff --git a/libass/ass_types.h b/libass/ass_types.h
index 4d9e0ec..c59ce15 100644
--- a/libass/ass_types.h
+++ b/libass/ass_types.h
@@ -125,11 +125,15 @@ typedef struct ass_style {
     int Alignment; // use `VALIGN_* | HALIGN_*` as value
     int MarginL;
     int MarginR;
-    int MarginV;
+    union {
+        int MarginV;
+        int MarginT;
+    };
+    int MarginB; // Applies only if set to a non-INT_MIN value
     int Encoding;
-    int treat_fontname_as_pattern; // does nothing (left in place for ABI-compatibility)
     double Blur; // sets a default \blur for the event; same values as \blur
     int Justify; // sets text justification independent of event alignment; use ASS_JUSTIFY_*
+    int RelativeTo; // 0 for dialogue, 1 for typesetting, 2 for auto
 } ASS_Style;
 
 
@@ -147,7 +151,11 @@ typedef struct ass_event {
     char *Name;
     int MarginL;
     int MarginR;
-    int MarginV;
+    union {
+        int MarginV;
+        int MarginT;
+    };
+    int MarginB;
     char *Effect;
     char *Text;
 
@@ -280,7 +288,8 @@ typedef struct ass_track {
     enum {
         TRACK_TYPE_UNKNOWN = 0,
         TRACK_TYPE_ASS,
-        TRACK_TYPE_SSA
+        TRACK_TYPE_SSA,
+        TRACK_TYPE_ASS2,
     } track_type;
 
     // Script header fields
@@ -305,4 +314,8 @@ typedef struct ass_track {
     // New fields can be added here in new ABI-compatible library releases.
 } ASS_Track;
 
+typedef struct {
+    double x, y;
+} ASS_DVector;
+
 #endif /* LIBASS_TYPES_H */
diff --git a/libass/ass_utils.c b/libass/ass_utils.c
index 01bb8a8..fbaca7a 100644
--- a/libass/ass_utils.c
+++ b/libass/ass_utils.c
@@ -56,6 +56,75 @@ char *ass_strndup_fallback(const char *s, size_t n)
 }
 #endif
 
+#if (defined(_WIN32) || defined(__APPLE__) || defined(__linux__) || defined(__FreeBSD__)) && \
+    (defined(__i386__) || defined(__x86_64__) || defined(__arm__) || defined(__aarch64__)) && \
+    (defined(__GNUC__))
+#define ALIGNED_OVERREADS_SAFE 1
+#endif
+
+#if __has_attribute(vector_size) && defined(ALIGNED_OVERREADS_SAFE)
+
+#if __has_builtin(__builtin_ia32_pmovmskb128) && defined(__SSE2__)
+#define VEC_ANY_ZERO(x) (__builtin_ia32_pmovmskb128((x) == 0) != 0)
+#define VECTOR_SIZE 16
+#elif __has_builtin(__builtin_neon_vminvq_u8) && defined(__ARM_NEON__)
+#define VEC_ANY_ZERO(x) (__builtin_neon_vminvq_u8(x) == 0)
+#define VECTOR_SIZE 16
+#elif __has_builtin(__builtin_aarch64_reduc_umin_scal_v16qi_uu) && defined(__ARM_NEON__)
+#define VEC_ANY_ZERO(x) (__builtin_aarch64_reduc_umin_scal_v16qi_uu(x) == 0)
+#define VECTOR_SIZE 16
+#endif
+
+#ifdef VEC_ANY_ZERO
+typedef unsigned char CheckVector __attribute__ ((vector_size(VECTOR_SIZE)));
+#endif
+
+#endif // __has_attribute(vector_size)
+
+#if ALIGNED_OVERREADS_SAFE && !defined(VEC_ANY_ZERO)
+
+typedef size_t CheckVector;
+#define USE_SCALAR_TYPE 1
+#define LOWS ((size_t)-1 / 0xFF)
+#define HIGHS (LOWS * 0x80)
+#define VEC_ANY_ZERO(vec) (((vec) - LOWS) & ~(vec) & HIGHS)
+#define BROADCAST(c) ((c) * LOWS)
+
+#endif // defined(ALIGNED_OVERREADS_SAFE) && !defined(VEC_ANY_ZERO)
+
+#if ALIGNED_OVERREADS_SAFE && __has_attribute(no_sanitize)
+__attribute__((no_sanitize("address")))
+#endif
+char *ass_strchrnul(const char *s, int c)
+{
+#if ALIGNED_OVERREADS_SAFE && !defined(__GLIBC__)
+    for (; (uintptr_t)s % sizeof(CheckVector); s++) {
+        if (!*s || *s == c)
+            return (char*)s;
+    }
+
+    for (;; s += sizeof(CheckVector)) {
+        CheckVector vec;
+        memcpy(&vec, s, sizeof(vec));
+#ifdef USE_SCALAR_TYPE
+        if (VEC_ANY_ZERO(vec) || VEC_ANY_ZERO(vec ^ BROADCAST(c)))
+            break;
+#else
+        if (VEC_ANY_ZERO(vec & (vec != (unsigned char)c)))
+            break;
+#endif
+    }
+#endif
+#ifdef HAVE_STRCHRNUL
+    return strchrnul(s, c);
+#else // HAVE_STRCHRNUL
+    while (*s && *s != c)
+        s++;
+
+    return (char*)s;
+#endif // HAVE_STRCHRNUL
+}
+
 void *ass_aligned_alloc(size_t alignment, size_t size, bool zero)
 {
     assert(!(alignment & (alignment - 1))); // alignment must be power of 2
diff --git a/libass/ass_utils.h b/libass/ass_utils.h
index 5b7cc7a..f6d60d3 100644
--- a/libass/ass_utils.h
+++ b/libass/ass_utils.h
@@ -93,6 +93,8 @@ void ass_aligned_free(void *ptr);
 void *ass_realloc_array(void *ptr, size_t nmemb, size_t size);
 void *ass_try_realloc_array(void *ptr, size_t nmemb, size_t size);
 
+char *ass_strchrnul(const char *s, int c);
+
 /**
  * Reallocate the array in ptr to at least count elements. For example, if
  * you do "int *ptr = NULL; ASS_REALLOC_ARRAY(ptr, 5)", you can access ptr[0]
diff --git a/libass/c/c_shift.c b/libass/c/c_shift.c
new file mode 100644
index 0000000..65d5fb1
--- /dev/null
+++ b/libass/c/c_shift.c
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2024-2026 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "config.h"
+#include "ass_compat.h"
+#include "ass_utils.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+
+#define ALIGNMENT  16
+
+/**
+ * \brief Shift a bitmap by the fraction of a pixel in negative x and y direction
+ * expressed in 26.6 fixed point
+ */
+void ass_shift_c(uint8_t *restrict buf, ptrdiff_t stride,
+                 size_t width, size_t height, int shift_x, int shift_y)
+{
+    ASSUME(!((uintptr_t) buf % ALIGNMENT) && !(stride % ALIGNMENT));
+    ASSUME(!(shift_x & ~63) && !(shift_y & ~63));
+    ASSUME(width > 1 && height > 1);
+
+    int16_t sy = shift_y << 9;
+    for (int32_t y = 1; y < height; y++) {
+        for (int32_t x = 1; x < width; x++) {
+            int16_t tmp1 = 64 * buf[x - 1] - (buf[x - 1] - buf[x]) * (int16_t) shift_x;
+            int16_t tmp2 = 64 * buf[x + stride - 1] - (buf[x + stride - 1] - buf[x + stride]) * (int16_t) shift_x;
+            buf[x - 1] = (int16_t) (tmp1 - (int16_t) ((int16_t) (2 * (tmp1 - tmp2)) * (int32_t) sy >> 16) + 32) >> 6;
+        }
+        int16_t tmp1 = 64 * buf[width - 1] - buf[width - 1] * (int16_t) shift_x;
+        int16_t tmp2 = 64 * buf[width + stride - 1] - buf[width + stride - 1] * (int16_t) shift_x;
+        buf[width - 1] = (int16_t) (tmp1 - (int16_t) ((int16_t) (2 * (tmp1 - tmp2)) * (int32_t) sy >> 16) + 32) >> 6;
+        buf += stride;
+    }
+    for (int32_t x = 1; x < width; x++) {
+        int16_t tmp1 = 64 * buf[x - 1] - (buf[x - 1] - buf[x]) * (int16_t) shift_x;
+        buf[x - 1] = (int16_t) (tmp1 - (int16_t) ((int16_t) (2 * tmp1) * (int32_t) sy >> 16) + 32) >> 6;
+    }
+    int16_t tmp1 = 64 * buf[width - 1] - buf[width - 1] * (int16_t) shift_x;
+    buf[width - 1] = (int16_t) (tmp1 - (int16_t) ((int16_t) (2 * tmp1) * (int32_t) sy >> 16) + 32) >> 6;
+}
diff --git a/libass/libass.sym b/libass/libass.sym
index 4cc6215..f8442bf 100644
--- a/libass/libass.sym
+++ b/libass/libass.sym
@@ -48,3 +48,4 @@ ass_malloc
 ass_free
 ass_prune_events
 ass_configure_prune
+ass_get_metrics
diff --git a/libass/meson.build b/libass/meson.build
index 9eda958..46a7201 100644
--- a/libass/meson.build
+++ b/libass/meson.build
@@ -29,6 +29,7 @@ src_x86 = files(
     'x86/blur.asm',
     'x86/cpuid.asm',
     'x86/rasterizer.asm',
+    'x86/shift.asm',
 )
 src_aarch64 = files(
     'aarch64/asm.S',
@@ -36,6 +37,7 @@ src_aarch64 = files(
     'aarch64/blend_bitmaps.S',
     'aarch64/blur.S',
     'aarch64/rasterizer.S',
+    'aarch64/shift.S',
 )
 src_fontconfig = files('ass_fontconfig.c')
 src_directwrite = files('ass_directwrite.c')
@@ -46,6 +48,7 @@ libass_src = files(
     'c/c_blend_bitmaps.c',
     'c/c_blur.c',
     'c/c_rasterizer.c',
+    'c/c_shift.c',
     'ass.c',
     'ass_arabic_charmap.c',
     'ass_bitmap.c',
@@ -70,7 +73,7 @@ libass_src = files(
 
 libass_link_with = []
 
-libass_headers = files('ass.h', 'ass_types.h')
+libass_headers = files('ass.h', 'ass_types.h', 'ass_metrics.h')
 
 if fontconfig
     libass_src += src_fontconfig
diff --git a/libass/x86/shift.asm b/libass/x86/shift.asm
new file mode 100644
index 0000000..584be0d
--- /dev/null
+++ b/libass/x86/shift.asm
@@ -0,0 +1,280 @@
+;******************************************************************************
+;* shift.asm: SSE2/AVX2 subpixel bitmap shift
+;******************************************************************************
+;* Copyright (C) 2026 libass contributors
+;*
+;* This file is part of libass.
+;*
+;* Permission to use, copy, modify, and distribute this software for any
+;* purpose with or without fee is hereby granted, provided that the above
+;* copyright notice and this permission notice appear in all copies.
+;*
+;* THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+;* WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+;* MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+;* ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+;* WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+;* ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+;* OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+;******************************************************************************
+
+%include "x86/utils.asm"
+
+SECTION .text
+
+;------------------------------------------------------------------------------
+; SHIFT_X 1:m_dst, 2:m_src1, 3:m_src2,
+;         4:m_tmp, 5:m_mul(64), 6:r_mul(32), 6:mode
+; Perform subpixel shift along X axis
+;------------------------------------------------------------------------------
+
+%macro SHIFT_X 7
+%ifidn %7, even
+    %assign %%mode 0
+%elifidn %7, odd
+    %assign %%mode 1
+%elifidn %7, last
+    %assign %%mode 2
+%else
+    %error "even/odd/last expected"
+%endif
+
+%if %%mode == 2 && mmsize != 32
+    psrldq m%1, m%2, 2
+%else
+    PALIGNR m%1,m%3,m%2, m%4, 2
+%endif
+%if %%mode == 1 && mmsize == 32
+    mova m%3, m%4
+%endif
+    psubw m%1, m%2
+%if ARCH_X86_64
+    pmullw m%1, m%5
+%else
+    BCASTD %4, %6d
+    pmullw m%1, m%4
+%endif
+%if %%mode == 0 && mmsize == 32
+    psllw m%4, m%2, 6
+    paddw m%1, m%4
+%else
+    psllw m%2, 6
+    paddw m%1, m%2
+%endif
+%endmacro
+
+;------------------------------------------------------------------------------
+; SHIFT_Y 1:m_dst/m_src1, 2:m_src2,
+;         3:m_tmp, 4:m_mul(64), 5:r_mul(32), 6:m_w32(64), 7:mode
+; Perform subpixel shift along Y axis
+;------------------------------------------------------------------------------
+
+%macro SHIFT_Y 6-7
+%ifidn %7, last
+    paddw m%3, m%1, m%1
+%else
+    psubw m%3, m%1, m%2
+    paddw m%3, m%3
+%endif
+%if ARCH_X86_64
+    pmulhw m%3, m%4
+    psubw m%1, m%3
+    paddw m%1, m%6
+%else
+    BCASTD %2, %5d
+    pmulhw m%3, m%2
+    psubw m%1, m%3
+    pcmpeqw m%2, m%2
+    psllw m%2, 5
+    psubw m%1, m%2
+%endif
+    psrlw m%1, 6
+%endmacro
+
+;------------------------------------------------------------------------------
+; SHIFT
+; void ass_shift_c(uint8_t *buf, ptrdiff_t stride,
+;                  size_t width, size_t height, int shift_x, int shift_y);
+;------------------------------------------------------------------------------
+
+%macro SHIFT 0
+%if ARCH_X86_64
+cglobal shift, 6,6,13
+    DECLARE_REG_TMP 5
+%else
+cglobal shift, 6,7,8
+    DECLARE_REG_TMP 6
+    sub r3, 1
+%endif
+
+    lea r0, [r0 + r2]
+    neg r2
+    imul r3, r1
+    add r3, r0
+
+%if ARCH_X86_64
+    pxor m9, m9
+    BCASTW 10, r4d
+    shl r5d, 9
+    BCASTW 11, r5d
+    mov r4d, 32
+    BCASTW 12, r4d
+    lea r4, [r0 + r1]
+%else
+    imul r4d, 0x10001
+    imul r5d, 0x10001 << 9
+%endif
+
+.height_loop:
+    mov t0, r2
+    mova m0, [r0 + t0]
+%if ARCH_X86_64
+    punpcklbw m2, m0, m9
+    punpckhbw m4, m0, m9
+    mova m0, [r4 + t0]
+    punpcklbw m3, m0, m9
+    punpckhbw m5, m0, m9
+%else
+    pxor m7, m7
+    add r0, r1
+    punpcklbw m2, m0, m7
+    punpckhbw m4, m0, m7
+    mova m0, [r0 + t0]
+    sub r0, r1
+    punpcklbw m3, m0, m7
+    punpckhbw m5, m0, m7
+%endif
+    jmp .loop_entry
+
+.width_loop:
+    mova m0, [r0 + t0]
+%if ARCH_X86_64
+
+%if mmsize == 32
+    punpcklbw m7, m0, m9
+    vperm2i128 m2, m2, m7, 0x21
+%else
+    punpcklbw m2, m0, m9
+%endif
+    SHIFT_X 6,4,2, 7,10,r4, odd
+    punpckhbw m4, m0, m9
+
+    mova m0, [r4 + t0]
+%if mmsize == 32
+    punpcklbw m8, m0, m9
+    vperm2i128 m3, m3, m8, 0x21
+%else
+    punpcklbw m3, m0, m9
+%endif
+    SHIFT_X 7,5,3, 8,10,r4, odd
+    punpckhbw m5, m0, m9
+
+%else
+
+    add r0, r1
+    pxor m6, m6
+%if mmsize == 32
+    punpcklbw m7, m0, m6
+    vperm2i128 m2, m2, m7, 0x21
+%else
+    punpcklbw m2, m0, m6
+%endif
+    SHIFT_X 6,4,2, 7,10,r4, odd
+
+    mova m7, [r0 + t0]
+    sub r0, r1
+    punpckhqdq m4, m0, m7
+    pxor m0, m0
+%if mmsize == 32
+    punpcklbw m0, m7, m0
+    vperm2i128 m3, m3, m0, 0x21
+%else
+    punpcklbw m3, m7, m0
+%endif
+    SHIFT_X 7,5,3, 0,10,r4, odd
+    pxor m0, m0
+    punpckhbw m5, m4, m0
+    punpcklbw m4, m4, m0
+
+%endif
+
+    SHIFT_Y 6,7, 0,11,r5,12
+    packuswb m1, m6
+    mova [r0 + t0 - mmsize], m1
+
+.loop_entry:
+    SHIFT_X 1,2,4, 0,10,r4, even
+    SHIFT_X 7,3,5, 0,10,r4, even
+    SHIFT_Y 1,7, 0,11,r5,12
+
+    add t0, mmsize
+    jnc .width_loop
+
+%if mmsize == 32
+    vperm2i128 m2, m2, m2, 0x81
+    vperm2i128 m3, m3, m2, 0x81
+%endif
+    SHIFT_X 6,4,2, 0,10,r4, last
+    SHIFT_X 7,5,3, 0,10,r4, last
+    SHIFT_Y 6,7, 0,11,r5,12
+    packuswb m1, m6
+    mova [r0 + t0 - mmsize], m1
+
+%if ARCH_X86_64
+    mov r0, r4
+    add r4, r1
+    cmp r4, r3
+%else
+    add r0, r1
+    cmp r0, r3
+%endif
+    jl .height_loop
+
+    mov t0, r2
+    mova m0, [r0 + t0]
+%if !ARCH_X86_64
+    pxor m3, m3
+    %define m9 m3
+%endif
+    punpcklbw m2, m0, m9
+    punpckhbw m4, m0, m9
+    jmp .last_loop_entry
+
+.last_width_loop:
+    mova m0, [r0 + t0]
+%if mmsize == 32
+    punpcklbw m7, m0, m9
+    vperm2i128 m2, m2, m7, 0x21
+%else
+    punpcklbw m2, m0, m9
+%endif
+    SHIFT_X 6,4,2, 7,10,r4, odd
+    punpckhbw m4, m0, m9
+
+    SHIFT_Y 6,7, 0,11,r5,12, last
+    packuswb m1, m6
+    mova [r0 + t0 - mmsize], m1
+
+.last_loop_entry:
+    SHIFT_X 1,2,4, 0,10,r4, even
+    SHIFT_Y 1,7, 0,11,r5,12, last
+
+    add t0, mmsize
+    jnc .last_width_loop
+
+%if mmsize == 32
+    vperm2i128 m2, m2, m2, 0x81
+%endif
+    SHIFT_X 6,4,2, 0,10,r4, last
+    SHIFT_Y 6,7, 0,11,r5,12, last
+    packuswb m1, m6
+    mova [r0 + t0 - mmsize], m1
+    RET
+%endmacro
+
+INIT_XMM sse2
+SHIFT
+INIT_XMM ssse3
+SHIFT
+INIT_YMM avx2
+SHIFT
diff --git a/meson.build b/meson.build
index 91a0435..56e9a59 100644
--- a/meson.build
+++ b/meson.build
@@ -60,7 +60,7 @@ add_project_arguments(
 
 # Configuration
 
-str_check_functions = ['strdup', 'strndup']
+str_check_functions = ['strdup', 'strndup', 'strchrnul']
 
 foreach name : str_check_functions
     if (
