diff --git a/libass/ass_parse.c b/libass/ass_parse.c
index e85fc5e..99833c5 100644
--- a/libass/ass_parse.c
+++ b/libass/ass_parse.c
@@ -966,6 +966,8 @@ void ass_apply_transition_effects(RenderContext *state)
         delay = ((int) FFMAX(delay / scale_x, 1)) * scale_x;
         state->scroll_shift =
             (render_priv->time - event->Start) / delay;
+        state->scroll_fade_width =
+            cnt >= 3 ? ((int) (v[2] * scale_x)) / scale_x : 0;
         state->evt_type |= EVENT_HSCROLL;
         state->detect_collisions = 0;
         state->wrap_style = 2;
@@ -1005,6 +1007,8 @@ void ass_apply_transition_effects(RenderContext *state)
         }
         state->scroll_y0 = y0;
         state->scroll_y1 = y1;
+        state->scroll_fade_width =
+            cnt >= 4 ? ((int) (v[3] * scale_y)) / scale_y : 0;
         state->evt_type |= EVENT_VSCROLL;
         state->detect_collisions = 0;
     }
diff --git a/libass/ass_render.c b/libass/ass_render.c
index 15bc813..40c1933 100644
--- a/libass/ass_render.c
+++ b/libass/ass_render.c
@@ -237,6 +237,20 @@ static ASS_Image *my_draw_bitmap(unsigned char *bitmap, int bitmap_w,
     return &img->result;
 }
 
+static inline ASS_Image *ass_image_borrow_copy(const ASS_Image *img)
+{
+    ASS_Image *nimg = my_draw_bitmap(img->bitmap, img->w, img->h, img->stride,
+                                     img->dst_x, img->dst_y, img->color,
+                                     ((ASS_ImagePriv *) img)->source);
+    if (!nimg)
+        return NULL;
+    nimg->next = img->next;
+    nimg->type = img->type;
+    // Borrow the buffer from the original ASS_Image
+    ((ASS_ImagePriv *) nimg)->buffer = NULL;
+    return nimg;
+}
+
 /**
  * \brief Mapping between script and screen coordinates
  */
@@ -694,6 +708,135 @@ static inline size_t bitmap_size(const Bitmap *bm)
     return bm->stride * bm->h;
 }
 
+static void apply_hscroll_fade(RenderContext *state, ASS_Image *head)
+{
+    double fade_width = state->scroll_fade_width * state->screen_scale_x;
+    if (fade_width <= 0)
+        return;
+
+    ASS_Renderer *render_priv = state->renderer;
+
+    for (ASS_Image *cur = head; cur; cur = cur->next) {
+        int event_start = cur->dst_x - render_priv->settings.left_margin;
+        int event_end = event_start + cur->w;
+
+        double leading_fade_start = 0.5 * state->blur_scale_x - 0.5;
+        int leading_fade_start_int = floor(leading_fade_start) + 1;
+        double leading_fade_end = leading_fade_start + fade_width;
+        int leading_fade_end_int = ceil(leading_fade_end);
+
+        // int trailing_fade_end_int = render_priv->frame_content_width;
+        // double trailing_fade_end = trailing_fade_end_int + leading_fade_start;
+        double trailing_fade_end = render_priv->frame_content_width + leading_fade_start;
+        int trailing_fade_end_int = ceil(trailing_fade_end);
+        double trailing_fade_start = trailing_fade_end - fade_width;
+        int trailing_fade_start_int = floor(trailing_fade_start) + 1;
+
+        int full_a = _a(cur->color);
+        cur->color |= 0xFF;  // starts out fully transparent
+
+        for (int x = FFMAX(event_start, leading_fade_start_int); x < event_end && x <= trailing_fade_end_int; ) {
+            double exact_a = 0xFF - full_a;
+            if (x < leading_fade_end_int)
+                exact_a *= (x - leading_fade_start) / fade_width;
+            if (x >= trailing_fade_start_int)
+                exact_a *= FFMAX(trailing_fade_end - x, 0) / fade_width;
+
+            int a = 0xFF - lround(exact_a);
+            if (a != _a(cur->color)) {
+                if (x > event_start) {
+                    ASS_Image *nimg = ass_image_borrow_copy(cur);
+                    nimg->dst_x = x + render_priv->settings.left_margin;
+                    cur->w = nimg->dst_x - cur->dst_x;
+                    nimg->w -= cur->w;
+                    nimg->bitmap += cur->w;
+                    cur->next = nimg;
+                    cur = nimg;
+                }
+                cur->color = (cur->color & ~0xFF) | a;
+            }
+
+            if (x >= leading_fade_end_int && x < trailing_fade_start_int)
+                x = trailing_fade_start_int;
+            else
+                x++;
+        }
+    }
+}
+
+static void apply_vscroll_fade(RenderContext *state, ASS_Image *head)
+{
+    double fade_height = state->scroll_fade_width * state->screen_scale_y;
+    if (fade_height <= 0)
+        return;
+
+    ASS_Renderer *render_priv = state->renderer;
+
+    for (ASS_Image *cur = head; cur; cur = cur->next) {
+        int event_start = cur->dst_y - render_priv->settings.top_margin;
+        int event_end = event_start + cur->h;
+
+        ASS_Vector layout_res = ass_layout_res(render_priv);
+        double scale_y = ((double) layout_res.y) / render_priv->track->PlayResY;
+        double half_pixel = 0.5 * state->blur_scale_y - 0.5;
+
+        double leading_fade_start =
+            ((int32_t) (state->scroll_y0 * scale_y * 8) >> 3) * state->blur_scale_y + half_pixel;
+        int leading_fade_start_int = floor(leading_fade_start) + 1;
+        double leading_fade_end = leading_fade_start + fade_height;
+        int leading_fade_end_int = ceil(leading_fade_end);
+
+        double trailing_fade_end =
+            ((int32_t) (state->scroll_y1 * scale_y * 8) >> 3) * state->blur_scale_y + half_pixel;
+        int trailing_fade_end_int = ceil(trailing_fade_end);
+        double trailing_fade_start = trailing_fade_end - fade_height;
+        int trailing_fade_start_int = floor(trailing_fade_start) + 1;
+
+        int full_a = _a(cur->color);
+        cur->color |= 0xFF;  // starts out fully transparent
+
+        for (int y = FFMAX(event_start, leading_fade_start_int); y < event_end && y <= trailing_fade_end_int; ) {
+            double exact_a = 0xFF - full_a;
+            if (y < leading_fade_end_int)
+                exact_a *= (y - leading_fade_start) / fade_height;
+            if (y >= trailing_fade_start_int)
+                exact_a *= FFMAX(trailing_fade_end - y, 0) / fade_height;
+
+            int a = 0xFF - lround(exact_a);
+            if (a != _a(cur->color)) {
+                if (y > event_start) {
+                    ASS_Image *nimg = ass_image_borrow_copy(cur);
+                    nimg->dst_y = y + render_priv->settings.top_margin;
+                    cur->h = nimg->dst_y - cur->dst_y;
+                    nimg->h -= cur->h;
+                    nimg->bitmap += cur->h * cur->stride;
+                    cur->next = nimg;
+                    cur = nimg;
+                }
+                cur->color = (cur->color & ~0xFF) | a;
+            }
+
+            if (y >= leading_fade_end_int && y < trailing_fade_start_int)
+                y = trailing_fade_start_int;
+            else
+                y++;
+        }
+    }
+}
+
+/**
+ * Iterate through a list of bitmaps and apply the fade gradient
+ * from a Banner or Scroll effect, if applicable. Each image is
+ * replaced by multiple cropped copies with different alphas.
+ */
+static void apply_scroll_fade(RenderContext *state, ASS_Image *head)
+{
+    if (state->evt_type & EVENT_HSCROLL)
+        apply_hscroll_fade(state, head);
+    else if (state->evt_type & EVENT_VSCROLL)
+        apply_vscroll_fade(state, head);
+}
+
 /**
  * Iterate through a list of bitmaps and blend with clip vector, if
  * applicable. The blended bitmaps are added to a free list which is freed
@@ -881,6 +1024,7 @@ static ASS_Image *render_text(RenderContext *state)
 
     *tail = 0;
     blend_vector_clip(state, head);
+    apply_scroll_fade(state, head);
 
     return head;
 }
diff --git a/libass/ass_render.h b/libass/ass_render.h
index f767e03..dae9494 100644
--- a/libass/ass_render.h
+++ b/libass/ass_render.h
@@ -276,6 +276,7 @@ struct render_context {
     } scroll_direction;         // for EVENT_HSCROLL, EVENT_VSCROLL
     double scroll_shift;
     int scroll_y0, scroll_y1;
+    double scroll_fade_width;
 
     // face properties
     ASS_StringView family;
